// @angular-three/core/esm2020/attributes/lib/uint8-buffer-attribute/uint8-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtUint8BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint8BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint8BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint8BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint8BufferAttribute, selector: "ngt-uint8-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint8BufferAttribute,
        },
    ], exportAs: ["ngtUint8BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint8-buffer-attribute',
                    exportAs: 'ngtUint8BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint8BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtUint8BufferAttributeModule {
}
NgtUint8BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint8BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, declarations: [NgtUint8BufferAttribute], exports: [NgtUint8BufferAttribute] });
NgtUint8BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint8BufferAttribute],
                    exports: [NgtUint8BufferAttribute],
                }]
        }] });
