// @angular-three/core/esm2020/meshes/mesh/mesh.directive.mjs
import { NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_OBJECT_TYPE, NgtCommonMesh, NgtMaterialGeometryControllerModule, } from '@angular-three/core';
import { Directive, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMesh extends NgtCommonMesh {
}
NgtMesh.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMesh, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMesh.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMesh, selector: "ngt-mesh", providers: [
        { provide: NgtCommonMesh, useExisting: NgtMesh },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        { provide: NGT_OBJECT_TYPE, useValue: THREE.Mesh },
    ], exportAs: ["ngtMesh"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMesh, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh',
                    exportAs: 'ngtMesh',
                    providers: [
                        { provide: NgtCommonMesh, useExisting: NgtMesh },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        { provide: NGT_OBJECT_TYPE, useValue: THREE.Mesh },
                    ],
                }]
        }] });
export class NgtMeshModule {
}
NgtMeshModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshModule, declarations: [NgtMesh], exports: [NgtMesh, NgtMaterialGeometryControllerModule] });
NgtMeshModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMesh],
                    exports: [NgtMesh, NgtMaterialGeometryControllerModule],
                }]
        }] });
