// @angular-three/core/esm2020/geometries/lib/lathe-geometry/lathe-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLatheGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.LatheGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtLatheGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLatheGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLatheGeometry, selector: "ngt-lathe-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtLatheGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtLatheGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-lathe-geometry',
                    exportAs: 'ngtLatheGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtLatheGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtLatheGeometryModule {
}
NgtLatheGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLatheGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometryModule, declarations: [NgtLatheGeometry], exports: [NgtLatheGeometry] });
NgtLatheGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLatheGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLatheGeometry],
                    exports: [NgtLatheGeometry],
                }]
        }] });
