// @angular-three/core/esm2020/lib/three/attribute.mjs
var _NgtAttribute_attributeArgs, _NgtAttribute_attribute, _NgtAttribute_defaultValue;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, Input, NgZone, Optional, } from '@angular/core';
import { NgtGeometry } from './geometry';
import * as i0 from "@angular/core";
import * as i1 from "./geometry";
export class NgtAttribute {
    constructor(ngZone, geometryDirective) {
        this.ngZone = ngZone;
        this.geometryDirective = geometryDirective;
        _NgtAttribute_attributeArgs.set(this, []);
        _NgtAttribute_attribute.set(this, void 0);
        _NgtAttribute_defaultValue.set(this, void 0);
    }
    set attributeArgs(v) {
        __classPrivateFieldSet(this, _NgtAttribute_attributeArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            this.init();
        });
    }
    ngOnChanges() {
        this.ngZone.runOutsideAngular(() => {
            if (this.attribute) {
                this.attribute.needsUpdate = true;
            }
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!this.attribute) {
                this.init();
            }
        });
    }
    init() {
        if (this.geometryDirective && this.attach) {
            __classPrivateFieldSet(this, _NgtAttribute_attribute, new this.attributeType(...__classPrivateFieldGet(this, _NgtAttribute_attributeArgs, "f")), "f");
            if (this.attribute) {
                __classPrivateFieldSet(this, _NgtAttribute_defaultValue, this.geometryDirective.geometry.attributes[this.attach], "f");
                this.geometryDirective.geometry.setAttribute(this.attach, this.attribute);
            }
        }
    }
    ngOnDestroy() {
        this.ngZone.runOutsideAngular(() => {
            if (this.geometryDirective && this.attach) {
                if (__classPrivateFieldGet(this, _NgtAttribute_defaultValue, "f") !== undefined) {
                    this.geometryDirective.geometry.setAttribute(this.attach, __classPrivateFieldGet(this, _NgtAttribute_defaultValue, "f"));
                }
                else {
                    this.geometryDirective.geometry.deleteAttribute(this.attach);
                }
            }
        });
    }
    get attribute() {
        return __classPrivateFieldGet(this, _NgtAttribute_attribute, "f");
    }
}
_NgtAttribute_attributeArgs = new WeakMap(), _NgtAttribute_attribute = new WeakMap(), _NgtAttribute_defaultValue = new WeakMap();
NgtAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAttribute, deps: [{ token: i0.NgZone }, { token: i1.NgtGeometry, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
NgtAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAttribute, inputs: { attach: "attach" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAttribute, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.NgtGeometry, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { attach: [{
                type: Input
            }] } });
