// @angular-three/core/esm2020/lib/three/audio.mjs
var _NgtCommonAudio_audioArgs, _NgtCommonAudio_audio;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, Input, NgZone, Output, } from '@angular/core';
import * as THREE from 'three';
import { NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, } from '../controllers/object-3d.controller';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/object-3d.controller";
export class NgtCommonAudio {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtCommonAudio_audioArgs.set(this, []);
        _NgtCommonAudio_audio.set(this, void 0);
        objectController.initFn = () => {
            if (!this.listener) {
                throw new Error('Cannot initialize Audio without an AudioListener');
            }
            __classPrivateFieldSet(this, _NgtCommonAudio_audio, new this.audioType(this.listener, ...__classPrivateFieldGet(this, _NgtCommonAudio_audioArgs, "f")), "f");
            return __classPrivateFieldGet(this, _NgtCommonAudio_audio, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.audio);
        };
    }
    set audioArgs(v) {
        __classPrivateFieldSet(this, _NgtCommonAudio_audioArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!__classPrivateFieldGet(this, _NgtCommonAudio_audio, "f")) {
                this.objectController.init();
            }
        });
    }
    get audio() {
        return __classPrivateFieldGet(this, _NgtCommonAudio_audio, "f");
    }
}
_NgtCommonAudio_audioArgs = new WeakMap(), _NgtCommonAudio_audio = new WeakMap();
NgtCommonAudio.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonAudio, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtCommonAudio.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCommonAudio, inputs: { listener: "listener" }, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonAudio, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { listener: [{
                type: Input
            }], ready: [{
                type: Output
            }] } });
