// @angular-three/core/esm2020/lib/three/camera.mjs
var _NgtCommonCamera_cameraArgs, _NgtCommonCamera_camera;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, NgZone, Output, } from '@angular/core';
import { NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, } from '../controllers/object-3d.controller';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/object-3d.controller";
export class NgtCommonCamera {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtCommonCamera_cameraArgs.set(this, []);
        _NgtCommonCamera_camera.set(this, void 0);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtCommonCamera_camera, new this.cameraType(...__classPrivateFieldGet(this, _NgtCommonCamera_cameraArgs, "f")), "f");
            return __classPrivateFieldGet(this, _NgtCommonCamera_camera, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.camera);
        };
    }
    set cameraArgs(v) {
        __classPrivateFieldSet(this, _NgtCommonCamera_cameraArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!__classPrivateFieldGet(this, _NgtCommonCamera_camera, "f")) {
                this.objectController.init();
            }
        });
    }
    get camera() {
        return __classPrivateFieldGet(this, _NgtCommonCamera_camera, "f");
    }
}
_NgtCommonCamera_cameraArgs = new WeakMap(), _NgtCommonCamera_camera = new WeakMap();
NgtCommonCamera.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonCamera, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtCommonCamera.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCommonCamera, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonCamera, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
