// @angular-three/core/esm2020/lib/three/curve.mjs
var _NgtCurve_instances, _NgtCurve_curveArgs, _NgtCurve_curve, _NgtCurve_init;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, Input, NgZone, Optional } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./geometry";
export class NgtCurve {
    constructor(ngZone, geometryDirective) {
        this.ngZone = ngZone;
        this.geometryDirective = geometryDirective;
        _NgtCurve_instances.add(this);
        _NgtCurve_curveArgs.set(this, []);
        _NgtCurve_curve.set(this, void 0);
    }
    set curveArgs(v) {
        __classPrivateFieldSet(this, _NgtCurve_curveArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            __classPrivateFieldGet(this, _NgtCurve_instances, "m", _NgtCurve_init).call(this);
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!this.curve) {
                __classPrivateFieldGet(this, _NgtCurve_instances, "m", _NgtCurve_init).call(this);
            }
        });
    }
    get curve() {
        return __classPrivateFieldGet(this, _NgtCurve_curve, "f");
    }
}
_NgtCurve_curveArgs = new WeakMap(), _NgtCurve_curve = new WeakMap(), _NgtCurve_instances = new WeakSet(), _NgtCurve_init = function _NgtCurve_init() {
    __classPrivateFieldSet(this, _NgtCurve_curve, new this.curveType(...__classPrivateFieldGet(this, _NgtCurve_curveArgs, "f")), "f");
    if (this.curve && this.geometryDirective) {
        const points = this.curve.getPoints(this.divisions);
        this.geometryDirective.geometry.setFromPoints(points);
    }
};
NgtCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCurve, deps: [{ token: i0.NgZone }, { token: i1.NgtGeometry, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
NgtCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCurve, inputs: { divisions: "divisions" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCurve, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.NgtGeometry, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { divisions: [{
                type: Input
            }] } });
