// @angular-three/core/esm2020/lib/three/geometry.mjs
var _NgtGeometry_instances, _NgtGeometry_geometryArgs, _NgtGeometry_init, _NgtGeometry_assign, _NgtGeometry_construct, _NgtGeometry_geometry;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, NgZone, Output, } from '@angular/core';
import { requestAnimationFrame } from '@rx-angular/cdk/zone-less';
import { NGT_OBJECT_3D } from '../di/object3d';
import * as i0 from "@angular/core";
export class NgtGeometry {
    constructor(ngZone, parentObject) {
        this.ngZone = ngZone;
        this.parentObject = parentObject;
        _NgtGeometry_instances.add(this);
        this.ready = new EventEmitter();
        _NgtGeometry_geometryArgs.set(this, []);
        _NgtGeometry_geometry.set(this, void 0);
    }
    set geometryArgs(v) {
        __classPrivateFieldSet(this, _NgtGeometry_geometryArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_init).call(this);
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!this.geometry) {
                __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_init).call(this);
            }
        });
    }
    get geometry() {
        return __classPrivateFieldGet(this, _NgtGeometry_geometry, "f");
    }
    ngOnDestroy() {
        this.ngZone.runOutsideAngular(() => {
            if (this.geometry) {
                this.geometry.dispose();
            }
        });
    }
}
_NgtGeometry_geometryArgs = new WeakMap(), _NgtGeometry_geometry = new WeakMap(), _NgtGeometry_instances = new WeakSet(), _NgtGeometry_init = function _NgtGeometry_init() {
    // geometry has changed. reconstruct
    if (this.geometry) {
        // cleanup
        if (this.parentObject) {
            const object3d = this.parentObject;
            if (object3d.geometry) {
                object3d.geometry.dispose();
            }
        }
        // reconstruct
        __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_construct).call(this);
        __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_assign).call(this);
    }
    else {
        __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_construct).call(this);
        __classPrivateFieldGet(this, _NgtGeometry_instances, "m", _NgtGeometry_assign).call(this);
    }
}, _NgtGeometry_assign = function _NgtGeometry_assign() {
    requestAnimationFrame(() => {
        const parentObject = this.parentObject();
        if (parentObject) {
            parentObject.geometry = this.geometry;
        }
    });
}, _NgtGeometry_construct = function _NgtGeometry_construct() {
    __classPrivateFieldSet(this, _NgtGeometry_geometry, new this.geometryType(...__classPrivateFieldGet(this, _NgtGeometry_geometryArgs, "f")), "f");
    this.ready.emit(this.geometry);
};
NgtGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGeometry, deps: [{ token: i0.NgZone }, { token: NGT_OBJECT_3D }], target: i0.ɵɵFactoryTarget.Directive });
NgtGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtGeometry, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGeometry, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_3D]
                }] }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
