// @angular-three/core/esm2020/lib/three/helper.mjs
var _NgtHelper_helper, _NgtHelper_helperArgs;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, NgZone, Output, } from '@angular/core';
import { NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, } from '../controllers/object-3d.controller';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/object-3d.controller";
export class NgtHelper {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtHelper_helper.set(this, void 0);
        _NgtHelper_helperArgs.set(this, []);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtHelper_helper, new this.helperType(...__classPrivateFieldGet(this, _NgtHelper_helperArgs, "f")), "f");
            return __classPrivateFieldGet(this, _NgtHelper_helper, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.helper);
        };
    }
    set helperArgs(v) {
        __classPrivateFieldSet(this, _NgtHelper_helperArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
    ngOnChanges() {
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!__classPrivateFieldGet(this, _NgtHelper_helper, "f")) {
                this.objectController.init();
            }
        });
    }
    get helper() {
        return __classPrivateFieldGet(this, _NgtHelper_helper, "f");
    }
}
_NgtHelper_helper = new WeakMap(), _NgtHelper_helperArgs = new WeakMap();
NgtHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHelper, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHelper, outputs: { ready: "ready" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHelper, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
