// @angular-three/core/esm2020/lib/three/light.mjs
var _NgtLight_lightArgs, _NgtLight_light;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, Input, NgZone, Output, } from '@angular/core';
import { NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, } from '../controllers/object-3d.controller';
import { applyProps } from '../utils/apply-props';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/object-3d.controller";
export class NgtLight {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtLight_lightArgs.set(this, []);
        _NgtLight_light.set(this, void 0);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtLight_light, new this.lightType(...__classPrivateFieldGet(this, _NgtLight_lightArgs, "f")), "f");
            if (this.intensity) {
                applyProps(this.light, { intensity: this.intensity });
            }
            if (this.shadow) {
                applyProps(this.light, this.shadow);
            }
            return __classPrivateFieldGet(this, _NgtLight_light, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.light);
        };
    }
    set lightArgs(v) {
        __classPrivateFieldSet(this, _NgtLight_lightArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!__classPrivateFieldGet(this, _NgtLight_light, "f")) {
                this.objectController.init();
            }
        });
    }
    get light() {
        return __classPrivateFieldGet(this, _NgtLight_light, "f");
    }
}
_NgtLight_lightArgs = new WeakMap(), _NgtLight_light = new WeakMap();
NgtLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLight, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLight, inputs: { intensity: "intensity", shadow: "shadow" }, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLight, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }], intensity: [{
                type: Input
            }], shadow: [{
                type: Input
            }] } });
