// @angular-three/core/esm2020/lib/three/line.mjs
import { Directive, EventEmitter, Inject, Optional, Output, } from '@angular/core';
import * as THREE from 'three';
import { NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, NGT_OBJECT_TYPE, NgtMaterialGeometryController, } from '../controllers/material-geometry.controller';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/material-geometry.controller";
export class NgtCommonLine {
    constructor(materialGeometryController) {
        this.materialGeometryController = materialGeometryController;
        this.ready = new EventEmitter();
        if (materialGeometryController) {
            materialGeometryController.readyFn = () => {
                this.ready.emit(this.line);
            };
        }
    }
    get line() {
        return this.materialGeometryController?.objectController.object3d;
    }
}
NgtCommonLine.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonLine, deps: [{ token: NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
NgtCommonLine.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCommonLine, outputs: { ready: "ready" }, providers: [{ provide: NGT_OBJECT_TYPE, useValue: THREE.Line }], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonLine, decorators: [{
            type: Directive,
            args: [{
                    providers: [{ provide: NGT_OBJECT_TYPE, useValue: THREE.Line }],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtMaterialGeometryController, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER]
                }] }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
