// @angular-three/core/esm2020/lib/three/material.mjs
var _NgtMaterial_instances, _NgtMaterial_parameters, _NgtMaterial_material, _NgtMaterial_convertColorToLinear;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Inject, Input, NgZone, Optional, Output, } from '@angular/core';
import { requestAnimationFrame } from '@rx-angular/cdk/zone-less';
import { NgtContentMaterialController } from '../controllers/content-material.controller';
import { NGT_OBJECT_3D } from '../di/object3d';
import { NgtStore } from '../stores/store';
import { makeColor } from '../utils/make';
import * as i0 from "@angular/core";
import * as i1 from "../stores/store";
import * as i2 from "../controllers/content-material.controller";
export class NgtMaterial {
    constructor(ngZone, store, parentObject, contentMaterialController) {
        this.ngZone = ngZone;
        this.store = store;
        this.parentObject = parentObject;
        this.contentMaterialController = contentMaterialController;
        _NgtMaterial_instances.add(this);
        this.ready = new EventEmitter();
        _NgtMaterial_parameters.set(this, void 0);
        _NgtMaterial_material.set(this, void 0);
    }
    set parameters(v) {
        requestAnimationFrame(() => {
            __classPrivateFieldSet(this, _NgtMaterial_parameters, v, "f");
            if (v && this.material) {
                __classPrivateFieldGet(this, _NgtMaterial_instances, "m", _NgtMaterial_convertColorToLinear).call(this, v);
                this.material.setValues(v);
                this.material.needsUpdate = true;
            }
        });
    }
    get parameters() {
        return __classPrivateFieldGet(this, _NgtMaterial_parameters, "f");
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (this.parameters) {
                __classPrivateFieldGet(this, _NgtMaterial_instances, "m", _NgtMaterial_convertColorToLinear).call(this, this.parameters);
            }
            __classPrivateFieldSet(this, _NgtMaterial_material, new this.materialType(this.parameters), "f");
            requestAnimationFrame(() => {
                const parentObject = this.parentObject();
                if (parentObject) {
                    if (this.contentMaterialController &&
                        this.contentMaterialController.isMaterialArray) {
                        if (!Array.isArray(parentObject.material)) {
                            parentObject.material = [];
                        }
                        parentObject.material.push(this.material);
                    }
                    else {
                        parentObject.material = this.material;
                    }
                }
                this.ready.emit(this.material);
            });
        });
    }
    get material() {
        return __classPrivateFieldGet(this, _NgtMaterial_material, "f");
    }
    ngOnDestroy() {
        this.ngZone.runOutsideAngular(() => {
            if (this.material) {
                this.material.dispose();
            }
        });
    }
}
_NgtMaterial_parameters = new WeakMap(), _NgtMaterial_material = new WeakMap(), _NgtMaterial_instances = new WeakSet(), _NgtMaterial_convertColorToLinear = function _NgtMaterial_convertColorToLinear(parameters) {
    if ('color' in parameters) {
        const colorParams = parameters['color'];
        parameters['color'] = makeColor(colorParams);
        if (!this.store.get('linear')) {
            parameters['color'].convertSRGBToLinear();
        }
    }
};
NgtMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMaterial, deps: [{ token: i0.NgZone }, { token: i1.NgtStore }, { token: NGT_OBJECT_3D }, { token: i2.NgtContentMaterialController, optional: true }], target: i0.ɵɵFactoryTarget.Directive });
NgtMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMaterial, inputs: { parameters: "parameters" }, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMaterial, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }, { type: i1.NgtStore }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_3D]
                }] }, { type: i2.NgtContentMaterialController, decorators: [{
                    type: Optional
                }] }]; }, propDecorators: { ready: [{
                type: Output
            }], parameters: [{
                type: Input
            }] } });
