// @angular-three/core/esm2020/lib/three/mesh.mjs
import { Directive, EventEmitter, Inject, NgZone, Optional, Output, } from '@angular/core';
import { NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, NgtMaterialGeometryController, } from '../controllers/material-geometry.controller';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/material-geometry.controller";
export class NgtCommonMesh {
    constructor(materialGeometryController, ngZone) {
        this.materialGeometryController = materialGeometryController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        if (materialGeometryController) {
            materialGeometryController.readyFn = () => {
                this.ready.emit(this.mesh);
            };
        }
    }
    get mesh() {
        return this.materialGeometryController.objectController.object3d;
    }
}
NgtCommonMesh.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonMesh, deps: [{ token: NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, optional: true }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtCommonMesh.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCommonMesh, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCommonMesh, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i1.NgtMaterialGeometryController, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
