// @angular-three/core/esm2020/lib/three/object-helper.mjs
var _NgtObjectHelper_object3d, _NgtObjectHelper_objectHelper;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, Inject, NgZone } from '@angular/core';
import { requestAnimationFrame } from '@rx-angular/cdk/zone-less';
import { NGT_OBJECT_3D } from '../di/object3d';
import { NgtAnimationFrameStore } from '../stores/animation-frame.store';
import { EnhancedRxState } from '../stores/enhanced-rx-state';
import { NgtStore } from '../stores/store';
import * as i0 from "@angular/core";
import * as i1 from "../stores/store";
import * as i2 from "../stores/animation-frame.store";
export class NgtObjectHelper extends EnhancedRxState {
    constructor(object3dFn, store, animationFrameStore, ngZone) {
        super();
        this.object3dFn = object3dFn;
        this.store = store;
        this.animationFrameStore = animationFrameStore;
        this.ngZone = ngZone;
        _NgtObjectHelper_object3d.set(this, void 0);
        _NgtObjectHelper_objectHelper.set(this, void 0);
        this.set({ args: [] });
    }
    set objectHelperArgs(v) {
        this.set({ args: Array.isArray(v) ? v : [v] });
    }
    ngOnInit() {
        requestAnimationFrame(() => {
            __classPrivateFieldSet(this, _NgtObjectHelper_object3d, this.object3dFn(), "f");
            if (!__classPrivateFieldGet(this, _NgtObjectHelper_object3d, "f")) {
                console.info('Parent is not an object3d');
                return;
            }
            this.holdEffect(this.select('args'), (args) => {
                __classPrivateFieldSet(this, _NgtObjectHelper_objectHelper, new this.objectHelperType(__classPrivateFieldGet(this, _NgtObjectHelper_object3d, "f"), ...args), "f");
                this.store.get('scene').add(this.objectHelper);
                const animationUuid = this.animationFrameStore.register({
                    callback: () => {
                        if (this.objectHelper) {
                            this.objectHelper.update();
                        }
                    },
                });
                return () => {
                    if (this.objectHelper) {
                        this.store.get('scene').remove(this.objectHelper);
                        this.animationFrameStore.actions.unsubscriberUuid(animationUuid);
                    }
                };
            });
        });
    }
    get objectHelper() {
        return __classPrivateFieldGet(this, _NgtObjectHelper_objectHelper, "f");
    }
}
_NgtObjectHelper_object3d = new WeakMap(), _NgtObjectHelper_objectHelper = new WeakMap();
NgtObjectHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtObjectHelper, deps: [{ token: NGT_OBJECT_3D }, { token: i1.NgtStore }, { token: i2.NgtAnimationFrameStore }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtObjectHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtObjectHelper, usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtObjectHelper, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: undefined, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_3D]
                }] }, { type: i1.NgtStore }, { type: i2.NgtAnimationFrameStore }, { type: i0.NgZone }]; } });
