// @angular-three/core/esm2020/lib/three/texture.mjs
var _NgtTexture_textureArgs, _NgtTexture_texture;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, NgZone, Output, } from '@angular/core';
import * as i0 from "@angular/core";
export class NgtTexture {
    constructor(ngZone) {
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtTexture_textureArgs.set(this, []);
        _NgtTexture_texture.set(this, void 0);
    }
    set textureArgs(v) {
        __classPrivateFieldSet(this, _NgtTexture_textureArgs, Array.isArray(v) ? v : [v], "f");
        this.ngZone.runOutsideAngular(() => {
            __classPrivateFieldSet(this, _NgtTexture_texture, new this.textureType(...__classPrivateFieldGet(this, _NgtTexture_textureArgs, "f")), "f");
        });
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (!this.texture) {
                __classPrivateFieldSet(this, _NgtTexture_texture, new this.textureType(...__classPrivateFieldGet(this, _NgtTexture_textureArgs, "f")), "f");
                this.ready.emit(this.texture);
            }
        });
    }
    get texture() {
        return __classPrivateFieldGet(this, _NgtTexture_texture, "f");
    }
    ngOnDestroy() {
        this.ngZone.runOutsideAngular(() => {
            if (this.texture) {
                this.texture.dispose();
            }
        });
    }
}
_NgtTexture_textureArgs = new WeakMap(), _NgtTexture_texture = new WeakMap();
NgtTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTexture, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtTexture, outputs: { ready: "ready" }, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTexture, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
