// @angular-three/core/esm2020/geometries/lib/cylinder-geometry/cylinder-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCylinderGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.CylinderGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtCylinderGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCylinderGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCylinderGeometry, selector: "ngt-cylinder-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtCylinderGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtCylinderGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cylinder-geometry',
                    exportAs: 'ngtCylinderGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtCylinderGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCylinderGeometryModule {
}
NgtCylinderGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCylinderGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometryModule, declarations: [NgtCylinderGeometry], exports: [NgtCylinderGeometry] });
NgtCylinderGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCylinderGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCylinderGeometry],
                    exports: [NgtCylinderGeometry],
                }]
        }] });
