// @angular-three/core/esm2020/geometries/lib/octahedron-geometry/octahedron-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtOctahedronGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.OctahedronGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtOctahedronGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtOctahedronGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtOctahedronGeometry, selector: "ngt-octahedron-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtOctahedronGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtOctahedronGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-octahedron-geometry',
                    exportAs: 'ngtOctahedronGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtOctahedronGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtOctahedronGeometryModule {
}
NgtOctahedronGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtOctahedronGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometryModule, declarations: [NgtOctahedronGeometry], exports: [NgtOctahedronGeometry] });
NgtOctahedronGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOctahedronGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtOctahedronGeometry],
                    exports: [NgtOctahedronGeometry],
                }]
        }] });
