// @angular-three/core/esm2020/lod/lib/lod.directive.mjs
var _NgtLod_lod;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { NGT_OBJECT_CONTROLLER_PROVIDER, NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, NgtObject3dControllerModule, } from '@angular-three/core';
import { Directive, EventEmitter, Inject, NgModule, NgZone, Output, } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtLod {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtLod_lod.set(this, void 0);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtLod_lod, new THREE.LOD(), "f");
            return __classPrivateFieldGet(this, _NgtLod_lod, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.lod);
        };
    }
    get lod() {
        return __classPrivateFieldGet(this, _NgtLod_lod, "f");
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
}
_NgtLod_lod = new WeakMap();
NgtLod.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLod, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtLod.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLod, selector: "ngt-lod", outputs: { ready: "ready" }, providers: [NGT_OBJECT_CONTROLLER_PROVIDER], exportAs: ["ngtLod"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLod, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-lod',
                    exportAs: 'ngtLod',
                    providers: [NGT_OBJECT_CONTROLLER_PROVIDER],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
export class NgtLodModule {
}
NgtLodModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLodModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLodModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLodModule, declarations: [NgtLod], exports: [NgtLod, NgtObject3dControllerModule] });
NgtLodModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLodModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLodModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLod],
                    exports: [NgtLod, NgtObject3dControllerModule],
                }]
        }] });
