// @angular-three/core/esm2020/geometries/lib/box-geometry/box-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtBoxGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.BoxGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtBoxGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBoxGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBoxGeometry, selector: "ngt-box-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtBoxGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtBoxGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-box-geometry',
                    exportAs: 'ngtBoxGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtBoxGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtBoxGeometryModule {
}
NgtBoxGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBoxGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometryModule, declarations: [NgtBoxGeometry], exports: [NgtBoxGeometry] });
NgtBoxGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBoxGeometry],
                    exports: [NgtBoxGeometry],
                }]
        }] });
