// @angular-three/core/esm2020/helpers/lib/box-helper/box-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtBoxHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.BoxHelper;
    }
    set ngtBoxHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtBoxHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBoxHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBoxHelper, selector: "[ngtBoxHelper]", inputs: { ngtBoxHelper: "ngtBoxHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtBoxHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtBoxHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtBoxHelper]',
                    exportAs: 'ngtBoxHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtBoxHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtBoxHelper: [{
                type: Input
            }] } });
export class NgtBoxHelperModule {
}
NgtBoxHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBoxHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, declarations: [NgtBoxHelper], exports: [NgtBoxHelper] });
NgtBoxHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBoxHelper],
                    exports: [NgtBoxHelper],
                }]
        }] });
