// @angular-three/core/esm2020/materials/lib/mesh-physical-material/mesh-physical-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshPhysicalMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshPhysicalMaterial;
    }
}
NgtMeshPhysicalMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshPhysicalMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshPhysicalMaterial, selector: "ngt-mesh-physical-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshPhysicalMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshPhysicalMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-physical-material',
                    exportAs: 'ngtMeshPhysicalMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshPhysicalMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshPhysicalMaterialModule {
}
NgtMeshPhysicalMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshPhysicalMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, declarations: [NgtMeshPhysicalMaterial], exports: [NgtMeshPhysicalMaterial] });
NgtMeshPhysicalMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshPhysicalMaterial],
                    exports: [NgtMeshPhysicalMaterial],
                }]
        }] });
