// @angular-three/core/esm2020/lines/lib/line/line.directive.mjs
// GENERATED
import { NgtCommonLine, NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_OBJECT_TYPE, NgtMaterialGeometryControllerModule, } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLine extends NgtCommonLine {
}
NgtLine.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLine, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLine.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLine, selector: "ngt-line", providers: [
        {
            provide: NgtCommonLine,
            useExisting: NgtLine,
        },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        {
            provide: NGT_OBJECT_TYPE,
            useValue: THREE.Line,
        },
    ], exportAs: ["ngtLine"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLine, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line',
                    exportAs: 'ngtLine',
                    providers: [
                        {
                            provide: NgtCommonLine,
                            useExisting: NgtLine,
                        },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        {
                            provide: NGT_OBJECT_TYPE,
                            useValue: THREE.Line,
                        },
                    ],
                }]
        }] });
export class NgtLineModule {
}
NgtLineModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, declarations: [NgtLine], exports: [NgtLine, NgtMaterialGeometryControllerModule] });
NgtLineModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLine],
                    exports: [NgtLine, NgtMaterialGeometryControllerModule],
                }]
        }] });
