// @angular-three/core/esm2020/lights/lib/directional-light/directional-light.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDirectionalLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.DirectionalLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtDirectionalLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDirectionalLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDirectionalLight, selector: "ngt-directional-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtDirectionalLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtDirectionalLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-directional-light',
                    exportAs: 'ngtDirectionalLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtDirectionalLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDirectionalLightModule {
}
NgtDirectionalLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDirectionalLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, declarations: [NgtDirectionalLight], exports: [NgtDirectionalLight, NgtObject3dControllerModule] });
NgtDirectionalLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDirectionalLight],
                    exports: [NgtDirectionalLight, NgtObject3dControllerModule],
                }]
        }] });
