// @angular-three/core/esm2020/curves/lib/quadratic-bezier-curve/quadratic-bezier-curve.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtQuadraticBezierCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.QuadraticBezierCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtQuadraticBezierCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtQuadraticBezierCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtQuadraticBezierCurve, selector: "ngt-quadratic-bezier-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtQuadraticBezierCurve,
        },
    ], exportAs: ["ngtQuadraticBezierCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-quadratic-bezier-curve',
                    exportAs: 'ngtQuadraticBezierCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtQuadraticBezierCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtQuadraticBezierCurveModule {
}
NgtQuadraticBezierCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtQuadraticBezierCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, declarations: [NgtQuadraticBezierCurve], exports: [NgtQuadraticBezierCurve] });
NgtQuadraticBezierCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtQuadraticBezierCurve],
                    exports: [NgtQuadraticBezierCurve],
                }]
        }] });
