// @angular-three/core/esm2020/materials/lib/points-material/points-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPointsMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.PointsMaterial;
    }
}
NgtPointsMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPointsMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPointsMaterial, selector: "ngt-points-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtPointsMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtPointsMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-points-material',
                    exportAs: 'ngtPointsMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtPointsMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtPointsMaterialModule {
}
NgtPointsMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointsMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, declarations: [NgtPointsMaterial], exports: [NgtPointsMaterial] });
NgtPointsMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPointsMaterial],
                    exports: [NgtPointsMaterial],
                }]
        }] });
