// @angular-three/soba/esm2020/abstractions/positional-audio/positional-audio.component.mjs
import { EnhancedRxState, NGT_AUDIO_CONTROLLER_PROVIDER, NGT_AUDIO_WATCHED_CONTROLLER, NgtAudioController, NgtAudioControllerModule, NgtLoaderService, NgtSobaExtender, NgtStore, } from '@angular-three/core';
import { NgtPositionalAudioModule } from '@angular-three/core/audios';
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Inject, Injectable, Input, NgModule, } from '@angular/core';
import { selectSlice } from '@rx-angular/state';
import { combineLatest, distinctUntilChanged, map, startWith, switchMap, } from 'rxjs';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
import * as i2 from "@angular/common";
import * as i3 from "@angular-three/core/audios";
export class NgtSobaPositionalAudioStore extends EnhancedRxState {
    constructor(loaderService, store) {
        super();
        this.loaderService = loaderService;
        this.store = store;
        this.buffer$ = this.select('url').pipe(switchMap((url) => this.loaderService.use(THREE.AudioLoader, url)));
        this.changes$ = combineLatest([
            this.buffer$,
            this.select(selectSlice(['distance', 'loop'])),
        ]).pipe(map(([buffer, { loop, distance }]) => ({ buffer, loop, distance })));
        this.set({ listener: new THREE.AudioListener(), distance: 1, loop: true });
        this.hold(this.changes$, ({ buffer, loop, distance }) => {
            const { autoplay, audio } = this.get();
            if (audio) {
                audio.setBuffer(buffer);
                audio.setRefDistance(distance);
                audio.setLoop(loop);
                if (autoplay && !audio.isPlaying)
                    audio.play();
            }
        });
        this.holdEffect(this.select('autoplay'), () => {
            const { audio, listener } = this.get();
            const camera = this.store.get('camera');
            if (camera) {
                camera.add(listener);
            }
            return () => {
                if (camera) {
                    camera.remove(listener);
                }
                if (audio) {
                    if (audio.isPlaying)
                        audio.stop();
                    if (audio.source &&
                        audio.source._connected) {
                        audio.disconnect();
                    }
                }
            };
        });
    }
}
NgtSobaPositionalAudioStore.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioStore, deps: [{ token: i1.NgtLoaderService }, { token: i1.NgtStore }], target: i0.ɵɵFactoryTarget.Injectable });
NgtSobaPositionalAudioStore.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioStore });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioStore, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.NgtLoaderService }, { type: i1.NgtStore }]; } });
export class NgtSobaPositionalAudio extends NgtSobaExtender {
    constructor(audioController, sobaPositionalAudioStore) {
        super();
        this.audioController = audioController;
        this.sobaPositionalAudioStore = sobaPositionalAudioStore;
        this.listener$ = this.sobaPositionalAudioStore.select('listener');
        sobaPositionalAudioStore.connect('autoplay', audioController.change$.pipe(map(() => audioController.autoplay || false), startWith(audioController.autoplay || false), distinctUntilChanged()));
    }
    set url(url) {
        this.sobaPositionalAudioStore.set({ url });
    }
    set distance(distance) {
        this.sobaPositionalAudioStore.set({ distance });
    }
    set loop(loop) {
        this.sobaPositionalAudioStore.set({ loop });
    }
    onPositionalAudioReady(audio) {
        this.object = audio;
        this.sobaPositionalAudioStore.set({ audio });
    }
}
NgtSobaPositionalAudio.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudio, deps: [{ token: NGT_AUDIO_WATCHED_CONTROLLER }, { token: NgtSobaPositionalAudioStore }], target: i0.ɵɵFactoryTarget.Component });
NgtSobaPositionalAudio.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaPositionalAudio, selector: "ngt-soba-positional-audio", inputs: { url: "url", distance: "distance", loop: "loop" }, providers: [
        NGT_AUDIO_CONTROLLER_PROVIDER,
        NgtSobaPositionalAudioStore,
        { provide: NgtSobaExtender, useExisting: NgtSobaPositionalAudio },
    ], usesInheritance: true, ngImport: i0, template: `
    <ngt-positional-audio
      *ngIf="listener$ | async as listener"
      (ready)="onPositionalAudioReady($event)"
      [listener]="listener"
      [audioController]="audioController"
    ></ngt-positional-audio>
  `, isInline: true, directives: [{ type: i2.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i3.NgtPositionalAudio, selector: "ngt-positional-audio", exportAs: ["ngtPositionalAudio"] }, { type: i1.NgtAudioController, selector: "    ngt-audio,    ngt-positional-audio,    ngt-soba-positional-audio,  ", inputs: ["autoplay", "buffer", "detune", "loop", "loopStart", "loopEnd", "offset", "duration", "playbackRate", "isPlaying", "hasPlaybackControl", "sourceType", "source", "filters", "audioController"], exportAs: ["ngtAudioController"] }, { type: i1.NgtObject3dController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera  ", exportAs: ["ngtObject3dController"] }, { type: i1.NgtObject3dInputsController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera,    ngt-soba-plane,    ngt-soba-box,    ngt-soba-cylinder,    ngt-soba-cone,    ngt-soba-circle,    ngt-soba-sphere,    ngt-soba-tube,    ngt-soba-torus,    ngt-soba-tetrahedron,    ngt-soba-ring,    ngt-soba-polyhedron,    ngt-soba-octahedron,    ngt-soba-dodecahedron,    ngt-soba-icosahedron,    ngt-soba-extrude,    ngt-soba-lathe,    ngt-soba-torus-knot,    ngt-soba-billboard,    ngt-soba-detailed,    ngt-soba-line,    ngt-soba-quadratic-bezier-line,    ngt-soba-cubic-bezier-line,    ngt-soba-orthographic-camera,    ngt-soba-gizmo-helper,    ngt-soba-gizmo-viewport,    ngt-soba-gizmo-axis-head,    ngt-soba-text  ", inputs: ["name", "position", "rotation", "quaternion", "scale", "color", "userData", "castShadow", "receiveShadow", "visible", "matrixAutoUpdate", "dispose", "raycast", "appendMode", "appendTo", "object3dInputsController"], outputs: ["click", "contextmenu", "dblclick", "pointerup", "pointerdown", "pointerover", "pointerout", "pointerenter", "pointerleave", "pointermove", "pointermissed", "pointercancel", "wheel"], exportAs: ["ngtObject3dInputsController"] }], pipes: { "async": i2.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudio, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngt-soba-positional-audio',
                    template: `
    <ngt-positional-audio
      *ngIf="listener$ | async as listener"
      (ready)="onPositionalAudioReady($event)"
      [listener]="listener"
      [audioController]="audioController"
    ></ngt-positional-audio>
  `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        NGT_AUDIO_CONTROLLER_PROVIDER,
                        NgtSobaPositionalAudioStore,
                        { provide: NgtSobaExtender, useExisting: NgtSobaPositionalAudio },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtAudioController, decorators: [{
                    type: Inject,
                    args: [NGT_AUDIO_WATCHED_CONTROLLER]
                }] }, { type: NgtSobaPositionalAudioStore }]; }, propDecorators: { url: [{
                type: Input
            }], distance: [{
                type: Input
            }], loop: [{
                type: Input
            }] } });
export class NgtSobaPositionalAudioModule {
}
NgtSobaPositionalAudioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaPositionalAudioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioModule, declarations: [NgtSobaPositionalAudio], imports: [NgtPositionalAudioModule, CommonModule], exports: [NgtSobaPositionalAudio, NgtAudioControllerModule] });
NgtSobaPositionalAudioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioModule, imports: [[NgtPositionalAudioModule, CommonModule], NgtAudioControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPositionalAudioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaPositionalAudio],
                    exports: [NgtSobaPositionalAudio, NgtAudioControllerModule],
                    imports: [NgtPositionalAudioModule, CommonModule],
                }]
        }] });
