// @angular-three/core/esm2020/textures/lib/data-texture2-darray/data-texture2-darray.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDataTexture2DArray extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.DataTexture2DArray;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtDataTexture2DArray.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArray, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDataTexture2DArray.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDataTexture2DArray, selector: "ngt-data-texture2-darray", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtDataTexture2DArray,
        },
    ], exportAs: ["ngtDataTexture2DArray"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArray, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-data-texture2-darray',
                    exportAs: 'ngtDataTexture2DArray',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtDataTexture2DArray,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDataTexture2DArrayModule {
}
NgtDataTexture2DArrayModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArrayModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDataTexture2DArrayModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArrayModule, declarations: [NgtDataTexture2DArray], exports: [NgtDataTexture2DArray] });
NgtDataTexture2DArrayModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArrayModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture2DArrayModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDataTexture2DArray],
                    exports: [NgtDataTexture2DArray],
                }]
        }] });
