// @angular-three/core/esm2020/textures/lib/compressed-texture/compressed-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCompressedTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.CompressedTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtCompressedTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCompressedTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCompressedTexture, selector: "ngt-compressed-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtCompressedTexture,
        },
    ], exportAs: ["ngtCompressedTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-compressed-texture',
                    exportAs: 'ngtCompressedTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtCompressedTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCompressedTextureModule {
}
NgtCompressedTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCompressedTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTextureModule, declarations: [NgtCompressedTexture], exports: [NgtCompressedTexture] });
NgtCompressedTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCompressedTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCompressedTexture],
                    exports: [NgtCompressedTexture],
                }]
        }] });
