// @angular-three/core/esm2020/attributes/lib/interleaved-buffer-attribute/interleaved-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInterleavedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.InterleavedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInterleavedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInterleavedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInterleavedBufferAttribute, selector: "ngt-interleaved-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInterleavedBufferAttribute,
        },
    ], exportAs: ["ngtInterleavedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-interleaved-buffer-attribute',
                    exportAs: 'ngtInterleavedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInterleavedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInterleavedBufferAttributeModule {
}
NgtInterleavedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInterleavedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, declarations: [NgtInterleavedBufferAttribute], exports: [NgtInterleavedBufferAttribute] });
NgtInterleavedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInterleavedBufferAttribute],
                    exports: [NgtInterleavedBufferAttribute],
                }]
        }] });
