// @angular-three/soba/esm2020/controls/lib/first-person-controls/first-person-controls.directive.mjs
import { EnhancedRxState, NgtAnimationFrameStore, NgtStore, } from '@angular-three/core';
import { Directive, NgModule, NgZone, Output } from '@angular/core';
import { selectSlice } from '@rx-angular/state';
import { FirstPersonControls } from 'three-stdlib';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtSobaFirstPersonControls extends EnhancedRxState {
    constructor(store, animationFrameStore, ngZone) {
        super();
        this.store = store;
        this.animationFrameStore = animationFrameStore;
        this.ngZone = ngZone;
        this.ready = this.select('controls');
    }
    ngOnInit() {
        this.hold(this.store.select(selectSlice(['camera', 'renderer'])), ({ renderer, camera }) => {
            this.ngZone.runOutsideAngular(() => {
                if (camera && renderer) {
                    this.set({
                        controls: new FirstPersonControls(camera, renderer.domElement),
                    });
                }
            });
        });
        this.holdEffect(this.select('controls'), (controls) => {
            const animationUuid = this.animationFrameStore.register({
                callback: ({ delta }) => {
                    controls.update(delta);
                },
            });
            return () => {
                this.animationFrameStore.actions.unsubscriberUuid(animationUuid);
            };
        });
    }
    get controls() {
        return this.get('controls');
    }
}
NgtSobaFirstPersonControls.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControls, deps: [{ token: i1.NgtStore }, { token: i1.NgtAnimationFrameStore }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtSobaFirstPersonControls.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaFirstPersonControls, selector: "ngt-soba-first-person-controls", outputs: { ready: "ready" }, exportAs: ["ngtSobaFirstPersonControls"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControls, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-soba-first-person-controls',
                    exportAs: 'ngtSobaFirstPersonControls',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtStore }, { type: i1.NgtAnimationFrameStore }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
export class NgtSobaFirstPersonControlsModule {
}
NgtSobaFirstPersonControlsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControlsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaFirstPersonControlsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControlsModule, declarations: [NgtSobaFirstPersonControls], exports: [NgtSobaFirstPersonControls] });
NgtSobaFirstPersonControlsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControlsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFirstPersonControlsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaFirstPersonControls],
                    exports: [NgtSobaFirstPersonControls],
                }]
        }] });
