// @angular-three/core/esm2020/audios/lib/positional-audio/positional-audio.directive.mjs
// GENERATED
import { NGT_AUDIO_CONTROLLER_PROVIDER, NGT_OBJECT_CONTROLLER_PROVIDER, NgtAudioControllerModule, NgtCommonAudio, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPositionalAudio extends NgtCommonAudio {
    constructor() {
        super(...arguments);
        this.audioType = THREE.PositionalAudio;
    }
}
NgtPositionalAudio.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudio, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPositionalAudio.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPositionalAudio, selector: "ngt-positional-audio", providers: [
        {
            provide: NgtCommonAudio,
            useExisting: NgtPositionalAudio,
        },
        NGT_AUDIO_CONTROLLER_PROVIDER,
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPositionalAudio"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudio, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-positional-audio',
                    exportAs: 'ngtPositionalAudio',
                    providers: [
                        {
                            provide: NgtCommonAudio,
                            useExisting: NgtPositionalAudio,
                        },
                        NGT_AUDIO_CONTROLLER_PROVIDER,
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }] });
export class NgtPositionalAudioModule {
}
NgtPositionalAudioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPositionalAudioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudioModule, declarations: [NgtPositionalAudio], exports: [NgtPositionalAudio, NgtAudioControllerModule,
        NgtObject3dControllerModule] });
NgtPositionalAudioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudioModule, imports: [NgtAudioControllerModule,
        NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPositionalAudioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPositionalAudio],
                    exports: [
                        NgtPositionalAudio,
                        NgtAudioControllerModule,
                        NgtObject3dControllerModule,
                    ],
                }]
        }] });
