// @angular-three/core/esm2020/attributes/lib/uint16-buffer-attribute/uint16-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtUint16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint16BufferAttribute, selector: "ngt-uint16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint16BufferAttribute,
        },
    ], exportAs: ["ngtUint16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint16-buffer-attribute',
                    exportAs: 'ngtUint16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtUint16BufferAttributeModule {
}
NgtUint16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, declarations: [NgtUint16BufferAttribute], exports: [NgtUint16BufferAttribute] });
NgtUint16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint16BufferAttribute],
                    exports: [NgtUint16BufferAttribute],
                }]
        }] });
