// @angular-three/core/esm2020/curves/lib/line-curve/line-curve.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLineCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.LineCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtLineCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineCurve, selector: "ngt-line-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtLineCurve,
        },
    ], exportAs: ["ngtLineCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-curve',
                    exportAs: 'ngtLineCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtLineCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtLineCurveModule {
}
NgtLineCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, declarations: [NgtLineCurve], exports: [NgtLineCurve] });
NgtLineCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineCurve],
                    exports: [NgtLineCurve],
                }]
        }] });
