// @angular-three/core/fesm2020/angular-three-core-points.mjs
import * as i1 from '@angular-three/core';
import { NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_OBJECT_TYPE, NgtMaterialGeometryControllerModule } from '@angular-three/core';
import * as i0 from '@angular/core';
import { EventEmitter, Directive, Inject, Output, NgModule } from '@angular/core';
import * as THREE from 'three';

class NgtPoints {
    constructor(materialGeometryController) {
        this.materialGeometryController = materialGeometryController;
        this.ready = new EventEmitter();
        materialGeometryController.readyFn = () => {
            this.ready.emit(this.points);
        };
    }
    get points() {
        return this.materialGeometryController.objectController
            .object3d;
    }
}
NgtPoints.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPoints, deps: [{ token: NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER }], target: i0.ɵɵFactoryTarget.Directive });
NgtPoints.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPoints, selector: "ngt-points", outputs: { ready: "ready" }, providers: [
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        { provide: NGT_OBJECT_TYPE, useValue: THREE.Points },
    ], exportAs: ["ngtPoints"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPoints, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-points',
                    exportAs: 'ngtPoints',
                    providers: [
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        { provide: NGT_OBJECT_TYPE, useValue: THREE.Points },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtMaterialGeometryController, decorators: [{
                    type: Inject,
                    args: [NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER]
                }] }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
class NgtPointsModule {
}
NgtPointsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsModule, declarations: [NgtPoints], exports: [NgtPoints, NgtMaterialGeometryControllerModule] });
NgtPointsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPoints],
                    exports: [NgtPoints, NgtMaterialGeometryControllerModule],
                }]
        }] });

/**
 * Generated bundle index. Do not edit.
 */

export { NgtPoints, NgtPointsModule };
