// @angular-three/core/fesm2020/angular-three-core-lines.mjs
import { NgtCommonLine, NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_OBJECT_TYPE, NgtMaterialGeometryControllerModule } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtLine extends NgtCommonLine {
}
NgtLine.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLine, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLine.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLine, selector: "ngt-line", providers: [
        {
            provide: NgtCommonLine,
            useExisting: NgtLine,
        },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        {
            provide: NGT_OBJECT_TYPE,
            useValue: THREE.Line,
        },
    ], exportAs: ["ngtLine"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLine, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line',
                    exportAs: 'ngtLine',
                    providers: [
                        {
                            provide: NgtCommonLine,
                            useExisting: NgtLine,
                        },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        {
                            provide: NGT_OBJECT_TYPE,
                            useValue: THREE.Line,
                        },
                    ],
                }]
        }] });
class NgtLineModule {
}
NgtLineModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, declarations: [NgtLine], exports: [NgtLine, NgtMaterialGeometryControllerModule] });
NgtLineModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLine],
                    exports: [NgtLine, NgtMaterialGeometryControllerModule],
                }]
        }] });

// GENERATED
class NgtLineLoop extends NgtCommonLine {
}
NgtLineLoop.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoop, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineLoop.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineLoop, selector: "ngt-line-loop", providers: [
        {
            provide: NgtCommonLine,
            useExisting: NgtLineLoop,
        },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        {
            provide: NGT_OBJECT_TYPE,
            useValue: THREE.LineLoop,
        },
    ], exportAs: ["ngtLineLoop"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoop, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-loop',
                    exportAs: 'ngtLineLoop',
                    providers: [
                        {
                            provide: NgtCommonLine,
                            useExisting: NgtLineLoop,
                        },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        {
                            provide: NGT_OBJECT_TYPE,
                            useValue: THREE.LineLoop,
                        },
                    ],
                }]
        }] });
class NgtLineLoopModule {
}
NgtLineLoopModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoopModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineLoopModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoopModule, declarations: [NgtLineLoop], exports: [NgtLineLoop, NgtMaterialGeometryControllerModule] });
NgtLineLoopModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoopModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineLoopModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineLoop],
                    exports: [NgtLineLoop, NgtMaterialGeometryControllerModule],
                }]
        }] });

// GENERATED
class NgtLineSegments extends NgtCommonLine {
}
NgtLineSegments.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegments, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineSegments.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineSegments, selector: "ngt-line-segments", providers: [
        {
            provide: NgtCommonLine,
            useExisting: NgtLineSegments,
        },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        {
            provide: NGT_OBJECT_TYPE,
            useValue: THREE.LineSegments,
        },
    ], exportAs: ["ngtLineSegments"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegments, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-segments',
                    exportAs: 'ngtLineSegments',
                    providers: [
                        {
                            provide: NgtCommonLine,
                            useExisting: NgtLineSegments,
                        },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        {
                            provide: NGT_OBJECT_TYPE,
                            useValue: THREE.LineSegments,
                        },
                    ],
                }]
        }] });
class NgtLineSegmentsModule {
}
NgtLineSegmentsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineSegmentsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, declarations: [NgtLineSegments], exports: [NgtLineSegments, NgtMaterialGeometryControllerModule] });
NgtLineSegmentsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineSegments],
                    exports: [NgtLineSegments, NgtMaterialGeometryControllerModule],
                }]
        }] });

// GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtLine, NgtLineLoop, NgtLineLoopModule, NgtLineModule, NgtLineSegments, NgtLineSegmentsModule };
