// @angular-three/core/fesm2020/angular-three-core-helpers.mjs
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, NgtObjectHelper, NGT_OBJECT_3D_PROVIDER } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtArrowHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.ArrowHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtArrowHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtArrowHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtArrowHelper, selector: "ngt-arrow-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtArrowHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtArrowHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-arrow-helper',
                    exportAs: 'ngtArrowHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtArrowHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtArrowHelperModule {
}
NgtArrowHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtArrowHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelperModule, declarations: [NgtArrowHelper], exports: [NgtArrowHelper, NgtObject3dControllerModule] });
NgtArrowHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrowHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtArrowHelper],
                    exports: [NgtArrowHelper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtAxesHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.AxesHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtAxesHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAxesHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAxesHelper, selector: "ngt-axes-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtAxesHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAxesHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-axes-helper',
                    exportAs: 'ngtAxesHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtAxesHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtAxesHelperModule {
}
NgtAxesHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAxesHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, declarations: [NgtAxesHelper], exports: [NgtAxesHelper, NgtObject3dControllerModule] });
NgtAxesHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAxesHelper],
                    exports: [NgtAxesHelper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtBox3Helper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.Box3Helper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtBox3Helper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3Helper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBox3Helper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBox3Helper, selector: "ngt-box3-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtBox3Helper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtBox3Helper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3Helper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-box3-helper',
                    exportAs: 'ngtBox3Helper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtBox3Helper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtBox3HelperModule {
}
NgtBox3HelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBox3HelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, declarations: [NgtBox3Helper], exports: [NgtBox3Helper, NgtObject3dControllerModule] });
NgtBox3HelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBox3Helper],
                    exports: [NgtBox3Helper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtGridHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.GridHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtGridHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtGridHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtGridHelper, selector: "ngt-grid-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtGridHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtGridHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-grid-helper',
                    exportAs: 'ngtGridHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtGridHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtGridHelperModule {
}
NgtGridHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtGridHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, declarations: [NgtGridHelper], exports: [NgtGridHelper, NgtObject3dControllerModule] });
NgtGridHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtGridHelper],
                    exports: [NgtGridHelper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtPlaneHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.PlaneHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtPlaneHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPlaneHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPlaneHelper, selector: "ngt-plane-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtPlaneHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPlaneHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-plane-helper',
                    exportAs: 'ngtPlaneHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtPlaneHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtPlaneHelperModule {
}
NgtPlaneHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPlaneHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, declarations: [NgtPlaneHelper], exports: [NgtPlaneHelper, NgtObject3dControllerModule] });
NgtPlaneHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPlaneHelper],
                    exports: [NgtPlaneHelper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtPolarGridHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.PolarGridHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtPolarGridHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPolarGridHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPolarGridHelper, selector: "ngt-polar-grid-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtPolarGridHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPolarGridHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-polar-grid-helper',
                    exportAs: 'ngtPolarGridHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtPolarGridHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtPolarGridHelperModule {
}
NgtPolarGridHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPolarGridHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, declarations: [NgtPolarGridHelper], exports: [NgtPolarGridHelper, NgtObject3dControllerModule] });
NgtPolarGridHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPolarGridHelper],
                    exports: [NgtPolarGridHelper, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtBoxHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.BoxHelper;
    }
    set ngtBoxHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtBoxHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBoxHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBoxHelper, selector: "[ngtBoxHelper]", inputs: { ngtBoxHelper: "ngtBoxHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtBoxHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtBoxHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtBoxHelper]',
                    exportAs: 'ngtBoxHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtBoxHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtBoxHelper: [{
                type: Input
            }] } });
class NgtBoxHelperModule {
}
NgtBoxHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBoxHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, declarations: [NgtBoxHelper], exports: [NgtBoxHelper] });
NgtBoxHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBoxHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBoxHelper],
                    exports: [NgtBoxHelper],
                }]
        }] });

// GENERATED
class NgtCameraHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.CameraHelper;
    }
    set ngtCameraHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtCameraHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCameraHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCameraHelper, selector: "[ngtCameraHelper]", inputs: { ngtCameraHelper: "ngtCameraHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtCameraHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtCameraHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtCameraHelper]',
                    exportAs: 'ngtCameraHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtCameraHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtCameraHelper: [{
                type: Input
            }] } });
class NgtCameraHelperModule {
}
NgtCameraHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCameraHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, declarations: [NgtCameraHelper], exports: [NgtCameraHelper] });
NgtCameraHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCameraHelper],
                    exports: [NgtCameraHelper],
                }]
        }] });

// GENERATED
class NgtDirectionalLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.DirectionalLightHelper;
    }
    set ngtDirectionalLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtDirectionalLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDirectionalLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDirectionalLightHelper, selector: "[ngtDirectionalLightHelper]", inputs: { ngtDirectionalLightHelper: "ngtDirectionalLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtDirectionalLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtDirectionalLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtDirectionalLightHelper]',
                    exportAs: 'ngtDirectionalLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtDirectionalLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtDirectionalLightHelper: [{
                type: Input
            }] } });
class NgtDirectionalLightHelperModule {
}
NgtDirectionalLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDirectionalLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, declarations: [NgtDirectionalLightHelper], exports: [NgtDirectionalLightHelper] });
NgtDirectionalLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDirectionalLightHelper],
                    exports: [NgtDirectionalLightHelper],
                }]
        }] });

// GENERATED
class NgtHemisphereLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.HemisphereLightHelper;
    }
    set ngtHemisphereLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtHemisphereLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLightHelper, selector: "[ngtHemisphereLightHelper]", inputs: { ngtHemisphereLightHelper: "ngtHemisphereLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtHemisphereLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtHemisphereLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtHemisphereLightHelper]',
                    exportAs: 'ngtHemisphereLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtHemisphereLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtHemisphereLightHelper: [{
                type: Input
            }] } });
class NgtHemisphereLightHelperModule {
}
NgtHemisphereLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, declarations: [NgtHemisphereLightHelper], exports: [NgtHemisphereLightHelper] });
NgtHemisphereLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLightHelper],
                    exports: [NgtHemisphereLightHelper],
                }]
        }] });

// GENERATED
class NgtPointLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.PointLightHelper;
    }
    set ngtPointLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtPointLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPointLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPointLightHelper, selector: "[ngtPointLightHelper]", inputs: { ngtPointLightHelper: "ngtPointLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtPointLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtPointLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtPointLightHelper]',
                    exportAs: 'ngtPointLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtPointLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtPointLightHelper: [{
                type: Input
            }] } });
class NgtPointLightHelperModule {
}
NgtPointLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, declarations: [NgtPointLightHelper], exports: [NgtPointLightHelper] });
NgtPointLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPointLightHelper],
                    exports: [NgtPointLightHelper],
                }]
        }] });

// GENERATED
class NgtSkeletonHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.SkeletonHelper;
    }
    set ngtSkeletonHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtSkeletonHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSkeletonHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSkeletonHelper, selector: "[ngtSkeletonHelper]", inputs: { ngtSkeletonHelper: "ngtSkeletonHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtSkeletonHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSkeletonHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtSkeletonHelper]',
                    exportAs: 'ngtSkeletonHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtSkeletonHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtSkeletonHelper: [{
                type: Input
            }] } });
class NgtSkeletonHelperModule {
}
NgtSkeletonHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSkeletonHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, declarations: [NgtSkeletonHelper], exports: [NgtSkeletonHelper] });
NgtSkeletonHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSkeletonHelper],
                    exports: [NgtSkeletonHelper],
                }]
        }] });

// GENERATED
class NgtSpotLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.SpotLightHelper;
    }
    set ngtSpotLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtSpotLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpotLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpotLightHelper, selector: "[ngtSpotLightHelper]", inputs: { ngtSpotLightHelper: "ngtSpotLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtSpotLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSpotLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtSpotLightHelper]',
                    exportAs: 'ngtSpotLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtSpotLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtSpotLightHelper: [{
                type: Input
            }] } });
class NgtSpotLightHelperModule {
}
NgtSpotLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpotLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, declarations: [NgtSpotLightHelper], exports: [NgtSpotLightHelper] });
NgtSpotLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpotLightHelper],
                    exports: [NgtSpotLightHelper],
                }]
        }] });

// GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtArrowHelper, NgtArrowHelperModule, NgtAxesHelper, NgtAxesHelperModule, NgtBox3Helper, NgtBox3HelperModule, NgtBoxHelper, NgtBoxHelperModule, NgtCameraHelper, NgtCameraHelperModule, NgtDirectionalLightHelper, NgtDirectionalLightHelperModule, NgtGridHelper, NgtGridHelperModule, NgtHemisphereLightHelper, NgtHemisphereLightHelperModule, NgtPlaneHelper, NgtPlaneHelperModule, NgtPointLightHelper, NgtPointLightHelperModule, NgtPolarGridHelper, NgtPolarGridHelperModule, NgtSkeletonHelper, NgtSkeletonHelperModule, NgtSpotLightHelper, NgtSpotLightHelperModule };
