// @angular-three/core/fesm2020/angular-three-core-attributes.mjs
import { NgtAttribute } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBufferAttribute, selector: "ngt-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtBufferAttribute,
        },
    ], exportAs: ["ngtBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-buffer-attribute',
                    exportAs: 'ngtBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtBufferAttributeModule {
}
NgtBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, declarations: [NgtBufferAttribute], exports: [NgtBufferAttribute] });
NgtBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBufferAttribute],
                    exports: [NgtBufferAttribute],
                }]
        }] });

// GENERATED
class NgtInstancedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.InstancedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInstancedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInstancedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInstancedBufferAttribute, selector: "ngt-instanced-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInstancedBufferAttribute,
        },
    ], exportAs: ["ngtInstancedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-instanced-buffer-attribute',
                    exportAs: 'ngtInstancedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInstancedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtInstancedBufferAttributeModule {
}
NgtInstancedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInstancedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, declarations: [NgtInstancedBufferAttribute], exports: [NgtInstancedBufferAttribute] });
NgtInstancedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInstancedBufferAttribute],
                    exports: [NgtInstancedBufferAttribute],
                }]
        }] });

// GENERATED
class NgtInterleavedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.InterleavedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInterleavedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInterleavedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInterleavedBufferAttribute, selector: "ngt-interleaved-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInterleavedBufferAttribute,
        },
    ], exportAs: ["ngtInterleavedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-interleaved-buffer-attribute',
                    exportAs: 'ngtInterleavedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInterleavedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtInterleavedBufferAttributeModule {
}
NgtInterleavedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInterleavedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, declarations: [NgtInterleavedBufferAttribute], exports: [NgtInterleavedBufferAttribute] });
NgtInterleavedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInterleavedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInterleavedBufferAttribute],
                    exports: [NgtInterleavedBufferAttribute],
                }]
        }] });

// GENERATED
class NgtFloat16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat16BufferAttribute, selector: "ngt-float16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat16BufferAttribute,
        },
    ], exportAs: ["ngtFloat16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float16-buffer-attribute',
                    exportAs: 'ngtFloat16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtFloat16BufferAttributeModule {
}
NgtFloat16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, declarations: [NgtFloat16BufferAttribute], exports: [NgtFloat16BufferAttribute] });
NgtFloat16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat16BufferAttribute],
                    exports: [NgtFloat16BufferAttribute],
                }]
        }] });

// GENERATED
class NgtFloat32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat32BufferAttribute, selector: "ngt-float32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat32BufferAttribute,
        },
    ], exportAs: ["ngtFloat32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float32-buffer-attribute',
                    exportAs: 'ngtFloat32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtFloat32BufferAttributeModule {
}
NgtFloat32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, declarations: [NgtFloat32BufferAttribute], exports: [NgtFloat32BufferAttribute] });
NgtFloat32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat32BufferAttribute],
                    exports: [NgtFloat32BufferAttribute],
                }]
        }] });

// GENERATED
class NgtFloat64BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float64BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat64BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat64BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat64BufferAttribute, selector: "ngt-float64-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat64BufferAttribute,
        },
    ], exportAs: ["ngtFloat64BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float64-buffer-attribute',
                    exportAs: 'ngtFloat64BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat64BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtFloat64BufferAttributeModule {
}
NgtFloat64BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat64BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, declarations: [NgtFloat64BufferAttribute], exports: [NgtFloat64BufferAttribute] });
NgtFloat64BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat64BufferAttribute],
                    exports: [NgtFloat64BufferAttribute],
                }]
        }] });

// GENERATED
class NgtInt8BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int8BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt8BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt8BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt8BufferAttribute, selector: "ngt-int8-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt8BufferAttribute,
        },
    ], exportAs: ["ngtInt8BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int8-buffer-attribute',
                    exportAs: 'ngtInt8BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt8BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtInt8BufferAttributeModule {
}
NgtInt8BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt8BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, declarations: [NgtInt8BufferAttribute], exports: [NgtInt8BufferAttribute] });
NgtInt8BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt8BufferAttribute],
                    exports: [NgtInt8BufferAttribute],
                }]
        }] });

// GENERATED
class NgtInt16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt16BufferAttribute, selector: "ngt-int16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt16BufferAttribute,
        },
    ], exportAs: ["ngtInt16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int16-buffer-attribute',
                    exportAs: 'ngtInt16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtInt16BufferAttributeModule {
}
NgtInt16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, declarations: [NgtInt16BufferAttribute], exports: [NgtInt16BufferAttribute] });
NgtInt16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt16BufferAttribute],
                    exports: [NgtInt16BufferAttribute],
                }]
        }] });

// GENERATED
class NgtInt32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt32BufferAttribute, selector: "ngt-int32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt32BufferAttribute,
        },
    ], exportAs: ["ngtInt32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int32-buffer-attribute',
                    exportAs: 'ngtInt32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtInt32BufferAttributeModule {
}
NgtInt32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, declarations: [NgtInt32BufferAttribute], exports: [NgtInt32BufferAttribute] });
NgtInt32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt32BufferAttribute],
                    exports: [NgtInt32BufferAttribute],
                }]
        }] });

// GENERATED
class NgtUint8BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint8BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint8BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint8BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint8BufferAttribute, selector: "ngt-uint8-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint8BufferAttribute,
        },
    ], exportAs: ["ngtUint8BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint8-buffer-attribute',
                    exportAs: 'ngtUint8BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint8BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtUint8BufferAttributeModule {
}
NgtUint8BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint8BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, declarations: [NgtUint8BufferAttribute], exports: [NgtUint8BufferAttribute] });
NgtUint8BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint8BufferAttribute],
                    exports: [NgtUint8BufferAttribute],
                }]
        }] });

// GENERATED
class NgtUint16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint16BufferAttribute, selector: "ngt-uint16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint16BufferAttribute,
        },
    ], exportAs: ["ngtUint16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint16-buffer-attribute',
                    exportAs: 'ngtUint16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtUint16BufferAttributeModule {
}
NgtUint16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, declarations: [NgtUint16BufferAttribute], exports: [NgtUint16BufferAttribute] });
NgtUint16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint16BufferAttribute],
                    exports: [NgtUint16BufferAttribute],
                }]
        }] });

// GENERATED
class NgtUint32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint32BufferAttribute, selector: "ngt-uint32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint32BufferAttribute,
        },
    ], exportAs: ["ngtUint32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint32-buffer-attribute',
                    exportAs: 'ngtUint32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtUint32BufferAttributeModule {
}
NgtUint32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, declarations: [NgtUint32BufferAttribute], exports: [NgtUint32BufferAttribute] });
NgtUint32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint32BufferAttribute],
                    exports: [NgtUint32BufferAttribute],
                }]
        }] });

// GENERATED
class NgtUint8ClampedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint8ClampedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint8ClampedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint8ClampedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint8ClampedBufferAttribute, selector: "ngt-uint8-clamped-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint8ClampedBufferAttribute,
        },
    ], exportAs: ["ngtUint8ClampedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint8-clamped-buffer-attribute',
                    exportAs: 'ngtUint8ClampedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint8ClampedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtUint8ClampedBufferAttributeModule {
}
NgtUint8ClampedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint8ClampedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, declarations: [NgtUint8ClampedBufferAttribute], exports: [NgtUint8ClampedBufferAttribute] });
NgtUint8ClampedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint8ClampedBufferAttribute],
                    exports: [NgtUint8ClampedBufferAttribute],
                }]
        }] });

//GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtBufferAttribute, NgtBufferAttributeModule, NgtFloat16BufferAttribute, NgtFloat16BufferAttributeModule, NgtFloat32BufferAttribute, NgtFloat32BufferAttributeModule, NgtFloat64BufferAttribute, NgtFloat64BufferAttributeModule, NgtInstancedBufferAttribute, NgtInstancedBufferAttributeModule, NgtInt16BufferAttribute, NgtInt16BufferAttributeModule, NgtInt32BufferAttribute, NgtInt32BufferAttributeModule, NgtInt8BufferAttribute, NgtInt8BufferAttributeModule, NgtInterleavedBufferAttribute, NgtInterleavedBufferAttributeModule, NgtUint16BufferAttribute, NgtUint16BufferAttributeModule, NgtUint32BufferAttribute, NgtUint32BufferAttributeModule, NgtUint8BufferAttribute, NgtUint8BufferAttributeModule, NgtUint8ClampedBufferAttribute, NgtUint8ClampedBufferAttributeModule };
