// @angular-three/core/fesm2020/angular-three-core-materials.mjs
import { NgtMaterial, NGT_OBJECT_3D_PROVIDER } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtShadowMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.ShadowMaterial;
    }
}
NgtShadowMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtShadowMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtShadowMaterial, selector: "ngt-shadow-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtShadowMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtShadowMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-shadow-material',
                    exportAs: 'ngtShadowMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtShadowMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtShadowMaterialModule {
}
NgtShadowMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtShadowMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, declarations: [NgtShadowMaterial], exports: [NgtShadowMaterial] });
NgtShadowMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtShadowMaterial],
                    exports: [NgtShadowMaterial],
                }]
        }] });

// GENERATED
class NgtSpriteMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.SpriteMaterial;
    }
}
NgtSpriteMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpriteMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpriteMaterial, selector: "ngt-sprite-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtSpriteMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSpriteMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-sprite-material',
                    exportAs: 'ngtSpriteMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtSpriteMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtSpriteMaterialModule {
}
NgtSpriteMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpriteMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, declarations: [NgtSpriteMaterial], exports: [NgtSpriteMaterial] });
NgtSpriteMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpriteMaterial],
                    exports: [NgtSpriteMaterial],
                }]
        }] });

// GENERATED
class NgtRawShaderMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.RawShaderMaterial;
    }
}
NgtRawShaderMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtRawShaderMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtRawShaderMaterial, selector: "ngt-raw-shader-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtRawShaderMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtRawShaderMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-raw-shader-material',
                    exportAs: 'ngtRawShaderMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtRawShaderMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtRawShaderMaterialModule {
}
NgtRawShaderMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtRawShaderMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, declarations: [NgtRawShaderMaterial], exports: [NgtRawShaderMaterial] });
NgtRawShaderMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtRawShaderMaterial],
                    exports: [NgtRawShaderMaterial],
                }]
        }] });

// GENERATED
class NgtShaderMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.ShaderMaterial;
    }
}
NgtShaderMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtShaderMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtShaderMaterial, selector: "ngt-shader-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtShaderMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtShaderMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-shader-material',
                    exportAs: 'ngtShaderMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtShaderMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtShaderMaterialModule {
}
NgtShaderMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtShaderMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, declarations: [NgtShaderMaterial], exports: [NgtShaderMaterial] });
NgtShaderMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtShaderMaterial],
                    exports: [NgtShaderMaterial],
                }]
        }] });

// GENERATED
class NgtPointsMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.PointsMaterial;
    }
}
NgtPointsMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPointsMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPointsMaterial, selector: "ngt-points-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtPointsMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtPointsMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-points-material',
                    exportAs: 'ngtPointsMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtPointsMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtPointsMaterialModule {
}
NgtPointsMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointsMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, declarations: [NgtPointsMaterial], exports: [NgtPointsMaterial] });
NgtPointsMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointsMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPointsMaterial],
                    exports: [NgtPointsMaterial],
                }]
        }] });

// GENERATED
class NgtMeshPhysicalMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshPhysicalMaterial;
    }
}
NgtMeshPhysicalMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshPhysicalMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshPhysicalMaterial, selector: "ngt-mesh-physical-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshPhysicalMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshPhysicalMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-physical-material',
                    exportAs: 'ngtMeshPhysicalMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshPhysicalMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshPhysicalMaterialModule {
}
NgtMeshPhysicalMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshPhysicalMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, declarations: [NgtMeshPhysicalMaterial], exports: [NgtMeshPhysicalMaterial] });
NgtMeshPhysicalMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhysicalMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshPhysicalMaterial],
                    exports: [NgtMeshPhysicalMaterial],
                }]
        }] });

// GENERATED
class NgtMeshStandardMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshStandardMaterial;
    }
}
NgtMeshStandardMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshStandardMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshStandardMaterial, selector: "ngt-mesh-standard-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshStandardMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshStandardMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-standard-material',
                    exportAs: 'ngtMeshStandardMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshStandardMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshStandardMaterialModule {
}
NgtMeshStandardMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshStandardMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, declarations: [NgtMeshStandardMaterial], exports: [NgtMeshStandardMaterial] });
NgtMeshStandardMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshStandardMaterial],
                    exports: [NgtMeshStandardMaterial],
                }]
        }] });

// GENERATED
class NgtMeshPhongMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshPhongMaterial;
    }
}
NgtMeshPhongMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshPhongMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshPhongMaterial, selector: "ngt-mesh-phong-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshPhongMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshPhongMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-phong-material',
                    exportAs: 'ngtMeshPhongMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshPhongMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshPhongMaterialModule {
}
NgtMeshPhongMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshPhongMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, declarations: [NgtMeshPhongMaterial], exports: [NgtMeshPhongMaterial] });
NgtMeshPhongMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshPhongMaterial],
                    exports: [NgtMeshPhongMaterial],
                }]
        }] });

// GENERATED
class NgtMeshToonMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshToonMaterial;
    }
}
NgtMeshToonMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshToonMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshToonMaterial, selector: "ngt-mesh-toon-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshToonMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshToonMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-toon-material',
                    exportAs: 'ngtMeshToonMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshToonMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshToonMaterialModule {
}
NgtMeshToonMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshToonMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, declarations: [NgtMeshToonMaterial], exports: [NgtMeshToonMaterial] });
NgtMeshToonMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshToonMaterial],
                    exports: [NgtMeshToonMaterial],
                }]
        }] });

// GENERATED
class NgtMeshNormalMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshNormalMaterial;
    }
}
NgtMeshNormalMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshNormalMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshNormalMaterial, selector: "ngt-mesh-normal-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshNormalMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshNormalMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-normal-material',
                    exportAs: 'ngtMeshNormalMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshNormalMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshNormalMaterialModule {
}
NgtMeshNormalMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshNormalMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, declarations: [NgtMeshNormalMaterial], exports: [NgtMeshNormalMaterial] });
NgtMeshNormalMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshNormalMaterial],
                    exports: [NgtMeshNormalMaterial],
                }]
        }] });

// GENERATED
class NgtMeshLambertMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshLambertMaterial;
    }
}
NgtMeshLambertMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshLambertMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshLambertMaterial, selector: "ngt-mesh-lambert-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshLambertMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshLambertMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-lambert-material',
                    exportAs: 'ngtMeshLambertMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshLambertMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshLambertMaterialModule {
}
NgtMeshLambertMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshLambertMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, declarations: [NgtMeshLambertMaterial], exports: [NgtMeshLambertMaterial] });
NgtMeshLambertMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshLambertMaterial],
                    exports: [NgtMeshLambertMaterial],
                }]
        }] });

// GENERATED
class NgtMeshDepthMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshDepthMaterial;
    }
}
NgtMeshDepthMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshDepthMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshDepthMaterial, selector: "ngt-mesh-depth-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshDepthMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshDepthMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-depth-material',
                    exportAs: 'ngtMeshDepthMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshDepthMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshDepthMaterialModule {
}
NgtMeshDepthMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshDepthMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, declarations: [NgtMeshDepthMaterial], exports: [NgtMeshDepthMaterial] });
NgtMeshDepthMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshDepthMaterial],
                    exports: [NgtMeshDepthMaterial],
                }]
        }] });

// GENERATED
class NgtMeshDistanceMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshDistanceMaterial;
    }
}
NgtMeshDistanceMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshDistanceMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshDistanceMaterial, selector: "ngt-mesh-distance-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshDistanceMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshDistanceMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-distance-material',
                    exportAs: 'ngtMeshDistanceMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshDistanceMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshDistanceMaterialModule {
}
NgtMeshDistanceMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshDistanceMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, declarations: [NgtMeshDistanceMaterial], exports: [NgtMeshDistanceMaterial] });
NgtMeshDistanceMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshDistanceMaterial],
                    exports: [NgtMeshDistanceMaterial],
                }]
        }] });

// GENERATED
class NgtMeshBasicMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshBasicMaterial;
    }
}
NgtMeshBasicMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshBasicMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshBasicMaterial, selector: "ngt-mesh-basic-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshBasicMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshBasicMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-basic-material',
                    exportAs: 'ngtMeshBasicMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshBasicMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshBasicMaterialModule {
}
NgtMeshBasicMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshBasicMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, declarations: [NgtMeshBasicMaterial], exports: [NgtMeshBasicMaterial] });
NgtMeshBasicMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshBasicMaterial],
                    exports: [NgtMeshBasicMaterial],
                }]
        }] });

// GENERATED
class NgtMeshMatcapMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshMatcapMaterial;
    }
}
NgtMeshMatcapMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshMatcapMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshMatcapMaterial, selector: "ngt-mesh-matcap-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshMatcapMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshMatcapMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-matcap-material',
                    exportAs: 'ngtMeshMatcapMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshMatcapMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtMeshMatcapMaterialModule {
}
NgtMeshMatcapMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshMatcapMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, declarations: [NgtMeshMatcapMaterial], exports: [NgtMeshMatcapMaterial] });
NgtMeshMatcapMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshMatcapMaterial],
                    exports: [NgtMeshMatcapMaterial],
                }]
        }] });

// GENERATED
class NgtLineDashedMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.LineDashedMaterial;
    }
}
NgtLineDashedMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineDashedMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineDashedMaterial, selector: "ngt-line-dashed-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtLineDashedMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtLineDashedMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-dashed-material',
                    exportAs: 'ngtLineDashedMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtLineDashedMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtLineDashedMaterialModule {
}
NgtLineDashedMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineDashedMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, declarations: [NgtLineDashedMaterial], exports: [NgtLineDashedMaterial] });
NgtLineDashedMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineDashedMaterial],
                    exports: [NgtLineDashedMaterial],
                }]
        }] });

// GENERATED
class NgtLineBasicMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.LineBasicMaterial;
    }
}
NgtLineBasicMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineBasicMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineBasicMaterial, selector: "ngt-line-basic-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtLineBasicMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtLineBasicMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-basic-material',
                    exportAs: 'ngtLineBasicMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtLineBasicMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
class NgtLineBasicMaterialModule {
}
NgtLineBasicMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineBasicMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, declarations: [NgtLineBasicMaterial], exports: [NgtLineBasicMaterial] });
NgtLineBasicMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineBasicMaterial],
                    exports: [NgtLineBasicMaterial],
                }]
        }] });

// GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtLineBasicMaterial, NgtLineBasicMaterialModule, NgtLineDashedMaterial, NgtLineDashedMaterialModule, NgtMeshBasicMaterial, NgtMeshBasicMaterialModule, NgtMeshDepthMaterial, NgtMeshDepthMaterialModule, NgtMeshDistanceMaterial, NgtMeshDistanceMaterialModule, NgtMeshLambertMaterial, NgtMeshLambertMaterialModule, NgtMeshMatcapMaterial, NgtMeshMatcapMaterialModule, NgtMeshNormalMaterial, NgtMeshNormalMaterialModule, NgtMeshPhongMaterial, NgtMeshPhongMaterialModule, NgtMeshPhysicalMaterial, NgtMeshPhysicalMaterialModule, NgtMeshStandardMaterial, NgtMeshStandardMaterialModule, NgtMeshToonMaterial, NgtMeshToonMaterialModule, NgtPointsMaterial, NgtPointsMaterialModule, NgtRawShaderMaterial, NgtRawShaderMaterialModule, NgtShaderMaterial, NgtShaderMaterialModule, NgtShadowMaterial, NgtShadowMaterialModule, NgtSpriteMaterial, NgtSpriteMaterialModule };
