// @angular-three/core/fesm2020/angular-three-core-curves.mjs
import { NgtCurve } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtCatmullRomCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CatmullRomCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCatmullRomCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCatmullRomCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCatmullRomCurve3, selector: "ngt-catmull-rom-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCatmullRomCurve3,
        },
    ], exportAs: ["ngtCatmullRomCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-catmull-rom-curve3',
                    exportAs: 'ngtCatmullRomCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCatmullRomCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtCatmullRomCurve3Module {
}
NgtCatmullRomCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCatmullRomCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, declarations: [NgtCatmullRomCurve3], exports: [NgtCatmullRomCurve3] });
NgtCatmullRomCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCatmullRomCurve3],
                    exports: [NgtCatmullRomCurve3],
                }]
        }] });

// GENERATED
class NgtCubicBezierCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CubicBezierCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCubicBezierCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCubicBezierCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubicBezierCurve, selector: "ngt-cubic-bezier-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCubicBezierCurve,
        },
    ], exportAs: ["ngtCubicBezierCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cubic-bezier-curve',
                    exportAs: 'ngtCubicBezierCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCubicBezierCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtCubicBezierCurveModule {
}
NgtCubicBezierCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubicBezierCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, declarations: [NgtCubicBezierCurve], exports: [NgtCubicBezierCurve] });
NgtCubicBezierCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubicBezierCurve],
                    exports: [NgtCubicBezierCurve],
                }]
        }] });

// GENERATED
class NgtCubicBezierCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CubicBezierCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCubicBezierCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCubicBezierCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubicBezierCurve3, selector: "ngt-cubic-bezier-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCubicBezierCurve3,
        },
    ], exportAs: ["ngtCubicBezierCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cubic-bezier-curve3',
                    exportAs: 'ngtCubicBezierCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCubicBezierCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtCubicBezierCurve3Module {
}
NgtCubicBezierCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubicBezierCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, declarations: [NgtCubicBezierCurve3], exports: [NgtCubicBezierCurve3] });
NgtCubicBezierCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubicBezierCurve3],
                    exports: [NgtCubicBezierCurve3],
                }]
        }] });

// GENERATED
class NgtEllipseCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.EllipseCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtEllipseCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtEllipseCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtEllipseCurve, selector: "ngt-ellipse-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtEllipseCurve,
        },
    ], exportAs: ["ngtEllipseCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ellipse-curve',
                    exportAs: 'ngtEllipseCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtEllipseCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtEllipseCurveModule {
}
NgtEllipseCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtEllipseCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, declarations: [NgtEllipseCurve], exports: [NgtEllipseCurve] });
NgtEllipseCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtEllipseCurve],
                    exports: [NgtEllipseCurve],
                }]
        }] });

// GENERATED
class NgtLineCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.LineCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtLineCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineCurve, selector: "ngt-line-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtLineCurve,
        },
    ], exportAs: ["ngtLineCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-curve',
                    exportAs: 'ngtLineCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtLineCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtLineCurveModule {
}
NgtLineCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, declarations: [NgtLineCurve], exports: [NgtLineCurve] });
NgtLineCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineCurve],
                    exports: [NgtLineCurve],
                }]
        }] });

// GENERATED
class NgtLineCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.LineCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtLineCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineCurve3, selector: "ngt-line-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtLineCurve3,
        },
    ], exportAs: ["ngtLineCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-curve3',
                    exportAs: 'ngtLineCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtLineCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtLineCurve3Module {
}
NgtLineCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, declarations: [NgtLineCurve3], exports: [NgtLineCurve3] });
NgtLineCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineCurve3],
                    exports: [NgtLineCurve3],
                }]
        }] });

// GENERATED
class NgtQuadraticBezierCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.QuadraticBezierCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtQuadraticBezierCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtQuadraticBezierCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtQuadraticBezierCurve, selector: "ngt-quadratic-bezier-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtQuadraticBezierCurve,
        },
    ], exportAs: ["ngtQuadraticBezierCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-quadratic-bezier-curve',
                    exportAs: 'ngtQuadraticBezierCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtQuadraticBezierCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtQuadraticBezierCurveModule {
}
NgtQuadraticBezierCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtQuadraticBezierCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, declarations: [NgtQuadraticBezierCurve], exports: [NgtQuadraticBezierCurve] });
NgtQuadraticBezierCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtQuadraticBezierCurve],
                    exports: [NgtQuadraticBezierCurve],
                }]
        }] });

// GENERATED
class NgtQuadraticBezierCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.QuadraticBezierCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtQuadraticBezierCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtQuadraticBezierCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtQuadraticBezierCurve3, selector: "ngt-quadratic-bezier-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtQuadraticBezierCurve3,
        },
    ], exportAs: ["ngtQuadraticBezierCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-quadratic-bezier-curve3',
                    exportAs: 'ngtQuadraticBezierCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtQuadraticBezierCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtQuadraticBezierCurve3Module {
}
NgtQuadraticBezierCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtQuadraticBezierCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, declarations: [NgtQuadraticBezierCurve3], exports: [NgtQuadraticBezierCurve3] });
NgtQuadraticBezierCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtQuadraticBezierCurve3],
                    exports: [NgtQuadraticBezierCurve3],
                }]
        }] });

// GENERATED
class NgtSplineCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.SplineCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtSplineCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSplineCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSplineCurve, selector: "ngt-spline-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtSplineCurve,
        },
    ], exportAs: ["ngtSplineCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-spline-curve',
                    exportAs: 'ngtSplineCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtSplineCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtSplineCurveModule {
}
NgtSplineCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSplineCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, declarations: [NgtSplineCurve], exports: [NgtSplineCurve] });
NgtSplineCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSplineCurve],
                    exports: [NgtSplineCurve],
                }]
        }] });

// GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtCatmullRomCurve3, NgtCatmullRomCurve3Module, NgtCubicBezierCurve, NgtCubicBezierCurve3, NgtCubicBezierCurve3Module, NgtCubicBezierCurveModule, NgtEllipseCurve, NgtEllipseCurveModule, NgtLineCurve, NgtLineCurve3, NgtLineCurve3Module, NgtLineCurveModule, NgtQuadraticBezierCurve, NgtQuadraticBezierCurve3, NgtQuadraticBezierCurve3Module, NgtQuadraticBezierCurveModule, NgtSplineCurve, NgtSplineCurveModule };
