// @angular-three/soba/esm2020/abstractions/billboard/billboard.component.mjs
import { NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER, NGT_OBJECT_INPUTS_WATCHED_CONTROLLER, NgtObject3dInputsController, NgtObject3dInputsControllerModule, NgtSobaExtender, } from '@angular-three/core';
import { NgtGroupModule } from '@angular-three/core/group';
import { ChangeDetectionStrategy, Component, ContentChild, Inject, Input, NgModule, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core/group";
import * as i2 from "@angular-three/core";
export class NgtSobaBillboard extends NgtSobaExtender {
    constructor(objectInputsController) {
        super();
        this.objectInputsController = objectInputsController;
        this.follow = true;
        this.lockX = false;
        this.lockY = false;
        this.lockZ = false;
    }
    onGroupAnimate(group, $event) {
        if (!this.follow)
            return;
        this.animateReady.emit($event);
        // save previous rotation in case we're locking an axis
        const prevRotation = group.rotation.clone();
        // always face the camera
        group.quaternion.copy($event.camera.quaternion);
        // readjust any axis that is locked
        if (this.lockX)
            group.rotation.x = prevRotation.x;
        if (this.lockY)
            group.rotation.y = prevRotation.y;
        if (this.lockZ)
            group.rotation.z = prevRotation.z;
    }
    onGroupReady(group) {
        this.object = group;
        if (this.childInputsController) {
            this.childInputsController.appendTo = group;
        }
    }
}
NgtSobaBillboard.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboard, deps: [{ token: NGT_OBJECT_INPUTS_WATCHED_CONTROLLER }], target: i0.ɵɵFactoryTarget.Component });
NgtSobaBillboard.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaBillboard, selector: "ngt-soba-billboard", inputs: { follow: "follow", lockX: "lockX", lockY: "lockY", lockZ: "lockZ" }, providers: [
        NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER,
        { provide: NgtSobaExtender, useExisting: NgtSobaBillboard },
    ], queries: [{ propertyName: "childInputsController", first: true, predicate: NgtObject3dInputsController, descendants: true }], usesInheritance: true, ngImport: i0, template: `
    <ngt-group
      #ngtGroup="ngtGroup"
      (ready)="onGroupReady($event)"
      (animateReady)="onGroupAnimate(ngtGroup.group!, $event)"
      [object3dInputsController]="objectInputsController"
    >
      <ng-content></ng-content>
    </ngt-group>
  `, isInline: true, directives: [{ type: i1.NgtGroup, selector: "ngt-group", outputs: ["ready"], exportAs: ["ngtGroup"] }, { type: i2.NgtObject3dController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera  ", exportAs: ["ngtObject3dController"] }, { type: i2.NgtObject3dInputsController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera,    ngt-soba-plane,    ngt-soba-box,    ngt-soba-cylinder,    ngt-soba-cone,    ngt-soba-circle,    ngt-soba-sphere,    ngt-soba-tube,    ngt-soba-torus,    ngt-soba-tetrahedron,    ngt-soba-ring,    ngt-soba-polyhedron,    ngt-soba-octahedron,    ngt-soba-dodecahedron,    ngt-soba-icosahedron,    ngt-soba-extrude,    ngt-soba-lathe,    ngt-soba-torus-knot,    ngt-soba-billboard,    ngt-soba-detailed,    ngt-soba-line,    ngt-soba-quadratic-bezier-line,    ngt-soba-cubic-bezier-line,    ngt-soba-orthographic-camera,    ngt-soba-gizmo-helper,    ngt-soba-gizmo-viewport,    ngt-soba-gizmo-axis-head,    ngt-soba-text  ", inputs: ["name", "position", "rotation", "quaternion", "scale", "color", "userData", "castShadow", "receiveShadow", "visible", "matrixAutoUpdate", "dispose", "raycast", "appendMode", "appendTo", "object3dInputsController"], outputs: ["click", "contextmenu", "dblclick", "pointerup", "pointerdown", "pointerover", "pointerout", "pointerenter", "pointerleave", "pointermove", "pointermissed", "pointercancel", "wheel"], exportAs: ["ngtObject3dInputsController"] }, { type: i2.NgtAnimationSubscriberController, selector: "[animateReady]", inputs: ["priority"], outputs: ["animateReady"], exportAs: ["ngtAnimationSubscriberController"] }], changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboard, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngt-soba-billboard',
                    template: `
    <ngt-group
      #ngtGroup="ngtGroup"
      (ready)="onGroupReady($event)"
      (animateReady)="onGroupAnimate(ngtGroup.group!, $event)"
      [object3dInputsController]="objectInputsController"
    >
      <ng-content></ng-content>
    </ngt-group>
  `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER,
                        { provide: NgtSobaExtender, useExisting: NgtSobaBillboard },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i2.NgtObject3dInputsController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_INPUTS_WATCHED_CONTROLLER]
                }] }]; }, propDecorators: { follow: [{
                type: Input
            }], lockX: [{
                type: Input
            }], lockY: [{
                type: Input
            }], lockZ: [{
                type: Input
            }], childInputsController: [{
                type: ContentChild,
                args: [NgtObject3dInputsController]
            }] } });
export class NgtSobaBillboardModule {
}
NgtSobaBillboardModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboardModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaBillboardModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboardModule, declarations: [NgtSobaBillboard], imports: [NgtGroupModule], exports: [NgtSobaBillboard, NgtObject3dInputsControllerModule] });
NgtSobaBillboardModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboardModule, imports: [[NgtGroupModule], NgtObject3dInputsControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaBillboardModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaBillboard],
                    exports: [NgtSobaBillboard, NgtObject3dInputsControllerModule],
                    imports: [NgtGroupModule],
                }]
        }] });
