// @angular-three/core/esm2020/curves/lib/catmull-rom-curve3/catmull-rom-curve3.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCatmullRomCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CatmullRomCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCatmullRomCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCatmullRomCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCatmullRomCurve3, selector: "ngt-catmull-rom-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCatmullRomCurve3,
        },
    ], exportAs: ["ngtCatmullRomCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-catmull-rom-curve3',
                    exportAs: 'ngtCatmullRomCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCatmullRomCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCatmullRomCurve3Module {
}
NgtCatmullRomCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCatmullRomCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, declarations: [NgtCatmullRomCurve3], exports: [NgtCatmullRomCurve3] });
NgtCatmullRomCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCatmullRomCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCatmullRomCurve3],
                    exports: [NgtCatmullRomCurve3],
                }]
        }] });
