// @angular-three/core/esm2020/materials/lib/mesh-toon-material/mesh-toon-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshToonMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshToonMaterial;
    }
}
NgtMeshToonMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshToonMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshToonMaterial, selector: "ngt-mesh-toon-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshToonMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshToonMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-toon-material',
                    exportAs: 'ngtMeshToonMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshToonMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshToonMaterialModule {
}
NgtMeshToonMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshToonMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, declarations: [NgtMeshToonMaterial], exports: [NgtMeshToonMaterial] });
NgtMeshToonMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshToonMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshToonMaterial],
                    exports: [NgtMeshToonMaterial],
                }]
        }] });
