// @angular-three/core/esm2020/materials/lib/raw-shader-material/raw-shader-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtRawShaderMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.RawShaderMaterial;
    }
}
NgtRawShaderMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtRawShaderMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtRawShaderMaterial, selector: "ngt-raw-shader-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtRawShaderMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtRawShaderMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-raw-shader-material',
                    exportAs: 'ngtRawShaderMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtRawShaderMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtRawShaderMaterialModule {
}
NgtRawShaderMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtRawShaderMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, declarations: [NgtRawShaderMaterial], exports: [NgtRawShaderMaterial] });
NgtRawShaderMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRawShaderMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtRawShaderMaterial],
                    exports: [NgtRawShaderMaterial],
                }]
        }] });
