// @angular-three/core/esm2020/curves/lib/spline-curve/spline-curve.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSplineCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.SplineCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtSplineCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSplineCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSplineCurve, selector: "ngt-spline-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtSplineCurve,
        },
    ], exportAs: ["ngtSplineCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-spline-curve',
                    exportAs: 'ngtSplineCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtSplineCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtSplineCurveModule {
}
NgtSplineCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSplineCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, declarations: [NgtSplineCurve], exports: [NgtSplineCurve] });
NgtSplineCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSplineCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSplineCurve],
                    exports: [NgtSplineCurve],
                }]
        }] });
