// @angular-three/core/esm2020/attributes/lib/float16-buffer-attribute/float16-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtFloat16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat16BufferAttribute, selector: "ngt-float16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat16BufferAttribute,
        },
    ], exportAs: ["ngtFloat16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float16-buffer-attribute',
                    exportAs: 'ngtFloat16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtFloat16BufferAttributeModule {
}
NgtFloat16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, declarations: [NgtFloat16BufferAttribute], exports: [NgtFloat16BufferAttribute] });
NgtFloat16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat16BufferAttribute],
                    exports: [NgtFloat16BufferAttribute],
                }]
        }] });
