// @angular-three/core/esm2020/geometries/lib/torus-knot-geometry/torus-knot-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtTorusKnotGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.TorusKnotGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtTorusKnotGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtTorusKnotGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtTorusKnotGeometry, selector: "ngt-torus-knot-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtTorusKnotGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtTorusKnotGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-torus-knot-geometry',
                    exportAs: 'ngtTorusKnotGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtTorusKnotGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtTorusKnotGeometryModule {
}
NgtTorusKnotGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtTorusKnotGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometryModule, declarations: [NgtTorusKnotGeometry], exports: [NgtTorusKnotGeometry] });
NgtTorusKnotGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTorusKnotGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtTorusKnotGeometry],
                    exports: [NgtTorusKnotGeometry],
                }]
        }] });
