// @angular-three/core/esm2020/attributes/lib/buffer-attribute/buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBufferAttribute, selector: "ngt-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtBufferAttribute,
        },
    ], exportAs: ["ngtBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-buffer-attribute',
                    exportAs: 'ngtBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtBufferAttributeModule {
}
NgtBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, declarations: [NgtBufferAttribute], exports: [NgtBufferAttribute] });
NgtBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBufferAttribute],
                    exports: [NgtBufferAttribute],
                }]
        }] });
