// @angular-three/core/esm2020/lights/lib/rect-area-light/rect-area-light.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtRectAreaLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.RectAreaLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtRectAreaLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtRectAreaLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtRectAreaLight, selector: "ngt-rect-area-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtRectAreaLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtRectAreaLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-rect-area-light',
                    exportAs: 'ngtRectAreaLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtRectAreaLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtRectAreaLightModule {
}
NgtRectAreaLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtRectAreaLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, declarations: [NgtRectAreaLight], exports: [NgtRectAreaLight, NgtObject3dControllerModule] });
NgtRectAreaLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtRectAreaLight],
                    exports: [NgtRectAreaLight, NgtObject3dControllerModule],
                }]
        }] });
