// @angular-three/core/esm2020/materials/lib/shader-material/shader-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtShaderMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.ShaderMaterial;
    }
}
NgtShaderMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtShaderMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtShaderMaterial, selector: "ngt-shader-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtShaderMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtShaderMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-shader-material',
                    exportAs: 'ngtShaderMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtShaderMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtShaderMaterialModule {
}
NgtShaderMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtShaderMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, declarations: [NgtShaderMaterial], exports: [NgtShaderMaterial] });
NgtShaderMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShaderMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtShaderMaterial],
                    exports: [NgtShaderMaterial],
                }]
        }] });
