// @angular-three/core/esm2020/geometries/lib/extrude-geometry/extrude-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtExtrudeGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.ExtrudeGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtExtrudeGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtExtrudeGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtExtrudeGeometry, selector: "ngt-extrude-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtExtrudeGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtExtrudeGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-extrude-geometry',
                    exportAs: 'ngtExtrudeGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtExtrudeGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtExtrudeGeometryModule {
}
NgtExtrudeGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtExtrudeGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometryModule, declarations: [NgtExtrudeGeometry], exports: [NgtExtrudeGeometry] });
NgtExtrudeGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtExtrudeGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtExtrudeGeometry],
                    exports: [NgtExtrudeGeometry],
                }]
        }] });
