// @angular-three/soba/esm2020/abstractions/gizmo-helper/gizmo-helper.component.mjs
import { NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER, NgtCoreModule, NgtSobaExtender, } from '@angular-three/core';
import { NgtGroupModule } from '@angular-three/core/group';
import { NgtSobaOrthographicCameraModule } from '@angular-three/soba/cameras';
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { NgtSobaGizmoHelperStore } from './gizmo-helper.store';
import * as i0 from "@angular/core";
import * as i1 from "./gizmo-helper.store";
import * as i2 from "@angular-three/soba/cameras";
import * as i3 from "@angular/common";
import * as i4 from "@angular-three/core/group";
import * as i5 from "@angular-three/core";
export class NgtSobaGizmoHelper extends NgtSobaExtender {
    constructor(sobaGizmoHelperStore) {
        super();
        this.sobaGizmoHelperStore = sobaGizmoHelperStore;
        this.gizmoProps$ = this.sobaGizmoHelperStore.gizmoProps$;
        this.virtualScene$ = this.sobaGizmoHelperStore.select('virtualScene');
    }
    set alignment(alignment) {
        this.sobaGizmoHelperStore.set({ alignment });
    }
    set margin(margin) {
        this.sobaGizmoHelperStore.set({ margin });
    }
    set renderPriority(renderPriority) {
        this.sobaGizmoHelperStore.set({ renderPriority });
    }
    ngOnInit() {
        this.sobaGizmoHelperStore.actions.init();
    }
    onCameraReady(camera) {
        this.sobaGizmoHelperStore.set({ virtualCamera: camera });
    }
    onGizmoReady(gizmo) {
        this.sobaGizmoHelperStore.set({ gizmo });
        this.object = gizmo;
    }
}
NgtSobaGizmoHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelper, deps: [{ token: i1.NgtSobaGizmoHelperStore }], target: i0.ɵɵFactoryTarget.Component });
NgtSobaGizmoHelper.ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaGizmoHelper, selector: "ngt-soba-gizmo-helper", inputs: { alignment: "alignment", margin: "margin", renderPriority: "renderPriority" }, providers: [
        NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER,
        NgtSobaGizmoHelperStore,
        {
            provide: NgtSobaExtender,
            useExisting: NgtSobaGizmoHelper,
        },
    ], usesInheritance: true, ngImport: i0, template: `
    <ng-container *ngIf="virtualScene$ | async as virtualScene">
      <ngt-group
        *ngIf="gizmoProps$ | async as gizmoProps"
        (ready)="onGizmoReady($event)"
        [appendTo]="virtualScene"
        [position]="[gizmoProps.x, gizmoProps.y, 0]"
        [object3dInputsController]="gizmoProps.objectInputsController"
      ></ngt-group>
      <ngt-soba-orthographic-camera
        [appendTo]="virtualScene"
        [makeDefault]="false"
        [position]="[0, 0, 200]"
        (ready)="onCameraReady($event)"
      ></ngt-soba-orthographic-camera>
      <ng-content></ng-content>
    </ng-container>
  `, isInline: true, components: [{ type: i2.NgtSobaOrthographicCamera, selector: "ngt-soba-orthographic-camera", inputs: ["makeDefault", "manual", "near", "far"] }], directives: [{ type: i3.NgIf, selector: "[ngIf]", inputs: ["ngIf", "ngIfThen", "ngIfElse"] }, { type: i4.NgtGroup, selector: "ngt-group", outputs: ["ready"], exportAs: ["ngtGroup"] }, { type: i5.NgtObject3dController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera  ", exportAs: ["ngtObject3dController"] }, { type: i5.NgtObject3dInputsController, selector: "    ngt-primitive,    ngt-bone,    ngt-group,    ngt-lod,    ngt-points,    ngt-mesh,    ngt-instanced-mesh,    ngt-skinned-mesh,    ngt-audio,    ngt-positional-audio,    ngt-line,    ngt-line-loop,    ngt-line-segments,    ngt-light-probe,    ngt-ambient-light,    ngt-ambient-light-probe,    ngt-hemisphere-light,    ngt-hemisphere-light-probe,    ngt-directional-light,    ngt-point-light,    ngt-spot-light,    ngt-rect-area-light,    ngt-arrow-helper,    ngt-axes-helper,    ngt-box3-helper,    ngt-grid-helper,    ngt-plane-helper,    ngt-polar-grid-helper,    ngt-sprite,    ngt-camera,    ngt-perspective-camera,    ngt-orthographic-camera,    ngt-array-camera,    ngt-stereo-camera,    ngt-cube-camera,    ngt-soba-plane,    ngt-soba-box,    ngt-soba-cylinder,    ngt-soba-cone,    ngt-soba-circle,    ngt-soba-sphere,    ngt-soba-tube,    ngt-soba-torus,    ngt-soba-tetrahedron,    ngt-soba-ring,    ngt-soba-polyhedron,    ngt-soba-octahedron,    ngt-soba-dodecahedron,    ngt-soba-icosahedron,    ngt-soba-extrude,    ngt-soba-lathe,    ngt-soba-torus-knot,    ngt-soba-billboard,    ngt-soba-detailed,    ngt-soba-line,    ngt-soba-quadratic-bezier-line,    ngt-soba-cubic-bezier-line,    ngt-soba-orthographic-camera,    ngt-soba-gizmo-helper,    ngt-soba-gizmo-viewport,    ngt-soba-gizmo-axis-head,    ngt-soba-text  ", inputs: ["name", "position", "rotation", "quaternion", "scale", "color", "userData", "castShadow", "receiveShadow", "visible", "matrixAutoUpdate", "dispose", "raycast", "appendMode", "appendTo", "object3dInputsController"], outputs: ["click", "contextmenu", "dblclick", "pointerup", "pointerdown", "pointerover", "pointerout", "pointerenter", "pointerleave", "pointermove", "pointermissed", "pointercancel", "wheel"], exportAs: ["ngtObject3dInputsController"] }], pipes: { "async": i3.AsyncPipe }, changeDetection: i0.ChangeDetectionStrategy.OnPush });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelper, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngt-soba-gizmo-helper',
                    template: `
    <ng-container *ngIf="virtualScene$ | async as virtualScene">
      <ngt-group
        *ngIf="gizmoProps$ | async as gizmoProps"
        (ready)="onGizmoReady($event)"
        [appendTo]="virtualScene"
        [position]="[gizmoProps.x, gizmoProps.y, 0]"
        [object3dInputsController]="gizmoProps.objectInputsController"
      ></ngt-group>
      <ngt-soba-orthographic-camera
        [appendTo]="virtualScene"
        [makeDefault]="false"
        [position]="[0, 0, 200]"
        (ready)="onCameraReady($event)"
      ></ngt-soba-orthographic-camera>
      <ng-content></ng-content>
    </ng-container>
  `,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    providers: [
                        NGT_OBJECT_INPUTS_CONTROLLER_PROVIDER,
                        NgtSobaGizmoHelperStore,
                        {
                            provide: NgtSobaExtender,
                            useExisting: NgtSobaGizmoHelper,
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtSobaGizmoHelperStore }]; }, propDecorators: { alignment: [{
                type: Input
            }], margin: [{
                type: Input
            }], renderPriority: [{
                type: Input
            }] } });
export class NgtSobaGizmoHelperModule {
}
NgtSobaGizmoHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaGizmoHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelperModule, declarations: [NgtSobaGizmoHelper], imports: [NgtCoreModule,
        NgtGroupModule,
        NgtSobaOrthographicCameraModule,
        CommonModule], exports: [NgtSobaGizmoHelper] });
NgtSobaGizmoHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelperModule, imports: [[
            NgtCoreModule,
            NgtGroupModule,
            NgtSobaOrthographicCameraModule,
            CommonModule,
        ]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaGizmoHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaGizmoHelper],
                    exports: [NgtSobaGizmoHelper],
                    imports: [
                        NgtCoreModule,
                        NgtGroupModule,
                        NgtSobaOrthographicCameraModule,
                        CommonModule,
                    ],
                }]
        }] });
