// @angular-three/core/esm2020/helpers/lib/box3-helper/box3-helper.directive.mjs
// GENERATED
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtBox3Helper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.Box3Helper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtBox3Helper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3Helper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBox3Helper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBox3Helper, selector: "ngt-box3-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtBox3Helper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtBox3Helper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3Helper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-box3-helper',
                    exportAs: 'ngtBox3Helper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtBox3Helper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtBox3HelperModule {
}
NgtBox3HelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBox3HelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, declarations: [NgtBox3Helper], exports: [NgtBox3Helper, NgtObject3dControllerModule] });
NgtBox3HelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBox3HelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBox3Helper],
                    exports: [NgtBox3Helper, NgtObject3dControllerModule],
                }]
        }] });
