// @angular-three/core/esm2020/materials/lib/mesh-depth-material/mesh-depth-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshDepthMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshDepthMaterial;
    }
}
NgtMeshDepthMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshDepthMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshDepthMaterial, selector: "ngt-mesh-depth-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshDepthMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshDepthMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-depth-material',
                    exportAs: 'ngtMeshDepthMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshDepthMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshDepthMaterialModule {
}
NgtMeshDepthMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshDepthMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, declarations: [NgtMeshDepthMaterial], exports: [NgtMeshDepthMaterial] });
NgtMeshDepthMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDepthMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshDepthMaterial],
                    exports: [NgtMeshDepthMaterial],
                }]
        }] });
