// @angular-three/core/esm2020/lib/directives/cursor.directive.mjs
import { DOCUMENT } from '@angular/common';
import { Directive, Inject, NgModule, Optional, Self } from '@angular/core';
import { mapTo, merge, takeUntil } from 'rxjs';
import { NgtObject3dInputsController } from '../controllers/object-3d-inputs.controller';
import { NgtDestroyedService } from '../services/destroyed.service';
import * as i0 from "@angular/core";
import * as i1 from "../controllers/object-3d-inputs.controller";
import * as i2 from "../services/destroyed.service";
export class NgtCursor {
    constructor(objectInputsController, document, destroyed) {
        if (!objectInputsController)
            return;
        merge(objectInputsController.pointerover.pipe(mapTo(true)), objectInputsController.pointerout.pipe(mapTo(false)))
            .pipe(takeUntil(destroyed))
            .subscribe((hovered) => {
            document.body.style.cursor = hovered ? 'pointer' : 'auto';
        });
    }
}
NgtCursor.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursor, deps: [{ token: i1.NgtObject3dInputsController, optional: true, self: true }, { token: DOCUMENT }, { token: i2.NgtDestroyedService }], target: i0.ɵɵFactoryTarget.Directive });
NgtCursor.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCursor, selector: "[ngtCursor]", providers: [NgtDestroyedService], exportAs: ["ngtCursor"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursor, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtCursor]',
                    exportAs: 'ngtCursor',
                    providers: [NgtDestroyedService],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dInputsController, decorators: [{
                    type: Optional
                }, {
                    type: Self
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: i2.NgtDestroyedService }]; } });
export class NgtCursorModule {
}
NgtCursorModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursorModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCursorModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursorModule, declarations: [NgtCursor], exports: [NgtCursor] });
NgtCursorModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursorModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCursorModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCursor],
                    exports: [NgtCursor],
                }]
        }] });
