// @angular-three/core/esm2020/attributes/lib/int16-buffer-attribute/int16-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInt16BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int16BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt16BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt16BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt16BufferAttribute, selector: "ngt-int16-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt16BufferAttribute,
        },
    ], exportAs: ["ngtInt16BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int16-buffer-attribute',
                    exportAs: 'ngtInt16BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt16BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInt16BufferAttributeModule {
}
NgtInt16BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt16BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, declarations: [NgtInt16BufferAttribute], exports: [NgtInt16BufferAttribute] });
NgtInt16BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt16BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt16BufferAttribute],
                    exports: [NgtInt16BufferAttribute],
                }]
        }] });
