// @angular-three/core/esm2020/geometries/lib/wireframe-geometry/wireframe-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtWireframeGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.WireframeGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtWireframeGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtWireframeGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtWireframeGeometry, selector: "ngt-wireframe-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtWireframeGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtWireframeGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-wireframe-geometry',
                    exportAs: 'ngtWireframeGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtWireframeGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtWireframeGeometryModule {
}
NgtWireframeGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtWireframeGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometryModule, declarations: [NgtWireframeGeometry], exports: [NgtWireframeGeometry] });
NgtWireframeGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtWireframeGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtWireframeGeometry],
                    exports: [NgtWireframeGeometry],
                }]
        }] });
