// @angular-three/soba/esm2020/loaders/lib/texture-loader/texture-loader.service.mjs
import { NgtLoaderService, NgtStore } from '@angular-three/core';
import { Injectable } from '@angular/core';
import { combineLatest, map, tap } from 'rxjs';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export const IsObject = (url) => url === Object(url) && !Array.isArray(url) && typeof url !== 'function';
export class NgtTextureLoaderService {
    constructor(store, loaderService) {
        this.store = store;
        this.loaderService = loaderService;
    }
    load(input) {
        return combineLatest([
            this.store.select('renderer'),
            this.loaderService.use(THREE.TextureLoader, IsObject(input) ? Object.values(input) : input),
        ]).pipe(tap(([renderer, textures]) => {
            if (renderer) {
                (Array.isArray(textures) ? textures : [textures]).forEach(renderer.initTexture.bind(renderer));
            }
        }), map(([, textures]) => {
            if (IsObject(input)) {
                return Object.keys(input).reduce((record, key, index) => {
                    record[key] = textures[index];
                    return record;
                }, {});
            }
            return textures;
        }));
    }
}
NgtTextureLoaderService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTextureLoaderService, deps: [{ token: i1.NgtStore }, { token: i1.NgtLoaderService }], target: i0.ɵɵFactoryTarget.Injectable });
NgtTextureLoaderService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTextureLoaderService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTextureLoaderService, decorators: [{
            type: Injectable,
            args: [{ providedIn: 'root' }]
        }], ctorParameters: function () { return [{ type: i1.NgtStore }, { type: i1.NgtLoaderService }]; } });
