// @angular-three/core/esm2020/geometries/lib/buffer-geometry/buffer-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtBufferGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.BufferGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtBufferGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtBufferGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtBufferGeometry, selector: "ngt-buffer-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtBufferGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtBufferGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-buffer-geometry',
                    exportAs: 'ngtBufferGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtBufferGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtBufferGeometryModule {
}
NgtBufferGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtBufferGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometryModule, declarations: [NgtBufferGeometry], exports: [NgtBufferGeometry] });
NgtBufferGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtBufferGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtBufferGeometry],
                    exports: [NgtBufferGeometry],
                }]
        }] });
