// @angular-three/core/esm2020/textures/lib/video-texture/video-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtVideoTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.VideoTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtVideoTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtVideoTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtVideoTexture, selector: "ngt-video-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtVideoTexture,
        },
    ], exportAs: ["ngtVideoTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-video-texture',
                    exportAs: 'ngtVideoTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtVideoTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtVideoTextureModule {
}
NgtVideoTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtVideoTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTextureModule, declarations: [NgtVideoTexture], exports: [NgtVideoTexture] });
NgtVideoTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVideoTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtVideoTexture],
                    exports: [NgtVideoTexture],
                }]
        }] });
