// @angular-three/core/esm2020/attributes/lib/uint8-clamped-buffer-attribute/uint8-clamped-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtUint8ClampedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint8ClampedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint8ClampedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint8ClampedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint8ClampedBufferAttribute, selector: "ngt-uint8-clamped-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint8ClampedBufferAttribute,
        },
    ], exportAs: ["ngtUint8ClampedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint8-clamped-buffer-attribute',
                    exportAs: 'ngtUint8ClampedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint8ClampedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtUint8ClampedBufferAttributeModule {
}
NgtUint8ClampedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint8ClampedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, declarations: [NgtUint8ClampedBufferAttribute], exports: [NgtUint8ClampedBufferAttribute] });
NgtUint8ClampedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint8ClampedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint8ClampedBufferAttribute],
                    exports: [NgtUint8ClampedBufferAttribute],
                }]
        }] });
