// @angular-three/core/esm2020/helpers/lib/hemisphere-light-helper/hemisphere-light-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtHemisphereLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.HemisphereLightHelper;
    }
    set ngtHemisphereLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtHemisphereLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLightHelper, selector: "[ngtHemisphereLightHelper]", inputs: { ngtHemisphereLightHelper: "ngtHemisphereLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtHemisphereLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtHemisphereLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtHemisphereLightHelper]',
                    exportAs: 'ngtHemisphereLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtHemisphereLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtHemisphereLightHelper: [{
                type: Input
            }] } });
export class NgtHemisphereLightHelperModule {
}
NgtHemisphereLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, declarations: [NgtHemisphereLightHelper], exports: [NgtHemisphereLightHelper] });
NgtHemisphereLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLightHelper],
                    exports: [NgtHemisphereLightHelper],
                }]
        }] });
