// @angular-three/core/esm2020/materials/lib/mesh-phong-material/mesh-phong-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshPhongMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshPhongMaterial;
    }
}
NgtMeshPhongMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshPhongMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshPhongMaterial, selector: "ngt-mesh-phong-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshPhongMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshPhongMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-phong-material',
                    exportAs: 'ngtMeshPhongMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshPhongMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshPhongMaterialModule {
}
NgtMeshPhongMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshPhongMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, declarations: [NgtMeshPhongMaterial], exports: [NgtMeshPhongMaterial] });
NgtMeshPhongMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshPhongMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshPhongMaterial],
                    exports: [NgtMeshPhongMaterial],
                }]
        }] });
