// @angular-three/core/esm2020/attributes/lib/int8-buffer-attribute/int8-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInt8BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int8BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt8BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt8BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt8BufferAttribute, selector: "ngt-int8-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt8BufferAttribute,
        },
    ], exportAs: ["ngtInt8BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int8-buffer-attribute',
                    exportAs: 'ngtInt8BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt8BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInt8BufferAttributeModule {
}
NgtInt8BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt8BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, declarations: [NgtInt8BufferAttribute], exports: [NgtInt8BufferAttribute] });
NgtInt8BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt8BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt8BufferAttribute],
                    exports: [NgtInt8BufferAttribute],
                }]
        }] });
