// @angular-three/core/esm2020/helpers/lib/spot-light-helper/spot-light-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSpotLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.SpotLightHelper;
    }
    set ngtSpotLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtSpotLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpotLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpotLightHelper, selector: "[ngtSpotLightHelper]", inputs: { ngtSpotLightHelper: "ngtSpotLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtSpotLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSpotLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtSpotLightHelper]',
                    exportAs: 'ngtSpotLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtSpotLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtSpotLightHelper: [{
                type: Input
            }] } });
export class NgtSpotLightHelperModule {
}
NgtSpotLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpotLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, declarations: [NgtSpotLightHelper], exports: [NgtSpotLightHelper] });
NgtSpotLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpotLightHelper],
                    exports: [NgtSpotLightHelper],
                }]
        }] });
