// @angular-three/core/fesm2015/angular-three-core-lights.mjs
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule } from '@angular-three/core';
import * as i0 from '@angular/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';

// GENERATED
class NgtLightProbe extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.LightProbe;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtLightProbe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbe, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLightProbe.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLightProbe, selector: "ngt-light-probe", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtLightProbe,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtLightProbe"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbe, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-light-probe',
                    exportAs: 'ngtLightProbe',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtLightProbe,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtLightProbeModule {
}
NgtLightProbeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLightProbeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbeModule, declarations: [NgtLightProbe], exports: [NgtLightProbe, NgtObject3dControllerModule] });
NgtLightProbeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbeModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLightProbeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLightProbe],
                    exports: [NgtLightProbe, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtAmbientLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.AmbientLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtAmbientLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAmbientLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAmbientLight, selector: "ngt-ambient-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtAmbientLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAmbientLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ambient-light',
                    exportAs: 'ngtAmbientLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtAmbientLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtAmbientLightModule {
}
NgtAmbientLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAmbientLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, declarations: [NgtAmbientLight], exports: [NgtAmbientLight, NgtObject3dControllerModule] });
NgtAmbientLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAmbientLight],
                    exports: [NgtAmbientLight, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtAmbientLightProbe extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.AmbientLightProbe;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtAmbientLightProbe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbe, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAmbientLightProbe.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAmbientLightProbe, selector: "ngt-ambient-light-probe", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtAmbientLightProbe,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAmbientLightProbe"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbe, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ambient-light-probe',
                    exportAs: 'ngtAmbientLightProbe',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtAmbientLightProbe,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtAmbientLightProbeModule {
}
NgtAmbientLightProbeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAmbientLightProbeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, declarations: [NgtAmbientLightProbe], exports: [NgtAmbientLightProbe, NgtObject3dControllerModule] });
NgtAmbientLightProbeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAmbientLightProbe],
                    exports: [NgtAmbientLightProbe, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtHemisphereLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.HemisphereLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtHemisphereLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLight, selector: "ngt-hemisphere-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtHemisphereLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtHemisphereLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-hemisphere-light',
                    exportAs: 'ngtHemisphereLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtHemisphereLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtHemisphereLightModule {
}
NgtHemisphereLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, declarations: [NgtHemisphereLight], exports: [NgtHemisphereLight, NgtObject3dControllerModule] });
NgtHemisphereLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLight],
                    exports: [NgtHemisphereLight, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtHemisphereLightProbe extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.HemisphereLightProbe;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtHemisphereLightProbe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbe, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLightProbe.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLightProbe, selector: "ngt-hemisphere-light-probe", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtHemisphereLightProbe,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtHemisphereLightProbe"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbe, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-hemisphere-light-probe',
                    exportAs: 'ngtHemisphereLightProbe',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtHemisphereLightProbe,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtHemisphereLightProbeModule {
}
NgtHemisphereLightProbeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightProbeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, declarations: [NgtHemisphereLightProbe], exports: [NgtHemisphereLightProbe, NgtObject3dControllerModule] });
NgtHemisphereLightProbeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLightProbe],
                    exports: [NgtHemisphereLightProbe, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtDirectionalLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.DirectionalLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtDirectionalLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDirectionalLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDirectionalLight, selector: "ngt-directional-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtDirectionalLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtDirectionalLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-directional-light',
                    exportAs: 'ngtDirectionalLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtDirectionalLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtDirectionalLightModule {
}
NgtDirectionalLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDirectionalLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, declarations: [NgtDirectionalLight], exports: [NgtDirectionalLight, NgtObject3dControllerModule] });
NgtDirectionalLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDirectionalLight],
                    exports: [NgtDirectionalLight, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtPointLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.PointLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtPointLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPointLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPointLight, selector: "ngt-point-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtPointLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPointLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-point-light',
                    exportAs: 'ngtPointLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtPointLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtPointLightModule {
}
NgtPointLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightModule, declarations: [NgtPointLight], exports: [NgtPointLight, NgtObject3dControllerModule] });
NgtPointLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPointLight],
                    exports: [NgtPointLight, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtSpotLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.SpotLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtSpotLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpotLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpotLight, selector: "ngt-spot-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtSpotLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtSpotLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-spot-light',
                    exportAs: 'ngtSpotLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtSpotLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtSpotLightModule {
}
NgtSpotLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpotLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, declarations: [NgtSpotLight], exports: [NgtSpotLight, NgtObject3dControllerModule] });
NgtSpotLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpotLight],
                    exports: [NgtSpotLight, NgtObject3dControllerModule],
                }]
        }] });

// GENERATED
class NgtRectAreaLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.RectAreaLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtRectAreaLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtRectAreaLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtRectAreaLight, selector: "ngt-rect-area-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtRectAreaLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtRectAreaLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-rect-area-light',
                    exportAs: 'ngtRectAreaLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtRectAreaLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
class NgtRectAreaLightModule {
}
NgtRectAreaLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtRectAreaLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, declarations: [NgtRectAreaLight], exports: [NgtRectAreaLight, NgtObject3dControllerModule] });
NgtRectAreaLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRectAreaLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtRectAreaLight],
                    exports: [NgtRectAreaLight, NgtObject3dControllerModule],
                }]
        }] });

//GENERATED

/**
 * Generated bundle index. Do not edit.
 */

export { NgtAmbientLight, NgtAmbientLightModule, NgtAmbientLightProbe, NgtAmbientLightProbeModule, NgtDirectionalLight, NgtDirectionalLightModule, NgtHemisphereLight, NgtHemisphereLightModule, NgtHemisphereLightProbe, NgtHemisphereLightProbeModule, NgtLightProbe, NgtLightProbeModule, NgtPointLight, NgtPointLightModule, NgtRectAreaLight, NgtRectAreaLightModule, NgtSpotLight, NgtSpotLightModule };
