// @angular-three/core/esm2020/stats/lib/stats.directive.mjs
var _NgtStats_node, _NgtStats_stats, _NgtStats_animationUuid;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { NgtAnimationFrameStore } from '@angular-three/core';
import { DOCUMENT } from '@angular/common';
import { Directive, Inject, Input, NgModule, NgZone, } from '@angular/core';
// import Stats from minified bundle for minimizing bundle size
// @ts-ignore
import Stats from 'three/examples/js/libs/stats.min';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtStats {
    constructor(animationFrameStore, ngZone, document) {
        this.animationFrameStore = animationFrameStore;
        this.ngZone = ngZone;
        _NgtStats_node.set(this, void 0);
        _NgtStats_stats.set(this, void 0);
        _NgtStats_animationUuid.set(this, '');
        __classPrivateFieldSet(this, _NgtStats_node, document.body, "f");
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            if (this.parent) {
                __classPrivateFieldSet(this, _NgtStats_node, this.parent, "f");
            }
            __classPrivateFieldSet(this, _NgtStats_stats, new Stats(), "f");
            __classPrivateFieldGet(this, _NgtStats_node, "f").appendChild(__classPrivateFieldGet(this, _NgtStats_stats, "f").dom);
            __classPrivateFieldSet(this, _NgtStats_animationUuid, this.animationFrameStore.register({
                callback: () => {
                    __classPrivateFieldGet(this, _NgtStats_stats, "f").update();
                },
            }), "f");
        });
    }
    ngOnDestroy() {
        this.ngZone.runOutsideAngular(() => {
            if (__classPrivateFieldGet(this, _NgtStats_stats, "f")) {
                this.animationFrameStore.actions.unsubscriberUuid(__classPrivateFieldGet(this, _NgtStats_animationUuid, "f"));
                __classPrivateFieldGet(this, _NgtStats_stats, "f").end();
                __classPrivateFieldGet(this, _NgtStats_node, "f").removeChild(__classPrivateFieldGet(this, _NgtStats_stats, "f").dom);
            }
        });
    }
}
_NgtStats_node = new WeakMap(), _NgtStats_stats = new WeakMap(), _NgtStats_animationUuid = new WeakMap();
NgtStats.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStats, deps: [{ token: i1.NgtAnimationFrameStore }, { token: i0.NgZone }, { token: DOCUMENT }], target: i0.ɵɵFactoryTarget.Directive });
NgtStats.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtStats, selector: "ngt-stats", inputs: { parent: "parent" }, exportAs: ["ngtStats"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStats, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-stats',
                    exportAs: 'ngtStats',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtAnimationFrameStore }, { type: i0.NgZone }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }]; }, propDecorators: { parent: [{
                type: Input
            }] } });
export class NgtStatsModule {
}
NgtStatsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStatsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtStatsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStatsModule, declarations: [NgtStats], exports: [NgtStats] });
NgtStatsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStatsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtStatsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtStats],
                    exports: [NgtStats],
                }]
        }] });
