// @angular-three/soba/esm2020/performances/preload/preload.directive.mjs
var _NgtSobaPreload_instances, _NgtSobaPreload_precompileParams_get;
import { __classPrivateFieldGet } from "tslib";
import { EnhancedRxState, NgtStore } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtSobaPreload extends EnhancedRxState {
    constructor(store) {
        super();
        this.store = store;
        _NgtSobaPreload_instances.add(this);
        this.set({
            all: false,
            scene: undefined,
            camera: undefined,
        });
    }
    set all(v) {
        this.set({ all: v });
    }
    set scene(v) {
        this.set({ scene: v });
    }
    set camera(v) {
        this.set({ camera: v });
    }
    ngOnInit() {
        this.hold(this.store.select('ready'), (ready) => {
            if (ready) {
                const { all, scene, camera, renderer } = __classPrivateFieldGet(this, _NgtSobaPreload_instances, "a", _NgtSobaPreload_precompileParams_get);
                const invisible = [];
                if (all) {
                    scene.traverse((object) => {
                        if (!object.visible) {
                            invisible.push(object);
                            object.visible = true;
                        }
                    });
                }
                // Now compile
                renderer.compile(scene, camera);
                // And for good measure, hit it with a cube camera
                const cubeRenderTarget = new THREE.WebGLCubeRenderTarget(128);
                const cubeCamera = new THREE.CubeCamera(0.01, 100000, cubeRenderTarget);
                cubeCamera.update(renderer, scene);
                cubeRenderTarget.dispose();
                // Flips these objects back
                invisible.forEach((object) => (object.visible = false));
            }
        });
    }
}
_NgtSobaPreload_instances = new WeakSet(), _NgtSobaPreload_precompileParams_get = function _NgtSobaPreload_precompileParams_get() {
    const { all, scene, camera } = this.get();
    const { camera: canvasCamera, scene: canvasScene, renderer, } = this.store.get();
    return {
        all,
        scene: scene || canvasScene,
        camera: camera || canvasCamera,
        renderer,
    };
};
NgtSobaPreload.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreload, deps: [{ token: i1.NgtStore }], target: i0.ɵɵFactoryTarget.Directive });
NgtSobaPreload.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaPreload, selector: "ngt-soba-preload", inputs: { all: "all", scene: "scene", camera: "camera" }, exportAs: ["ngtSobaPreload"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreload, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-soba-preload',
                    exportAs: 'ngtSobaPreload',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtStore }]; }, propDecorators: { all: [{
                type: Input
            }], scene: [{
                type: Input
            }], camera: [{
                type: Input
            }] } });
export class NgtSobaPreloadModule {
}
NgtSobaPreloadModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreloadModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaPreloadModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreloadModule, declarations: [NgtSobaPreload], exports: [NgtSobaPreload] });
NgtSobaPreloadModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreloadModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaPreloadModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaPreload],
                    exports: [NgtSobaPreload],
                }]
        }] });
