// @angular-three/core/esm2020/geometries/lib/ring-geometry/ring-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtRingGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.RingGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtRingGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtRingGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtRingGeometry, selector: "ngt-ring-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtRingGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtRingGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ring-geometry',
                    exportAs: 'ngtRingGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtRingGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtRingGeometryModule {
}
NgtRingGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtRingGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometryModule, declarations: [NgtRingGeometry], exports: [NgtRingGeometry] });
NgtRingGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtRingGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtRingGeometry],
                    exports: [NgtRingGeometry],
                }]
        }] });
