// @angular-three/core/esm2020/materials/lib/line-dashed-material/line-dashed-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLineDashedMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.LineDashedMaterial;
    }
}
NgtLineDashedMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineDashedMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineDashedMaterial, selector: "ngt-line-dashed-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtLineDashedMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtLineDashedMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-dashed-material',
                    exportAs: 'ngtLineDashedMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtLineDashedMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtLineDashedMaterialModule {
}
NgtLineDashedMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineDashedMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, declarations: [NgtLineDashedMaterial], exports: [NgtLineDashedMaterial] });
NgtLineDashedMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineDashedMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineDashedMaterial],
                    exports: [NgtLineDashedMaterial],
                }]
        }] });
