// @angular-three/core/esm2020/materials/lib/mesh-basic-material/mesh-basic-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshBasicMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshBasicMaterial;
    }
}
NgtMeshBasicMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshBasicMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshBasicMaterial, selector: "ngt-mesh-basic-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshBasicMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshBasicMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-basic-material',
                    exportAs: 'ngtMeshBasicMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshBasicMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshBasicMaterialModule {
}
NgtMeshBasicMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshBasicMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, declarations: [NgtMeshBasicMaterial], exports: [NgtMeshBasicMaterial] });
NgtMeshBasicMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshBasicMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshBasicMaterial],
                    exports: [NgtMeshBasicMaterial],
                }]
        }] });
