// @angular-three/core/esm2020/curves/lib/line-curve3/line-curve3.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLineCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.LineCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtLineCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineCurve3, selector: "ngt-line-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtLineCurve3,
        },
    ], exportAs: ["ngtLineCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-curve3',
                    exportAs: 'ngtLineCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtLineCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtLineCurve3Module {
}
NgtLineCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, declarations: [NgtLineCurve3], exports: [NgtLineCurve3] });
NgtLineCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineCurve3],
                    exports: [NgtLineCurve3],
                }]
        }] });
