// @angular-three/core/esm2020/attributes/lib/float32-buffer-attribute/float32-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtFloat32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat32BufferAttribute, selector: "ngt-float32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat32BufferAttribute,
        },
    ], exportAs: ["ngtFloat32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float32-buffer-attribute',
                    exportAs: 'ngtFloat32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtFloat32BufferAttributeModule {
}
NgtFloat32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, declarations: [NgtFloat32BufferAttribute], exports: [NgtFloat32BufferAttribute] });
NgtFloat32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat32BufferAttribute],
                    exports: [NgtFloat32BufferAttribute],
                }]
        }] });
