// @angular-three/core/esm2020/helpers/lib/point-light-helper/point-light-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPointLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.PointLightHelper;
    }
    set ngtPointLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtPointLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPointLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPointLightHelper, selector: "[ngtPointLightHelper]", inputs: { ngtPointLightHelper: "ngtPointLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtPointLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtPointLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtPointLightHelper]',
                    exportAs: 'ngtPointLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtPointLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtPointLightHelper: [{
                type: Input
            }] } });
export class NgtPointLightHelperModule {
}
NgtPointLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPointLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, declarations: [NgtPointLightHelper], exports: [NgtPointLightHelper] });
NgtPointLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPointLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPointLightHelper],
                    exports: [NgtPointLightHelper],
                }]
        }] });
