// @angular-three/core/esm2020/lib/core.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgtCanvas } from './canvas.component';
import * as i0 from "@angular/core";
export class NgtCoreModule {
}
NgtCoreModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCoreModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCoreModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCoreModule, declarations: [NgtCanvas], imports: [CommonModule], exports: [NgtCanvas] });
NgtCoreModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCoreModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCoreModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCanvas],
                    imports: [CommonModule],
                    exports: [NgtCanvas],
                }]
        }] });
