// @angular-three/core/esm2020/lights/lib/ambient-light/ambient-light.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtAmbientLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.AmbientLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtAmbientLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAmbientLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAmbientLight, selector: "ngt-ambient-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtAmbientLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAmbientLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ambient-light',
                    exportAs: 'ngtAmbientLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtAmbientLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtAmbientLightModule {
}
NgtAmbientLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAmbientLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, declarations: [NgtAmbientLight], exports: [NgtAmbientLight, NgtObject3dControllerModule] });
NgtAmbientLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAmbientLight],
                    exports: [NgtAmbientLight, NgtObject3dControllerModule],
                }]
        }] });
