// @angular-three/core/esm2020/audios/lib/audio-listener/audio-listener.directive.mjs
var _NgtAudioListener_listener;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { applyProps, EnhancedRxState, NgtStore } from '@angular-three/core';
import { Directive, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtAudioListener extends EnhancedRxState {
    constructor(store) {
        super();
        this.store = store;
        this.ready = new EventEmitter();
        _NgtAudioListener_listener.set(this, void 0);
    }
    get listener() {
        return __classPrivateFieldGet(this, _NgtAudioListener_listener, "f");
    }
    ngOnInit() {
        __classPrivateFieldSet(this, _NgtAudioListener_listener, new THREE.AudioListener(), "f");
        const props = {
            filter: this.filter,
            timeDelta: this.timeDelta,
        };
        applyProps(this.listener, props);
        this.holdEffect(this.store.select('ready'), (ready) => {
            const camera = this.store.get('camera');
            if (ready && camera) {
                camera.add(this.listener);
                this.ready.emit(this.listener);
            }
            return () => {
                if (ready && camera) {
                    this.listener.clear();
                    camera.remove(this.listener);
                }
            };
        });
    }
}
_NgtAudioListener_listener = new WeakMap();
NgtAudioListener.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListener, deps: [{ token: i1.NgtStore }], target: i0.ɵɵFactoryTarget.Directive });
NgtAudioListener.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAudioListener, selector: "ngt-audio-listener", inputs: { filter: "filter", timeDelta: "timeDelta" }, outputs: { ready: "ready" }, exportAs: ["ngtAudioListener"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListener, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-audio-listener',
                    exportAs: 'ngtAudioListener',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtStore }]; }, propDecorators: { filter: [{
                type: Input
            }], timeDelta: [{
                type: Input
            }], ready: [{
                type: Output
            }] } });
export class NgtAudioListenerModule {
}
NgtAudioListenerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListenerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAudioListenerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListenerModule, declarations: [NgtAudioListener], exports: [NgtAudioListener] });
NgtAudioListenerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListenerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioListenerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAudioListener],
                    exports: [NgtAudioListener],
                }]
        }] });
