// @angular-three/soba/esm2020/controls/lib/fly-controls/fly-controls.directive.mjs
import { EnhancedRxState, NgtAnimationFrameStore, NgtLoopService, NgtStore, } from '@angular-three/core';
import { Directive, EventEmitter, NgModule, NgZone, Output, } from '@angular/core';
import { selectSlice } from '@rx-angular/state';
import { FlyControls } from 'three-stdlib';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtSobaFlyControls extends EnhancedRxState {
    constructor(store, loopService, animationFrameStore, ngZone) {
        super();
        this.store = store;
        this.loopService = loopService;
        this.animationFrameStore = animationFrameStore;
        this.ngZone = ngZone;
        this.ready = this.select('controls');
        this.change = new EventEmitter();
    }
    ngOnInit() {
        this.holdEffect(this.select('controls'), (controls) => {
            const callback = (e) => {
                this.loopService.invalidate();
                if (this.change.observed) {
                    this.change.emit(e);
                }
            };
            controls.addEventListener('change', callback);
            return () => {
                controls.removeEventListener('change', callback);
            };
        });
        this.hold(this.store.select(selectSlice(['camera', 'renderer'])), ({ renderer, camera }) => {
            this.ngZone.runOutsideAngular(() => {
                if (camera && renderer) {
                    this.set({
                        controls: new FlyControls(camera, renderer.domElement),
                    });
                }
            });
        });
        this.holdEffect(this.select('controls'), (controls) => {
            const animationUuid = this.animationFrameStore.register({
                callback: ({ delta }) => {
                    controls.update(delta);
                },
            });
            return () => {
                this.animationFrameStore.actions.unsubscriberUuid(animationUuid);
            };
        });
    }
    get controls() {
        return this.get('controls');
    }
}
NgtSobaFlyControls.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControls, deps: [{ token: i1.NgtStore }, { token: i1.NgtLoopService }, { token: i1.NgtAnimationFrameStore }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtSobaFlyControls.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaFlyControls, selector: "ngt-soba-fly-controls", outputs: { ready: "ready", change: "change" }, exportAs: ["ngtSobaFlyControls"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControls, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-soba-fly-controls',
                    exportAs: 'ngtSobaFlyControls',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtStore }, { type: i1.NgtLoopService }, { type: i1.NgtAnimationFrameStore }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }], change: [{
                type: Output
            }] } });
export class NgtSobaFlyControlsModule {
}
NgtSobaFlyControlsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControlsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaFlyControlsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControlsModule, declarations: [NgtSobaFlyControls], exports: [NgtSobaFlyControls] });
NgtSobaFlyControlsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControlsModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaFlyControlsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaFlyControls],
                    exports: [NgtSobaFlyControls],
                }]
        }] });
