// @angular-three/core/esm2020/materials/lib/sprite-material/sprite-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSpriteMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.SpriteMaterial;
    }
}
NgtSpriteMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpriteMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpriteMaterial, selector: "ngt-sprite-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtSpriteMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSpriteMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-sprite-material',
                    exportAs: 'ngtSpriteMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtSpriteMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtSpriteMaterialModule {
}
NgtSpriteMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpriteMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, declarations: [NgtSpriteMaterial], exports: [NgtSpriteMaterial] });
NgtSpriteMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpriteMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpriteMaterial],
                    exports: [NgtSpriteMaterial],
                }]
        }] });
