// @angular-three/core/esm2020/textures/lib/data-texture3-d/data-texture3-d.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDataTexture3D extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.DataTexture3D;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtDataTexture3D.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3D, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDataTexture3D.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDataTexture3D, selector: "ngt-data-texture3-d", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtDataTexture3D,
        },
    ], exportAs: ["ngtDataTexture3D"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3D, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-data-texture3-d',
                    exportAs: 'ngtDataTexture3D',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtDataTexture3D,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDataTexture3DModule {
}
NgtDataTexture3DModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3DModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDataTexture3DModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3DModule, declarations: [NgtDataTexture3D], exports: [NgtDataTexture3D] });
NgtDataTexture3DModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3DModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture3DModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDataTexture3D],
                    exports: [NgtDataTexture3D],
                }]
        }] });
