// @angular-three/soba/esm2020/staging/stars/star-material.directive.mjs
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { Directive, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class StarFieldMaterial extends THREE.ShaderMaterial {
    constructor() {
        super({
            uniforms: { time: { value: 0.0 }, fade: { value: 1.0 } },
            vertexShader: /* glsl */ `
      uniform float time;
      attribute float size;
      varying vec3 vColor;
      void main() {
        vColor = color;
        vec4 mvPosition = modelViewMatrix * vec4(position, 0.5);
        gl_PointSize = size * (30.0 / -mvPosition.z) * (3.0 + sin(mvPosition.x + 2.0 * time + 100.0));
        gl_Position = projectionMatrix * mvPosition;
      }`,
            fragmentShader: /* glsl */ `
      uniform sampler2D pointTexture;
      uniform float fade;
      varying vec3 vColor;
      void main() {
        float opacity = 1.0;
        if (fade == 1.0) {
          float d = distance(gl_PointCoord, vec2(0.5, 0.5));
          opacity = 1.0 / (1.0 + exp(16.0 * (d - 0.25)));
        }
        gl_FragColor = vec4(vColor, opacity);
      }`,
        });
    }
}
export class NgtSobaStarFieldMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = StarFieldMaterial;
    }
}
NgtSobaStarFieldMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSobaStarFieldMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSobaStarFieldMaterial, selector: "ngt-soba-star-field-material", providers: [
        { provide: NgtMaterial, useExisting: NgtSobaStarFieldMaterial },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSobaStarFieldMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-soba-star-field-material',
                    exportAs: 'ngtSobaStarFieldMaterial',
                    providers: [
                        { provide: NgtMaterial, useExisting: NgtSobaStarFieldMaterial },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtSobaStarFieldMaterialModule {
}
NgtSobaStarFieldMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSobaStarFieldMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterialModule, declarations: [NgtSobaStarFieldMaterial], exports: [NgtSobaStarFieldMaterial] });
NgtSobaStarFieldMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSobaStarFieldMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSobaStarFieldMaterial],
                    exports: [NgtSobaStarFieldMaterial],
                }]
        }] });
