// @angular-three/core/esm2020/geometries/lib/icosahedron-geometry/icosahedron-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtIcosahedronGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.IcosahedronGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtIcosahedronGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtIcosahedronGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtIcosahedronGeometry, selector: "ngt-icosahedron-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtIcosahedronGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtIcosahedronGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-icosahedron-geometry',
                    exportAs: 'ngtIcosahedronGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtIcosahedronGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtIcosahedronGeometryModule {
}
NgtIcosahedronGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtIcosahedronGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometryModule, declarations: [NgtIcosahedronGeometry], exports: [NgtIcosahedronGeometry] });
NgtIcosahedronGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtIcosahedronGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtIcosahedronGeometry],
                    exports: [NgtIcosahedronGeometry],
                }]
        }] });
