// @angular-three/core/esm2020/lights/lib/ambient-light-probe/ambient-light-probe.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtAmbientLightProbe extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.AmbientLightProbe;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtAmbientLightProbe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbe, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAmbientLightProbe.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAmbientLightProbe, selector: "ngt-ambient-light-probe", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtAmbientLightProbe,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAmbientLightProbe"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbe, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ambient-light-probe',
                    exportAs: 'ngtAmbientLightProbe',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtAmbientLightProbe,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtAmbientLightProbeModule {
}
NgtAmbientLightProbeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAmbientLightProbeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, declarations: [NgtAmbientLightProbe], exports: [NgtAmbientLightProbe, NgtObject3dControllerModule] });
NgtAmbientLightProbeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAmbientLightProbeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAmbientLightProbe],
                    exports: [NgtAmbientLightProbe, NgtObject3dControllerModule],
                }]
        }] });
