// @angular-three/core/esm2020/meshes/instanced-mesh/instanced-mesh.directive.mjs
import { NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, NGT_OBJECT_POST_INIT, NGT_OBJECT_TYPE, NgtCommonMesh, NgtMaterialGeometryController, NgtMaterialGeometryControllerModule, } from '@angular-three/core';
import { Directive, Inject, Input, NgModule, NgZone, Optional, } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtInstancedMesh extends NgtCommonMesh {
    constructor(materialGeometryController, ngZone) {
        super(materialGeometryController, ngZone);
    }
    set args(value) {
        if (this.materialGeometryController) {
            this.materialGeometryController.meshArgs = value;
        }
    }
}
NgtInstancedMesh.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMesh, deps: [{ token: NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER, optional: true }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtInstancedMesh.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInstancedMesh, selector: "ngt-instanced-mesh", inputs: { args: "args" }, providers: [
        { provide: NgtCommonMesh, useExisting: NgtInstancedMesh },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        { provide: NGT_OBJECT_TYPE, useValue: THREE.InstancedMesh },
        {
            provide: NGT_OBJECT_POST_INIT,
            useValue: (object) => {
                object.instanceMatrix.setUsage(THREE.DynamicDrawUsage);
            },
        },
    ], exportAs: ["ngtInstancedMesh"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMesh, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-instanced-mesh',
                    exportAs: 'ngtInstancedMesh',
                    providers: [
                        { provide: NgtCommonMesh, useExisting: NgtInstancedMesh },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        { provide: NGT_OBJECT_TYPE, useValue: THREE.InstancedMesh },
                        {
                            provide: NGT_OBJECT_POST_INIT,
                            useValue: (object) => {
                                object.instanceMatrix.setUsage(THREE.DynamicDrawUsage);
                            },
                        },
                    ],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtMaterialGeometryController, decorators: [{
                    type: Optional
                }, {
                    type: Inject,
                    args: [NGT_MATERIAL_GEOMETRY_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInstancedMeshModule {
}
NgtInstancedMeshModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMeshModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInstancedMeshModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMeshModule, declarations: [NgtInstancedMesh], exports: [NgtInstancedMesh, NgtMaterialGeometryControllerModule] });
NgtInstancedMeshModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMeshModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedMeshModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInstancedMesh],
                    exports: [NgtInstancedMesh, NgtMaterialGeometryControllerModule],
                }]
        }] });
