// @angular-three/core/esm2020/cameras/lib/array-camera/array-camera.directive.mjs
// GENERATED
import { NgtCommonCamera, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtArrayCamera extends NgtCommonCamera {
    constructor() {
        super(...arguments);
        this.cameraType = THREE.ArrayCamera;
    }
    set args(v) {
        this.cameraArgs = v;
    }
}
NgtArrayCamera.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCamera, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtArrayCamera.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtArrayCamera, selector: "ngt-array-camera", inputs: { args: "args" }, providers: [
        {
            provide: NgtCommonCamera,
            useExisting: NgtArrayCamera,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtArrayCamera"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCamera, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-array-camera',
                    exportAs: 'ngtArrayCamera',
                    providers: [
                        {
                            provide: NgtCommonCamera,
                            useExisting: NgtArrayCamera,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtArrayCameraModule {
}
NgtArrayCameraModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCameraModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtArrayCameraModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCameraModule, declarations: [NgtArrayCamera], exports: [NgtArrayCamera, NgtObject3dControllerModule] });
NgtArrayCameraModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCameraModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtArrayCameraModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtArrayCamera],
                    exports: [NgtArrayCamera, NgtObject3dControllerModule],
                }]
        }] });
