// @angular-three/core/esm2020/curves/lib/ellipse-curve/ellipse-curve.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtEllipseCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.EllipseCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtEllipseCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtEllipseCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtEllipseCurve, selector: "ngt-ellipse-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtEllipseCurve,
        },
    ], exportAs: ["ngtEllipseCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-ellipse-curve',
                    exportAs: 'ngtEllipseCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtEllipseCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtEllipseCurveModule {
}
NgtEllipseCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtEllipseCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, declarations: [NgtEllipseCurve], exports: [NgtEllipseCurve] });
NgtEllipseCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtEllipseCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtEllipseCurve],
                    exports: [NgtEllipseCurve],
                }]
        }] });
