// @angular-three/core/esm2020/primitive/lib/primitive.directive.mjs
var _NgtPrimitive_object;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { NGT_OBJECT_CONTROLLER_PROVIDER, NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, NgtObject3dControllerModule, } from '@angular-three/core';
import { Directive, EventEmitter, Inject, Input, NgModule, NgZone, Output, } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtPrimitive {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtPrimitive_object.set(this, void 0);
    }
    set object(value) {
        __classPrivateFieldSet(this, _NgtPrimitive_object, value, "f");
        this.objectController.initFn = () => value;
        this.objectController.readyFn = () => {
            this.ready.emit(this.object);
        };
    }
    get object() {
        return __classPrivateFieldGet(this, _NgtPrimitive_object, "f");
    }
    ngOnChanges() {
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
}
_NgtPrimitive_object = new WeakMap();
NgtPrimitive.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitive, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtPrimitive.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPrimitive, selector: "ngt-primitive[object]", inputs: { object: "object" }, outputs: { ready: "ready" }, providers: [NGT_OBJECT_CONTROLLER_PROVIDER], exportAs: ["ngtPrimitive"], usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitive, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-primitive[object]',
                    exportAs: 'ngtPrimitive',
                    providers: [NGT_OBJECT_CONTROLLER_PROVIDER],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }], object: [{
                type: Input
            }] } });
export class NgtPrimitiveModule {
}
NgtPrimitiveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitiveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPrimitiveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitiveModule, declarations: [NgtPrimitive], exports: [NgtPrimitive, NgtObject3dControllerModule] });
NgtPrimitiveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitiveModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPrimitiveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPrimitive],
                    exports: [NgtPrimitive, NgtObject3dControllerModule],
                }]
        }] });
