// @angular-three/core/esm2020/attributes/lib/float64-buffer-attribute/float64-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtFloat64BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Float64BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtFloat64BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtFloat64BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtFloat64BufferAttribute, selector: "ngt-float64-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtFloat64BufferAttribute,
        },
    ], exportAs: ["ngtFloat64BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-float64-buffer-attribute',
                    exportAs: 'ngtFloat64BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtFloat64BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtFloat64BufferAttributeModule {
}
NgtFloat64BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFloat64BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, declarations: [NgtFloat64BufferAttribute], exports: [NgtFloat64BufferAttribute] });
NgtFloat64BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFloat64BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFloat64BufferAttribute],
                    exports: [NgtFloat64BufferAttribute],
                }]
        }] });
