// @angular-three/core/esm2020/textures/lib/canvas-texture/canvas-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCanvasTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.CanvasTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtCanvasTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCanvasTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCanvasTexture, selector: "ngt-canvas-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtCanvasTexture,
        },
    ], exportAs: ["ngtCanvasTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-canvas-texture',
                    exportAs: 'ngtCanvasTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtCanvasTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCanvasTextureModule {
}
NgtCanvasTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCanvasTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTextureModule, declarations: [NgtCanvasTexture], exports: [NgtCanvasTexture] });
NgtCanvasTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCanvasTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCanvasTexture],
                    exports: [NgtCanvasTexture],
                }]
        }] });
