// @angular-three/core/esm2020/textures/lib/depth-texture/depth-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDepthTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.DepthTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtDepthTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDepthTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDepthTexture, selector: "ngt-depth-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtDepthTexture,
        },
    ], exportAs: ["ngtDepthTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-depth-texture',
                    exportAs: 'ngtDepthTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtDepthTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDepthTextureModule {
}
NgtDepthTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDepthTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTextureModule, declarations: [NgtDepthTexture], exports: [NgtDepthTexture] });
NgtDepthTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDepthTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDepthTexture],
                    exports: [NgtDepthTexture],
                }]
        }] });
