// @angular-three/core/esm2020/geometries/lib/circle-geometry/circle-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCircleGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.CircleGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtCircleGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCircleGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCircleGeometry, selector: "ngt-circle-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtCircleGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtCircleGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-circle-geometry',
                    exportAs: 'ngtCircleGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtCircleGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCircleGeometryModule {
}
NgtCircleGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCircleGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometryModule, declarations: [NgtCircleGeometry], exports: [NgtCircleGeometry] });
NgtCircleGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCircleGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCircleGeometry],
                    exports: [NgtCircleGeometry],
                }]
        }] });
