// @angular-three/core/esm2020/materials/lib/mesh-matcap-material/mesh-matcap-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshMatcapMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshMatcapMaterial;
    }
}
NgtMeshMatcapMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshMatcapMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshMatcapMaterial, selector: "ngt-mesh-matcap-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshMatcapMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshMatcapMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-matcap-material',
                    exportAs: 'ngtMeshMatcapMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshMatcapMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshMatcapMaterialModule {
}
NgtMeshMatcapMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshMatcapMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, declarations: [NgtMeshMatcapMaterial], exports: [NgtMeshMatcapMaterial] });
NgtMeshMatcapMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshMatcapMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshMatcapMaterial],
                    exports: [NgtMeshMatcapMaterial],
                }]
        }] });
