// @angular-three/core/esm2020/helpers/lib/camera-helper/camera-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCameraHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.CameraHelper;
    }
    set ngtCameraHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtCameraHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCameraHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCameraHelper, selector: "[ngtCameraHelper]", inputs: { ngtCameraHelper: "ngtCameraHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtCameraHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtCameraHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtCameraHelper]',
                    exportAs: 'ngtCameraHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtCameraHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtCameraHelper: [{
                type: Input
            }] } });
export class NgtCameraHelperModule {
}
NgtCameraHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCameraHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, declarations: [NgtCameraHelper], exports: [NgtCameraHelper] });
NgtCameraHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCameraHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCameraHelper],
                    exports: [NgtCameraHelper],
                }]
        }] });
