// @angular-three/core/esm2020/materials/lib/mesh-distance-material/mesh-distance-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshDistanceMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshDistanceMaterial;
    }
}
NgtMeshDistanceMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshDistanceMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshDistanceMaterial, selector: "ngt-mesh-distance-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshDistanceMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshDistanceMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-distance-material',
                    exportAs: 'ngtMeshDistanceMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshDistanceMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshDistanceMaterialModule {
}
NgtMeshDistanceMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshDistanceMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, declarations: [NgtMeshDistanceMaterial], exports: [NgtMeshDistanceMaterial] });
NgtMeshDistanceMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshDistanceMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshDistanceMaterial],
                    exports: [NgtMeshDistanceMaterial],
                }]
        }] });
