// @angular-three/core/esm2020/materials/lib/mesh-normal-material/mesh-normal-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshNormalMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshNormalMaterial;
    }
}
NgtMeshNormalMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshNormalMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshNormalMaterial, selector: "ngt-mesh-normal-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshNormalMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshNormalMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-normal-material',
                    exportAs: 'ngtMeshNormalMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshNormalMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshNormalMaterialModule {
}
NgtMeshNormalMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshNormalMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, declarations: [NgtMeshNormalMaterial], exports: [NgtMeshNormalMaterial] });
NgtMeshNormalMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshNormalMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshNormalMaterial],
                    exports: [NgtMeshNormalMaterial],
                }]
        }] });
