// @angular-three/core/esm2020/textures/lib/cube-texture/cube-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCubeTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.CubeTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtCubeTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCubeTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubeTexture, selector: "ngt-cube-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtCubeTexture,
        },
    ], exportAs: ["ngtCubeTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cube-texture',
                    exportAs: 'ngtCubeTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtCubeTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCubeTextureModule {
}
NgtCubeTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubeTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTextureModule, declarations: [NgtCubeTexture], exports: [NgtCubeTexture] });
NgtCubeTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubeTexture],
                    exports: [NgtCubeTexture],
                }]
        }] });
