// @angular-three/core/esm2020/geometries/lib/tube-geometry/tube-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtTubeGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.TubeGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtTubeGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtTubeGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtTubeGeometry, selector: "ngt-tube-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtTubeGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtTubeGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-tube-geometry',
                    exportAs: 'ngtTubeGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtTubeGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtTubeGeometryModule {
}
NgtTubeGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtTubeGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometryModule, declarations: [NgtTubeGeometry], exports: [NgtTubeGeometry] });
NgtTubeGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtTubeGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtTubeGeometry],
                    exports: [NgtTubeGeometry],
                }]
        }] });
