// @angular-three/core/esm2020/cameras/lib/cube-camera/cube-camera.directive.mjs
var _NgtCubeCamera_cubeCamera;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { NGT_OBJECT_CONTROLLER_PROVIDER, NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, } from '@angular-three/core';
import { Directive, EventEmitter, Inject, Input, NgModule, NgZone, Output, } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtCubeCamera {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtCubeCamera_cubeCamera.set(this, void 0);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtCubeCamera_cubeCamera, new THREE.CubeCamera(...this.args), "f");
            return __classPrivateFieldGet(this, _NgtCubeCamera_cubeCamera, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.cubeCamera);
        };
    }
    get cubeCamera() {
        return __classPrivateFieldGet(this, _NgtCubeCamera_cubeCamera, "f");
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
}
_NgtCubeCamera_cubeCamera = new WeakMap();
NgtCubeCamera.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCamera, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtCubeCamera.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubeCamera, selector: "ngt-cube-camera", inputs: { args: "args" }, outputs: { ready: "ready" }, providers: [NGT_OBJECT_CONTROLLER_PROVIDER], exportAs: ["ngtCubeCamera"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCamera, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cube-camera',
                    exportAs: 'ngtCubeCamera',
                    providers: [NGT_OBJECT_CONTROLLER_PROVIDER],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }], args: [{
                type: Input
            }] } });
export class NgtCubeCameraModule {
}
NgtCubeCameraModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCameraModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubeCameraModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCameraModule, declarations: [NgtCubeCamera], exports: [NgtCubeCamera] });
NgtCubeCameraModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCameraModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubeCameraModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubeCamera],
                    exports: [NgtCubeCamera],
                }]
        }] });
