// @rx-angular/cdk/esm2015/render-strategies/lib/native-strategies.js
import { ɵmarkDirty as markDirty } from '@angular/core';
import { Observable } from 'rxjs';
import { tap } from 'rxjs/operators';
import { coalesceWith } from '@rx-angular/cdk/coalescing';
import { cancelAnimationFrame, requestAnimationFrame, } from '@rx-angular/cdk/zone-less';
const animationFrameTick = () => new Observable((subscriber) => {
    const id = requestAnimationFrame(() => {
        subscriber.next(0);
        subscriber.complete();
    });
    return () => {
        cancelAnimationFrame(id);
    };
});
const ɵ0 = animationFrameTick;
const ɵ1 = (cdRef, _, notification) => {
    cdRef.detectChanges();
}, ɵ2 = (work, scope) => (o$) => o$.pipe(coalesceWith(animationFrameTick(), scope), tap(() => work()));
const localCredentials = {
    name: 'local',
    work: ɵ1,
    behavior: ɵ2,
};
const ɵ3 = (_, context) => markDirty(context), ɵ4 = (work) => (o$) => o$.pipe(tap(() => work()));
const globalCredentials = {
    name: 'global',
    work: ɵ3,
    behavior: ɵ4,
};
const ɵ5 = () => void 0, ɵ6 = () => (o$) => o$;
const noopCredentials = {
    name: 'noop',
    work: ɵ5,
    behavior: ɵ6,
};
const ɵ7 = (cdRef) => cdRef.markForCheck(), ɵ8 = (work) => (o$) => o$.pipe(tap(() => work()));
const nativeCredentials = {
    name: 'native',
    work: ɵ7,
    behavior: ɵ8,
};
export const RX_NATIVE_STRATEGIES = {
    global: globalCredentials,
    native: nativeCredentials,
    noop: noopCredentials,
    local: localCredentials,
};
export { ɵ0, ɵ1, ɵ2, ɵ3, ɵ4, ɵ5, ɵ6, ɵ7, ɵ8 };
