// @angular-three/core/esm2020/group/lib/group.directive.mjs
var _NgtGroup_group;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { NGT_OBJECT_CONTROLLER_PROVIDER, NGT_OBJECT_WATCHED_CONTROLLER, NgtObject3dController, NgtObject3dControllerModule, } from '@angular-three/core';
import { Directive, EventEmitter, Inject, NgModule, NgZone, Output, } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
import * as i1 from "@angular-three/core";
export class NgtGroup {
    constructor(objectController, ngZone) {
        this.objectController = objectController;
        this.ngZone = ngZone;
        this.ready = new EventEmitter();
        _NgtGroup_group.set(this, void 0);
        objectController.initFn = () => {
            __classPrivateFieldSet(this, _NgtGroup_group, new THREE.Group(), "f");
            return __classPrivateFieldGet(this, _NgtGroup_group, "f");
        };
        objectController.readyFn = () => {
            this.ready.emit(this.group);
        };
    }
    get group() {
        return __classPrivateFieldGet(this, _NgtGroup_group, "f");
    }
    ngOnInit() {
        this.ngZone.runOutsideAngular(() => {
            this.objectController.init();
        });
    }
}
_NgtGroup_group = new WeakMap();
NgtGroup.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroup, deps: [{ token: NGT_OBJECT_WATCHED_CONTROLLER }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtGroup.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtGroup, selector: "ngt-group", outputs: { ready: "ready" }, providers: [NGT_OBJECT_CONTROLLER_PROVIDER], exportAs: ["ngtGroup"], ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroup, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-group',
                    exportAs: 'ngtGroup',
                    providers: [NGT_OBJECT_CONTROLLER_PROVIDER],
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtObject3dController, decorators: [{
                    type: Inject,
                    args: [NGT_OBJECT_WATCHED_CONTROLLER]
                }] }, { type: i0.NgZone }]; }, propDecorators: { ready: [{
                type: Output
            }] } });
export class NgtGroupModule {
}
NgtGroupModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroupModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtGroupModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroupModule, declarations: [NgtGroup], exports: [NgtGroup, NgtObject3dControllerModule] });
NgtGroupModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroupModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGroupModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtGroup],
                    exports: [NgtGroup, NgtObject3dControllerModule],
                }]
        }] });
