// @angular-three/core/esm2020/curves/lib/cubic-bezier-curve/cubic-bezier-curve.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCubicBezierCurve extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CubicBezierCurve;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCubicBezierCurve.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCubicBezierCurve.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubicBezierCurve, selector: "ngt-cubic-bezier-curve", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCubicBezierCurve,
        },
    ], exportAs: ["ngtCubicBezierCurve"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cubic-bezier-curve',
                    exportAs: 'ngtCubicBezierCurve',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCubicBezierCurve,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCubicBezierCurveModule {
}
NgtCubicBezierCurveModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubicBezierCurveModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, declarations: [NgtCubicBezierCurve], exports: [NgtCubicBezierCurve] });
NgtCubicBezierCurveModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurveModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubicBezierCurve],
                    exports: [NgtCubicBezierCurve],
                }]
        }] });
