// @angular-three/core/esm2020/lib/services/loader.service.mjs
import { Injectable } from '@angular/core';
import { catchError, defer, forkJoin, map, of, ReplaySubject, share, tap, throwError, } from 'rxjs';
import { buildGraph } from '../utils/build-graph';
import * as i0 from "@angular/core";
export class NgtLoaderService {
    constructor() {
        this.cached = new Map();
    }
    use(loaderConstructor, input, extensions, onProgress) {
        const keys = (Array.isArray(input) ? input : [input]);
        const loader = new loaderConstructor();
        if (extensions) {
            extensions(loader);
        }
        const results$ = forkJoin(keys.map((key) => {
            if (this.cached.has(key)) {
                return of(this.cached.get(key));
            }
            return defer(() => loader.loadAsync(key, onProgress)).pipe(tap((data) => {
                if (data.scene) {
                    Object.assign(data, buildGraph(data.scene));
                }
                this.cached.set(key, data);
            }), catchError((err) => {
                console.error(`Error loading ${key}: ${err.message}`);
                return throwError(err);
            }));
        }));
        return defer(() => Array.isArray(input)
            ? results$
            : results$.pipe(map((results) => results[0]))).pipe(share({
            connector: () => new ReplaySubject(),
            resetOnRefCountZero: true,
            resetOnError: true,
        }));
    }
    ngOnDestroy() {
        this.cached.clear();
    }
}
NgtLoaderService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoaderService, deps: [], target: i0.ɵɵFactoryTarget.Injectable });
NgtLoaderService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoaderService, providedIn: 'root' });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoaderService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }] });
