// @angular-three/core/esm2020/lib/services/loop.service.mjs
var _NgtLoopService_running, _NgtLoopService_repeat, _NgtLoopService_frames;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Injectable, NgZone } from '@angular/core';
import { requestAnimationFrame } from '@rx-angular/cdk/zone-less';
import { NgtAnimationFrameStore } from '../stores/animation-frame.store';
import { NgtStore } from '../stores/store';
import * as i0 from "@angular/core";
import * as i1 from "../stores/store";
import * as i2 from "../stores/animation-frame.store";
export class NgtLoopService {
    constructor(store, animationFrameStore, ngZone) {
        this.store = store;
        this.animationFrameStore = animationFrameStore;
        this.ngZone = ngZone;
        _NgtLoopService_running.set(this, false);
        _NgtLoopService_repeat.set(this, void 0);
        _NgtLoopService_frames.set(this, 1);
    }
    render(timestamp, state, animationFrameState) {
        return this.ngZone.runOutsideAngular(() => {
            const { clock, frameloop, camera, scene, renderer, mouse, size, viewport, } = state;
            const { subscribers, hasPriority } = animationFrameState;
            if (renderer) {
                let delta = clock.getDelta();
                // In frameloop='never' mode, clock times are updated using the provided timestamp
                if (frameloop === 'never' && typeof timestamp === 'number') {
                    delta = timestamp - clock.elapsedTime;
                    clock.oldTime = clock.elapsedTime;
                    clock.elapsedTime = timestamp;
                }
                const renderState = {
                    clock,
                    camera,
                    scene,
                    renderer,
                    mouse,
                    size,
                    viewport,
                    delta,
                };
                for (const subscriber of subscribers) {
                    subscriber.callback(renderState, subscriber.obj);
                }
                if (!hasPriority) {
                    renderer.render(scene, camera);
                }
            }
            __classPrivateFieldSet(this, _NgtLoopService_frames, Math.max(0, __classPrivateFieldGet(this, _NgtLoopService_frames, "f") - 1), "f");
            return state.frameloop === 'always' ? 1 : __classPrivateFieldGet(this, _NgtLoopService_frames, "f");
        });
    }
    loop(timestamp) {
        return this.ngZone.runOutsideAngular(() => {
            __classPrivateFieldSet(this, _NgtLoopService_running, true, "f");
            __classPrivateFieldSet(this, _NgtLoopService_repeat, 0, "f");
            const state = this.store.get();
            if (state.ready && (state.frameloop === 'always' || __classPrivateFieldGet(this, _NgtLoopService_frames, "f") > 0)) {
                __classPrivateFieldSet(this, _NgtLoopService_repeat, __classPrivateFieldGet(this, _NgtLoopService_repeat, "f") + this.render(timestamp, state, this.animationFrameStore.get()), "f");
            }
            if (__classPrivateFieldGet(this, _NgtLoopService_repeat, "f") > 0)
                return requestAnimationFrame(this.loop.bind(this));
            __classPrivateFieldSet(this, _NgtLoopService_running, false, "f");
            return;
        });
    }
    invalidate(state = this.store.get()) {
        if (state.vr) {
            state.renderer?.setAnimationLoop((timestamp) => {
                this.render(timestamp, this.store.get(), this.animationFrameStore.get());
            });
            return;
        }
        if (!state.ready || state.frameloop === 'never')
            return;
        // Increase frames, do not go higher than 60
        __classPrivateFieldSet(this, _NgtLoopService_frames, Math.min(60, __classPrivateFieldGet(this, _NgtLoopService_frames, "f") + 1), "f");
        // If the render-loop isn't active, start it
        if (!__classPrivateFieldGet(this, _NgtLoopService_running, "f")) {
            __classPrivateFieldSet(this, _NgtLoopService_running, true, "f");
            requestAnimationFrame(this.loop.bind(this));
        }
    }
}
_NgtLoopService_running = new WeakMap(), _NgtLoopService_repeat = new WeakMap(), _NgtLoopService_frames = new WeakMap();
NgtLoopService.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoopService, deps: [{ token: i1.NgtStore }, { token: i2.NgtAnimationFrameStore }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Injectable });
NgtLoopService.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoopService });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLoopService, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i1.NgtStore }, { type: i2.NgtAnimationFrameStore }, { type: i0.NgZone }]; } });
