// @angular-three/core/esm2020/lib/services/resize.service.mjs
import { DOCUMENT } from '@angular/common';
import { ElementRef, Inject, Injectable, NgZone } from '@angular/core';
import { debounceTime, fromEvent, Observable, pipe, ReplaySubject, share, Subject, takeUntil, } from 'rxjs';
import { NGT_RESIZE_OBSERVER_SUPPORT, NGT_RESIZE_OPTIONS, } from '../di/resize';
import * as i0 from "@angular/core";
export class NgtResize extends Observable {
    constructor({ nativeElement }, ngZone, document, isSupport, { box, offsetSize, scroll, debounce }) {
        let observer;
        let lastBounds;
        let lastEntries = [];
        const torndown$ = new Subject();
        const scrollContainers = findScrollContainers(nativeElement);
        // set actual debounce values early, so effects know if they should react accordingly
        const scrollDebounce = debounce
            ? typeof debounce === 'number'
                ? debounce
                : debounce.scroll
            : null;
        const resizeDebounce = debounce
            ? typeof debounce === 'number'
                ? debounce
                : debounce.resize
            : null;
        const debounceAndDestroy = (debounce) => {
            return pipe(debounceTime(debounce ?? 0), takeUntil(torndown$));
        };
        super((subscriber) => {
            if (!isSupport) {
                subscriber.error('ResizeObserver is not supported in your browser. Please use a polyfill');
                return;
            }
            ngZone.runOutsideAngular(() => {
                const callback = (entries) => {
                    lastEntries = entries;
                    const { left, top, width, height, bottom, right, x, y } = nativeElement.getBoundingClientRect();
                    const size = {
                        left,
                        top,
                        width,
                        height,
                        bottom,
                        right,
                        x,
                        y,
                    };
                    if (nativeElement instanceof HTMLElement && offsetSize) {
                        size.height = nativeElement.offsetHeight;
                        size.width = nativeElement.offsetWidth;
                    }
                    Object.freeze(size);
                    subscriber.next({
                        entries,
                        dpr: document.defaultView.devicePixelRatio,
                        ...size,
                    });
                    if (!areBoundsEqual(lastBounds || {}, size)) {
                        lastBounds = size;
                    }
                };
                const boundEntriesCallback = () => {
                    callback(lastEntries);
                };
                observer = new ResizeObserver(callback);
                observer.observe(nativeElement, { box });
                if (scroll) {
                    if (scrollContainers) {
                        scrollContainers.forEach((scrollContainer) => {
                            fromEvent(scrollContainer, 'scroll', {
                                capture: true,
                                passive: true,
                            })
                                .pipe(debounceAndDestroy(scrollDebounce))
                                .subscribe(boundEntriesCallback);
                        });
                    }
                    fromEvent(document.defaultView, 'scroll', {
                        capture: true,
                        passive: true,
                    })
                        .pipe(debounceAndDestroy(scrollDebounce))
                        .subscribe(boundEntriesCallback);
                }
                fromEvent(document.defaultView, 'resize')
                    .pipe(debounceAndDestroy(resizeDebounce))
                    .subscribe(boundEntriesCallback);
            });
            return () => {
                if (observer) {
                    observer.unobserve(nativeElement);
                    observer.disconnect();
                }
                torndown$.next();
                torndown$.complete();
            };
        });
        return this.pipe(debounceTime(scrollDebounce || 0), share({
            connector: () => new ReplaySubject(1),
            resetOnRefCountZero: true,
            resetOnComplete: true,
        }));
    }
}
NgtResize.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtResize, deps: [{ token: ElementRef }, { token: NgZone }, { token: DOCUMENT }, { token: NGT_RESIZE_OBSERVER_SUPPORT }, { token: NGT_RESIZE_OPTIONS }], target: i0.ɵɵFactoryTarget.Injectable });
NgtResize.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtResize });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtResize, decorators: [{
            type: Injectable
        }], ctorParameters: function () { return [{ type: i0.ElementRef, decorators: [{
                    type: Inject,
                    args: [ElementRef]
                }] }, { type: i0.NgZone, decorators: [{
                    type: Inject,
                    args: [NgZone]
                }] }, { type: Document, decorators: [{
                    type: Inject,
                    args: [DOCUMENT]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGT_RESIZE_OBSERVER_SUPPORT]
                }] }, { type: undefined, decorators: [{
                    type: Inject,
                    args: [NGT_RESIZE_OPTIONS]
                }] }]; } });
// Returns a list of scroll offsets
function findScrollContainers(element) {
    const result = [];
    if (!element || element === document.body)
        return result;
    const { overflow, overflowX, overflowY } = window.getComputedStyle(element);
    if ([overflow, overflowX, overflowY].some((prop) => prop === 'auto' || prop === 'scroll'))
        result.push(element);
    return [
        ...result,
        ...findScrollContainers(element.parentElement),
    ];
}
// Checks if element boundaries are equal
const keys = [
    'x',
    'y',
    'top',
    'bottom',
    'left',
    'right',
    'width',
    'height',
];
const areBoundsEqual = (a, b) => keys.every((key) => a[key] === b[key]);
