// @angular-three/core/esm2020/cameras/lib/orthographic-camera/orthographic-camera.directive.mjs
// GENERATED
import { NgtCommonCamera, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtOrthographicCamera extends NgtCommonCamera {
    constructor() {
        super(...arguments);
        this.cameraType = THREE.OrthographicCamera;
    }
    set args(v) {
        this.cameraArgs = v;
    }
}
NgtOrthographicCamera.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCamera, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtOrthographicCamera.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtOrthographicCamera, selector: "ngt-orthographic-camera", inputs: { args: "args" }, providers: [
        {
            provide: NgtCommonCamera,
            useExisting: NgtOrthographicCamera,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtOrthographicCamera"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCamera, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-orthographic-camera',
                    exportAs: 'ngtOrthographicCamera',
                    providers: [
                        {
                            provide: NgtCommonCamera,
                            useExisting: NgtOrthographicCamera,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtOrthographicCameraModule {
}
NgtOrthographicCameraModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCameraModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtOrthographicCameraModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCameraModule, declarations: [NgtOrthographicCamera], exports: [NgtOrthographicCamera, NgtObject3dControllerModule] });
NgtOrthographicCameraModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCameraModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtOrthographicCameraModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtOrthographicCamera],
                    exports: [NgtOrthographicCamera, NgtObject3dControllerModule],
                }]
        }] });
