// @angular-three/core/esm2020/textures/lib/data-texture/data-texture.directive.mjs
// GENERATED
import { NgtTexture } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDataTexture extends NgtTexture {
    constructor() {
        super(...arguments);
        this.textureType = THREE.DataTexture;
    }
    set args(v) {
        this.textureArgs = v;
    }
}
NgtDataTexture.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDataTexture.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDataTexture, selector: "ngt-data-texture", inputs: { args: "args" }, providers: [
        {
            provide: NgtTexture,
            useExisting: NgtDataTexture,
        },
    ], exportAs: ["ngtDataTexture"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTexture, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-data-texture',
                    exportAs: 'ngtDataTexture',
                    providers: [
                        {
                            provide: NgtTexture,
                            useExisting: NgtDataTexture,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDataTextureModule {
}
NgtDataTextureModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTextureModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDataTextureModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTextureModule, declarations: [NgtDataTexture], exports: [NgtDataTexture] });
NgtDataTextureModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTextureModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDataTextureModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDataTexture],
                    exports: [NgtDataTexture],
                }]
        }] });
