// @angular-three/core/esm2020/geometries/lib/polyhedron-geometry/polyhedron-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPolyhedronGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.PolyhedronGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtPolyhedronGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPolyhedronGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPolyhedronGeometry, selector: "ngt-polyhedron-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtPolyhedronGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtPolyhedronGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-polyhedron-geometry',
                    exportAs: 'ngtPolyhedronGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtPolyhedronGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtPolyhedronGeometryModule {
}
NgtPolyhedronGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPolyhedronGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometryModule, declarations: [NgtPolyhedronGeometry], exports: [NgtPolyhedronGeometry] });
NgtPolyhedronGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolyhedronGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPolyhedronGeometry],
                    exports: [NgtPolyhedronGeometry],
                }]
        }] });
