// @angular-three/core/esm2020/helpers/lib/skeleton-helper/skeleton-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSkeletonHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.SkeletonHelper;
    }
    set ngtSkeletonHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtSkeletonHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSkeletonHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSkeletonHelper, selector: "[ngtSkeletonHelper]", inputs: { ngtSkeletonHelper: "ngtSkeletonHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtSkeletonHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSkeletonHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtSkeletonHelper]',
                    exportAs: 'ngtSkeletonHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtSkeletonHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtSkeletonHelper: [{
                type: Input
            }] } });
export class NgtSkeletonHelperModule {
}
NgtSkeletonHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSkeletonHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, declarations: [NgtSkeletonHelper], exports: [NgtSkeletonHelper] });
NgtSkeletonHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSkeletonHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSkeletonHelper],
                    exports: [NgtSkeletonHelper],
                }]
        }] });
