// @angular-three/core/esm2020/helpers/lib/axes-helper/axes-helper.directive.mjs
// GENERATED
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtAxesHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.AxesHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtAxesHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAxesHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAxesHelper, selector: "ngt-axes-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtAxesHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAxesHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-axes-helper',
                    exportAs: 'ngtAxesHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtAxesHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtAxesHelperModule {
}
NgtAxesHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAxesHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, declarations: [NgtAxesHelper], exports: [NgtAxesHelper, NgtObject3dControllerModule] });
NgtAxesHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAxesHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAxesHelper],
                    exports: [NgtAxesHelper, NgtObject3dControllerModule],
                }]
        }] });
