// @angular-three/core/esm2020/geometries/lib/cone-geometry/cone-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtConeGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.ConeGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtConeGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtConeGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtConeGeometry, selector: "ngt-cone-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtConeGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtConeGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cone-geometry',
                    exportAs: 'ngtConeGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtConeGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtConeGeometryModule {
}
NgtConeGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtConeGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometryModule, declarations: [NgtConeGeometry], exports: [NgtConeGeometry] });
NgtConeGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtConeGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtConeGeometry],
                    exports: [NgtConeGeometry],
                }]
        }] });
