// @angular-three/core/esm2020/helpers/lib/grid-helper/grid-helper.directive.mjs
// GENERATED
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtGridHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.GridHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtGridHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtGridHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtGridHelper, selector: "ngt-grid-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtGridHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtGridHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-grid-helper',
                    exportAs: 'ngtGridHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtGridHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtGridHelperModule {
}
NgtGridHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtGridHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, declarations: [NgtGridHelper], exports: [NgtGridHelper, NgtObject3dControllerModule] });
NgtGridHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtGridHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtGridHelper],
                    exports: [NgtGridHelper, NgtObject3dControllerModule],
                }]
        }] });
