// @angular-three/core/esm2020/lib/pipes/color.pipe.mjs
import { NgModule, Pipe } from '@angular/core';
import { makeColor } from '../utils/make';
import * as i0 from "@angular/core";
export class NgtColorPipe {
    /**
     * ConstructorParameters<typeof THREE.Color> has a limitation on THREE.Color constructor overloads
     */
    transform(args) {
        return makeColor(args);
    }
}
NgtColorPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtColorPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipe, name: "color" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'color',
                }]
        }] });
export class NgtColorPipeModule {
}
NgtColorPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtColorPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipeModule, declarations: [NgtColorPipe], exports: [NgtColorPipe] });
NgtColorPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtColorPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtColorPipe],
                    exports: [NgtColorPipe],
                }]
        }] });
