// @angular-three/core/esm2020/lib/pipes/fog.pipe.mjs
import { NgModule, Pipe } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtFogPipe {
    transform(value) {
        return new THREE.Fog(...value);
    }
}
NgtFogPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtFogPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipe, name: "fog" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fog',
                }]
        }] });
export class NgtFogExp2Pipe {
    transform(value) {
        return new THREE.FogExp2(...value);
    }
}
NgtFogExp2Pipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogExp2Pipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtFogExp2Pipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogExp2Pipe, name: "fogExp2" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogExp2Pipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'fogExp2',
                }]
        }] });
export class NgtFogPipeModule {
}
NgtFogPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtFogPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipeModule, declarations: [NgtFogPipe, NgtFogExp2Pipe], exports: [NgtFogPipe, NgtFogExp2Pipe] });
NgtFogPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtFogPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtFogPipe, NgtFogExp2Pipe],
                    exports: [NgtFogPipe, NgtFogExp2Pipe],
                }]
        }] });
