// @angular-three/core/esm2020/lib/pipes/math.pipe.mjs
import { NgModule, Pipe } from '@angular/core';
import * as i0 from "@angular/core";
export class NgtMathPipe {
    transform(value, keyOfMath) {
        const params = Array.isArray(value) ? value : [value];
        return Math[keyOfMath](...params);
    }
}
NgtMathPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtMathPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipe, name: "math" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'math',
                }]
        }] });
export class NgtMathConstantPipe {
    transform(value, keyOfMathConst) {
        if (keyOfMathConst === 'random') {
            return Math.random();
        }
        return value * Math[keyOfMathConst];
    }
}
NgtMathConstantPipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathConstantPipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtMathConstantPipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathConstantPipe, name: "mathConst" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathConstantPipe, decorators: [{
            type: Pipe,
            args: [{
                    name: 'mathConst',
                }]
        }] });
export class NgtMathPipeModule {
}
NgtMathPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMathPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipeModule, declarations: [NgtMathPipe, NgtMathConstantPipe], exports: [NgtMathPipe, NgtMathConstantPipe] });
NgtMathPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMathPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMathPipe, NgtMathConstantPipe],
                    exports: [NgtMathPipe, NgtMathConstantPipe],
                }]
        }] });
