// @angular-three/core/esm2020/lib/pipes/side.pipe.mjs
import { NgModule, Pipe } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSidePipe {
    transform(value, side) {
        switch (side) {
            case 'front':
                return THREE.FrontSide;
            case 'back':
                return THREE.BackSide;
            case 'double':
                return THREE.DoubleSide;
        }
    }
}
NgtSidePipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtSidePipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipe, name: "side" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipe, decorators: [{
            type: Pipe,
            args: [{ name: 'side' }]
        }] });
export class NgtSidePipeModule {
}
NgtSidePipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSidePipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipeModule, declarations: [NgtSidePipe], exports: [NgtSidePipe] });
NgtSidePipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSidePipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSidePipe],
                    exports: [NgtSidePipe],
                }]
        }] });
