// @angular-three/core/esm2020/lib/pipes/vector.pipe.mjs
import { NgModule, Pipe } from '@angular/core';
import { makeVector2, makeVector3, makeVector4 } from '../utils/make';
import * as i0 from "@angular/core";
export class NgtVector2Pipe {
    transform(value) {
        return makeVector2(value);
    }
}
NgtVector2Pipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector2Pipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtVector2Pipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector2Pipe, name: "vector2" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector2Pipe, decorators: [{
            type: Pipe,
            args: [{ name: 'vector2' }]
        }] });
export class NgtVector3Pipe {
    transform(value) {
        return makeVector3(value);
    }
}
NgtVector3Pipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector3Pipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtVector3Pipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector3Pipe, name: "vector3" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector3Pipe, decorators: [{
            type: Pipe,
            args: [{ name: 'vector3' }]
        }] });
export class NgtVector4Pipe {
    transform(value) {
        return makeVector4(value);
    }
}
NgtVector4Pipe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector4Pipe, deps: [], target: i0.ɵɵFactoryTarget.Pipe });
NgtVector4Pipe.ɵpipe = i0.ɵɵngDeclarePipe({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector4Pipe, name: "vector4" });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVector4Pipe, decorators: [{
            type: Pipe,
            args: [{ name: 'vector4' }]
        }] });
export class NgtVectorPipeModule {
}
NgtVectorPipeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVectorPipeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtVectorPipeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVectorPipeModule, declarations: [NgtVector2Pipe, NgtVector3Pipe, NgtVector4Pipe], exports: [NgtVector2Pipe, NgtVector3Pipe, NgtVector4Pipe] });
NgtVectorPipeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVectorPipeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtVectorPipeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtVector2Pipe, NgtVector3Pipe, NgtVector4Pipe],
                    exports: [NgtVector2Pipe, NgtVector3Pipe, NgtVector4Pipe],
                }]
        }] });
