// @angular-three/core/esm2020/helpers/lib/polar-grid-helper/polar-grid-helper.directive.mjs
// GENERATED
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPolarGridHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.PolarGridHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtPolarGridHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPolarGridHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPolarGridHelper, selector: "ngt-polar-grid-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtPolarGridHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPolarGridHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-polar-grid-helper',
                    exportAs: 'ngtPolarGridHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtPolarGridHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtPolarGridHelperModule {
}
NgtPolarGridHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPolarGridHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, declarations: [NgtPolarGridHelper], exports: [NgtPolarGridHelper, NgtObject3dControllerModule] });
NgtPolarGridHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPolarGridHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPolarGridHelper],
                    exports: [NgtPolarGridHelper, NgtObject3dControllerModule],
                }]
        }] });
