// @angular-three/core/esm2020/lib/controllers/animation-subscriber.controller.mjs
var _NgtAnimationSubscriberController_subscriber;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { Directive, EventEmitter, Input, NgModule, NgZone, Output, } from '@angular/core';
import { NgtAnimationFrameStore } from '../stores/animation-frame.store';
import { Controller, createControllerProviderFactory } from './controller';
import * as i0 from "@angular/core";
import * as i1 from "../stores/animation-frame.store";
export class NgtAnimationSubscriberController extends Controller {
    constructor(animationFrameStore, ngZone) {
        super(ngZone);
        this.animationFrameStore = animationFrameStore;
        this.priority = 0;
        this.animateReady = new EventEmitter();
        _NgtAnimationSubscriberController_subscriber.set(this, void 0);
    }
    prepare(obj) {
        // only subscribe to animation frame if there's an output handler
        if (this.animateReady.observed) {
            __classPrivateFieldSet(this, _NgtAnimationSubscriberController_subscriber, obj, "f");
            this.animationFrameStore.register({
                obj,
                callback: this.animateReady.emit.bind(this.animateReady),
                priority: this.priority,
            });
        }
    }
    ngOnDestroy() {
        if (__classPrivateFieldGet(this, _NgtAnimationSubscriberController_subscriber, "f")) {
            this.animationFrameStore.actions.unsubscriberUuid(__classPrivateFieldGet(this, _NgtAnimationSubscriberController_subscriber, "f").uuid);
        }
    }
    get controller() {
        return undefined;
    }
    get props() {
        return [];
    }
}
_NgtAnimationSubscriberController_subscriber = new WeakMap();
NgtAnimationSubscriberController.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberController, deps: [{ token: i1.NgtAnimationFrameStore }, { token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtAnimationSubscriberController.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAnimationSubscriberController, selector: "[animateReady]", inputs: { priority: "priority" }, outputs: { animateReady: "animateReady" }, exportAs: ["ngtAnimationSubscriberController"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberController, decorators: [{
            type: Directive,
            args: [{
                    selector: '[animateReady]',
                    exportAs: 'ngtAnimationSubscriberController',
                }]
        }], ctorParameters: function () { return [{ type: i1.NgtAnimationFrameStore }, { type: i0.NgZone }]; }, propDecorators: { priority: [{
                type: Input
            }], animateReady: [{
                type: Output
            }] } });
export class NgtAnimationSubscriberControllerModule {
}
NgtAnimationSubscriberControllerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberControllerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAnimationSubscriberControllerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberControllerModule, declarations: [NgtAnimationSubscriberController], exports: [NgtAnimationSubscriberController] });
NgtAnimationSubscriberControllerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberControllerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAnimationSubscriberControllerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAnimationSubscriberController],
                    exports: [NgtAnimationSubscriberController],
                }]
        }] });
export const [NGT_ANIMATION_SUBSCRIBER_WATCHED_CONTROLLER, NGT_ANIMATION_SUBSCRIBER_CONTROLLER_PROVIDER,] = createControllerProviderFactory({
    watchedControllerTokenName: 'Watched AnimationSubscriberController',
    controller: NgtAnimationSubscriberController,
});
