// @angular-three/core/esm2020/lib/controllers/content-geometry.controller.mjs
var _NgtContentGeometryController_instances, _NgtContentGeometryController_geometryInput, _NgtContentGeometryController_geometry, _NgtContentGeometryController_getGeometry;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
// GENERATED
import { Directive, Input, NgModule, NgZone } from '@angular/core';
import * as THREE from 'three';
import { Controller, createControllerProviderFactory } from './controller';
import * as i0 from "@angular/core";
export class NgtContentGeometryController extends Controller {
    constructor(ngZone) {
        super(ngZone);
        _NgtContentGeometryController_instances.add(this);
        _NgtContentGeometryController_geometryInput.set(this, void 0);
        _NgtContentGeometryController_geometry.set(this, undefined);
    }
    set geometry(v) {
        if (v) {
            __classPrivateFieldSet(this, _NgtContentGeometryController_geometryInput, v, "f");
            this.construct();
        }
    }
    get geometry() {
        return __classPrivateFieldGet(this, _NgtContentGeometryController_geometry, "f");
    }
    construct() {
        __classPrivateFieldSet(this, _NgtContentGeometryController_geometry, __classPrivateFieldGet(this, _NgtContentGeometryController_instances, "m", _NgtContentGeometryController_getGeometry).call(this, __classPrivateFieldGet(this, _NgtContentGeometryController_geometryInput, "f")), "f");
    }
    get controller() {
        return this.contentGeometryController;
    }
    get props() {
        return ['geometry'];
    }
}
_NgtContentGeometryController_geometryInput = new WeakMap(), _NgtContentGeometryController_geometry = new WeakMap(), _NgtContentGeometryController_instances = new WeakSet(), _NgtContentGeometryController_getGeometry = function _NgtContentGeometryController_getGeometry(input) {
    if (input instanceof THREE.BufferGeometry) {
        return input;
    }
    return undefined;
};
NgtContentGeometryController.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryController, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtContentGeometryController.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtContentGeometryController, selector: "\n    ngt-mesh,\n    ngt-instanced-mesh,\n    ngt-skinned-mesh,\n    ngt-line,\n    ngt-line-loop,\n    ngt-line-segments,\n    ngt-points\n  ", inputs: { geometry: "geometry", contentGeometryController: "contentGeometryController" }, exportAs: ["ngtContentGeometryController"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryController, decorators: [{
            type: Directive,
            args: [{
                    selector: `
    ngt-mesh,
    ngt-instanced-mesh,
    ngt-skinned-mesh,
    ngt-line,
    ngt-line-loop,
    ngt-line-segments,
    ngt-points
  `,
                    exportAs: 'ngtContentGeometryController',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { geometry: [{
                type: Input
            }], contentGeometryController: [{
                type: Input
            }] } });
export class NgtContentGeometryControllerModule {
}
NgtContentGeometryControllerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryControllerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtContentGeometryControllerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryControllerModule, declarations: [NgtContentGeometryController], exports: [NgtContentGeometryController] });
NgtContentGeometryControllerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryControllerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentGeometryControllerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtContentGeometryController],
                    exports: [NgtContentGeometryController],
                }]
        }] });
export const [NGT_CONTENT_GEOMETRY_WATCHED_CONTROLLER, NGT_CONTENT_GEOMETRY_CONTROLLER_PROVIDER,] = createControllerProviderFactory({
    controller: NgtContentGeometryController,
    watchedControllerTokenName: 'Watched ContentGeometryController',
});
