// @angular-three/core/esm2020/lib/controllers/content-material.controller.mjs
var _NgtContentMaterialController_instances, _NgtContentMaterialController_materialInput, _NgtContentMaterialController_material, _NgtContentMaterialController_getMaterial;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
// GENERATED
import { Directive, Input, NgModule, NgZone } from '@angular/core';
import * as THREE from 'three';
import { Controller, createControllerProviderFactory } from './controller';
import * as i0 from "@angular/core";
export class NgtContentMaterialController extends Controller {
    constructor(ngZone) {
        super(ngZone);
        _NgtContentMaterialController_instances.add(this);
        _NgtContentMaterialController_materialInput.set(this, void 0);
        this.isMaterialArray = false;
        _NgtContentMaterialController_material.set(this, undefined);
    }
    set material(v) {
        if (v) {
            if (!(Array.isArray(v) && !v.length)) {
                __classPrivateFieldSet(this, _NgtContentMaterialController_materialInput, v, "f");
            }
            this.construct();
        }
    }
    get material() {
        return __classPrivateFieldGet(this, _NgtContentMaterialController_material, "f");
    }
    construct() {
        __classPrivateFieldSet(this, _NgtContentMaterialController_material, __classPrivateFieldGet(this, _NgtContentMaterialController_instances, "m", _NgtContentMaterialController_getMaterial).call(this, __classPrivateFieldGet(this, _NgtContentMaterialController_materialInput, "f")), "f");
    }
    get controller() {
        return this.contentMaterialController;
    }
    get props() {
        return ['material', 'isArrayMaterial'];
    }
}
_NgtContentMaterialController_materialInput = new WeakMap(), _NgtContentMaterialController_material = new WeakMap(), _NgtContentMaterialController_instances = new WeakSet(), _NgtContentMaterialController_getMaterial = function _NgtContentMaterialController_getMaterial(input) {
    if (Array.isArray(input)) {
        if (!input.length)
            return undefined;
        if (input[0] instanceof THREE.Material) {
            return input;
        }
    }
    if (input instanceof THREE.Material) {
        return input;
    }
    return undefined;
};
NgtContentMaterialController.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialController, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
NgtContentMaterialController.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtContentMaterialController, selector: "\n    ngt-mesh,\n    ngt-instanced-mesh,\n    ngt-skinned-mesh,\n    ngt-line,\n    ngt-line-loop,\n    ngt-line-segments,\n    ngt-soba-plane,\n    ngt-soba-box,\n    ngt-soba-cylinder,\n    ngt-soba-cone,\n    ngt-soba-circle,\n    ngt-soba-sphere,\n    ngt-soba-tube,\n    ngt-soba-torus,\n    ngt-soba-tetrahedron,\n    ngt-soba-ring,\n    ngt-soba-polyhedron,\n    ngt-soba-octahedron,\n    ngt-soba-dodecahedron,\n    ngt-soba-icosahedron,\n    ngt-soba-extrude,\n    ngt-soba-lathe,\n    ngt-soba-torus-knot,\n    ngt-soba-text,,\n    ngt-points\n  ", inputs: { material: "material", isMaterialArray: "isMaterialArray", contentMaterialController: "contentMaterialController" }, exportAs: ["ngtContentMaterialController"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialController, decorators: [{
            type: Directive,
            args: [{
                    selector: `
    ngt-mesh,
    ngt-instanced-mesh,
    ngt-skinned-mesh,
    ngt-line,
    ngt-line-loop,
    ngt-line-segments,
    ngt-soba-plane,
    ngt-soba-box,
    ngt-soba-cylinder,
    ngt-soba-cone,
    ngt-soba-circle,
    ngt-soba-sphere,
    ngt-soba-tube,
    ngt-soba-torus,
    ngt-soba-tetrahedron,
    ngt-soba-ring,
    ngt-soba-polyhedron,
    ngt-soba-octahedron,
    ngt-soba-dodecahedron,
    ngt-soba-icosahedron,
    ngt-soba-extrude,
    ngt-soba-lathe,
    ngt-soba-torus-knot,
    ngt-soba-text,,
    ngt-points
  `,
                    exportAs: 'ngtContentMaterialController',
                }]
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { material: [{
                type: Input
            }], isMaterialArray: [{
                type: Input
            }], contentMaterialController: [{
                type: Input
            }] } });
export class NgtContentMaterialControllerModule {
}
NgtContentMaterialControllerModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialControllerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtContentMaterialControllerModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialControllerModule, declarations: [NgtContentMaterialController], exports: [NgtContentMaterialController] });
NgtContentMaterialControllerModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialControllerModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtContentMaterialControllerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtContentMaterialController],
                    exports: [NgtContentMaterialController],
                }]
        }] });
export const [NGT_CONTENT_MATERIAL_WATCHED_CONTROLLER, NGT_CONTENT_MATERIAL_CONTROLLER_PROVIDER,] = createControllerProviderFactory({
    controller: NgtContentMaterialController,
    watchedControllerTokenName: 'Watched ContentMaterialController',
});
