// @angular-three/core/esm2020/lib/controllers/controller.mjs
var _Controller_instances, _Controller_readyFn, _Controller_assignProps;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "tslib";
import { ChangeDetectorRef, Directive, InjectionToken, Input, NgZone, Optional, } from '@angular/core';
import { ReplaySubject, takeUntil } from 'rxjs';
import { NgtDestroyedService } from '../services/destroyed.service';
import * as i0 from "@angular/core";
export class Controller {
    constructor(ngZone) {
        this.ngZone = ngZone;
        _Controller_instances.add(this);
        this.disabled = false;
        this.change$ = new ReplaySubject(1);
        _Controller_readyFn.set(this, void 0);
    }
    set readyFn(v) {
        __classPrivateFieldSet(this, _Controller_readyFn, v, "f");
    }
    get readyFn() {
        return __classPrivateFieldGet(this, _Controller_readyFn, "f");
    }
    ngOnChanges(changes) {
        if (this.controller) {
            __classPrivateFieldGet(this, _Controller_instances, "m", _Controller_assignProps).call(this);
            this.controller.ngOnChanges(changes);
            this.change$.next(changes);
        }
        else {
            this.change$.next(changes);
        }
    }
    ngOnInit() {
        if (this.controller) {
            __classPrivateFieldGet(this, _Controller_instances, "m", _Controller_assignProps).call(this);
        }
    }
}
_Controller_readyFn = new WeakMap(), _Controller_instances = new WeakSet(), _Controller_assignProps = function _Controller_assignProps() {
    this.ngZone.runOutsideAngular(() => {
        this.props.forEach((prop) => {
            const selfController = this;
            const inputController = this.controller;
            selfController[prop] = selfController[prop] ?? inputController[prop];
        });
    });
};
Controller.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: Controller, deps: [{ token: i0.NgZone }], target: i0.ɵɵFactoryTarget.Directive });
Controller.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: Controller, inputs: { disabled: "disabled" }, usesOnChanges: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: Controller, decorators: [{
            type: Directive
        }], ctorParameters: function () { return [{ type: i0.NgZone }]; }, propDecorators: { disabled: [{
                type: Input
            }] } });
export function controllerFactory(newInstanceOnNull = false, controllerType) {
    return (controller, changeDetectorRef, destroyed) => {
        if (!controller) {
            return newInstanceOnNull ? new controllerType() : null;
        }
        controller.change$.pipe(takeUntil(destroyed)).subscribe(() => {
            changeDetectorRef.markForCheck();
        });
        return controller;
    };
}
export function createControllerProviderFactory({ watchedControllerTokenName, controller, newInstanceOnNull = false, }) {
    const watchedControllerToken = new InjectionToken(watchedControllerTokenName);
    const controllerProvider = [
        NgtDestroyedService,
        {
            provide: watchedControllerToken,
            deps: [
                [new Optional(), controller],
                ChangeDetectorRef,
                NgtDestroyedService,
            ],
            useFactory: controllerFactory(newInstanceOnNull, controller),
        },
    ];
    return [watchedControllerToken, controllerProvider];
}
