// @angular-three/core/esm2020/attributes/lib/uint32-buffer-attribute/uint32-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtUint32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Uint32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtUint32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtUint32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtUint32BufferAttribute, selector: "ngt-uint32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtUint32BufferAttribute,
        },
    ], exportAs: ["ngtUint32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-uint32-buffer-attribute',
                    exportAs: 'ngtUint32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtUint32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtUint32BufferAttributeModule {
}
NgtUint32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtUint32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, declarations: [NgtUint32BufferAttribute], exports: [NgtUint32BufferAttribute] });
NgtUint32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtUint32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtUint32BufferAttribute],
                    exports: [NgtUint32BufferAttribute],
                }]
        }] });
