// @angular-three/core/esm2020/attributes/lib/int32-buffer-attribute/int32-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInt32BufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.Int32BufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInt32BufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInt32BufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInt32BufferAttribute, selector: "ngt-int32-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInt32BufferAttribute,
        },
    ], exportAs: ["ngtInt32BufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-int32-buffer-attribute',
                    exportAs: 'ngtInt32BufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInt32BufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInt32BufferAttributeModule {
}
NgtInt32BufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInt32BufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, declarations: [NgtInt32BufferAttribute], exports: [NgtInt32BufferAttribute] });
NgtInt32BufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInt32BufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInt32BufferAttribute],
                    exports: [NgtInt32BufferAttribute],
                }]
        }] });
