// @angular-three/core/esm2020/audios/lib/audio/audio.directive.mjs
// GENERATED
import { NGT_AUDIO_CONTROLLER_PROVIDER, NGT_OBJECT_CONTROLLER_PROVIDER, NgtAudioControllerModule, NgtCommonAudio, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtAudio extends NgtCommonAudio {
    constructor() {
        super(...arguments);
        this.audioType = THREE.Audio;
    }
}
NgtAudio.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudio, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtAudio.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtAudio, selector: "ngt-audio", providers: [
        {
            provide: NgtCommonAudio,
            useExisting: NgtAudio,
        },
        NGT_AUDIO_CONTROLLER_PROVIDER,
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtAudio"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudio, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-audio',
                    exportAs: 'ngtAudio',
                    providers: [
                        {
                            provide: NgtCommonAudio,
                            useExisting: NgtAudio,
                        },
                        NGT_AUDIO_CONTROLLER_PROVIDER,
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }] });
export class NgtAudioModule {
}
NgtAudioModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtAudioModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioModule, declarations: [NgtAudio], exports: [NgtAudio, NgtAudioControllerModule, NgtObject3dControllerModule] });
NgtAudioModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioModule, imports: [NgtAudioControllerModule, NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtAudioModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtAudio],
                    exports: [NgtAudio, NgtAudioControllerModule, NgtObject3dControllerModule],
                }]
        }] });
