// @angular-three/core/esm2020/materials/lib/mesh-standard-material/mesh-standard-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshStandardMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshStandardMaterial;
    }
}
NgtMeshStandardMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshStandardMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshStandardMaterial, selector: "ngt-mesh-standard-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshStandardMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshStandardMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-standard-material',
                    exportAs: 'ngtMeshStandardMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshStandardMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshStandardMaterialModule {
}
NgtMeshStandardMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshStandardMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, declarations: [NgtMeshStandardMaterial], exports: [NgtMeshStandardMaterial] });
NgtMeshStandardMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshStandardMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshStandardMaterial],
                    exports: [NgtMeshStandardMaterial],
                }]
        }] });
