// @angular-three/core/esm2020/lights/lib/hemisphere-light-probe/hemisphere-light-probe.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtHemisphereLightProbe extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.HemisphereLightProbe;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtHemisphereLightProbe.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbe, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLightProbe.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLightProbe, selector: "ngt-hemisphere-light-probe", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtHemisphereLightProbe,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtHemisphereLightProbe"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbe, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-hemisphere-light-probe',
                    exportAs: 'ngtHemisphereLightProbe',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtHemisphereLightProbe,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtHemisphereLightProbeModule {
}
NgtHemisphereLightProbeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightProbeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, declarations: [NgtHemisphereLightProbe], exports: [NgtHemisphereLightProbe, NgtObject3dControllerModule] });
NgtHemisphereLightProbeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightProbeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLightProbe],
                    exports: [NgtHemisphereLightProbe, NgtObject3dControllerModule],
                }]
        }] });
