// @angular-three/core/esm2020/cameras/lib/perspective-camera/perspective-camera.directive.mjs
// GENERATED
import { NgtCommonCamera, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPerspectiveCamera extends NgtCommonCamera {
    constructor() {
        super(...arguments);
        this.cameraType = THREE.PerspectiveCamera;
    }
    set args(v) {
        this.cameraArgs = v;
    }
}
NgtPerspectiveCamera.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCamera, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPerspectiveCamera.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPerspectiveCamera, selector: "ngt-perspective-camera", inputs: { args: "args" }, providers: [
        {
            provide: NgtCommonCamera,
            useExisting: NgtPerspectiveCamera,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPerspectiveCamera"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCamera, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-perspective-camera',
                    exportAs: 'ngtPerspectiveCamera',
                    providers: [
                        {
                            provide: NgtCommonCamera,
                            useExisting: NgtPerspectiveCamera,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtPerspectiveCameraModule {
}
NgtPerspectiveCameraModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCameraModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPerspectiveCameraModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCameraModule, declarations: [NgtPerspectiveCamera], exports: [NgtPerspectiveCamera, NgtObject3dControllerModule] });
NgtPerspectiveCameraModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCameraModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPerspectiveCameraModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPerspectiveCamera],
                    exports: [NgtPerspectiveCamera, NgtObject3dControllerModule],
                }]
        }] });
