// @angular-three/core/esm2020/materials/lib/mesh-lambert-material/mesh-lambert-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtMeshLambertMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.MeshLambertMaterial;
    }
}
NgtMeshLambertMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtMeshLambertMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtMeshLambertMaterial, selector: "ngt-mesh-lambert-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtMeshLambertMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtMeshLambertMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-mesh-lambert-material',
                    exportAs: 'ngtMeshLambertMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtMeshLambertMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtMeshLambertMaterialModule {
}
NgtMeshLambertMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtMeshLambertMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, declarations: [NgtMeshLambertMaterial], exports: [NgtMeshLambertMaterial] });
NgtMeshLambertMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtMeshLambertMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtMeshLambertMaterial],
                    exports: [NgtMeshLambertMaterial],
                }]
        }] });
