// @angular-three/core/esm2020/materials/lib/shadow-material/shadow-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtShadowMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.ShadowMaterial;
    }
}
NgtShadowMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtShadowMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtShadowMaterial, selector: "ngt-shadow-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtShadowMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtShadowMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-shadow-material',
                    exportAs: 'ngtShadowMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtShadowMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtShadowMaterialModule {
}
NgtShadowMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtShadowMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, declarations: [NgtShadowMaterial], exports: [NgtShadowMaterial] });
NgtShadowMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShadowMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtShadowMaterial],
                    exports: [NgtShadowMaterial],
                }]
        }] });
