// @angular-three/core/esm2020/curves/lib/quadratic-bezier-curve3/quadratic-bezier-curve3.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtQuadraticBezierCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.QuadraticBezierCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtQuadraticBezierCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtQuadraticBezierCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtQuadraticBezierCurve3, selector: "ngt-quadratic-bezier-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtQuadraticBezierCurve3,
        },
    ], exportAs: ["ngtQuadraticBezierCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-quadratic-bezier-curve3',
                    exportAs: 'ngtQuadraticBezierCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtQuadraticBezierCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtQuadraticBezierCurve3Module {
}
NgtQuadraticBezierCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtQuadraticBezierCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, declarations: [NgtQuadraticBezierCurve3], exports: [NgtQuadraticBezierCurve3] });
NgtQuadraticBezierCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtQuadraticBezierCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtQuadraticBezierCurve3],
                    exports: [NgtQuadraticBezierCurve3],
                }]
        }] });
