// @angular-three/core/esm2020/lights/lib/spot-light/spot-light.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSpotLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.SpotLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtSpotLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSpotLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSpotLight, selector: "ngt-spot-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtSpotLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtSpotLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-spot-light',
                    exportAs: 'ngtSpotLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtSpotLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtSpotLightModule {
}
NgtSpotLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSpotLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, declarations: [NgtSpotLight], exports: [NgtSpotLight, NgtObject3dControllerModule] });
NgtSpotLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSpotLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSpotLight],
                    exports: [NgtSpotLight, NgtObject3dControllerModule],
                }]
        }] });
