// @angular-three/core/esm2020/geometries/lib/sphere-geometry/sphere-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtSphereGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.SphereGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtSphereGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtSphereGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtSphereGeometry, selector: "ngt-sphere-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtSphereGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtSphereGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-sphere-geometry',
                    exportAs: 'ngtSphereGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtSphereGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtSphereGeometryModule {
}
NgtSphereGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtSphereGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometryModule, declarations: [NgtSphereGeometry], exports: [NgtSphereGeometry] });
NgtSphereGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtSphereGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtSphereGeometry],
                    exports: [NgtSphereGeometry],
                }]
        }] });
