// @angular-three/core/esm2020/helpers/lib/plane-helper/plane-helper.directive.mjs
// GENERATED
import { NgtHelper, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtPlaneHelper extends NgtHelper {
    constructor() {
        super(...arguments);
        this.helperType = THREE.PlaneHelper;
    }
    set args(v) {
        this.helperArgs = v;
    }
}
NgtPlaneHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtPlaneHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtPlaneHelper, selector: "ngt-plane-helper", inputs: { args: "args" }, providers: [
        {
            provide: NgtHelper,
            useExisting: NgtPlaneHelper,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtPlaneHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-plane-helper',
                    exportAs: 'ngtPlaneHelper',
                    providers: [
                        {
                            provide: NgtHelper,
                            useExisting: NgtPlaneHelper,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtPlaneHelperModule {
}
NgtPlaneHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtPlaneHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, declarations: [NgtPlaneHelper], exports: [NgtPlaneHelper, NgtObject3dControllerModule] });
NgtPlaneHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtPlaneHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtPlaneHelper],
                    exports: [NgtPlaneHelper, NgtObject3dControllerModule],
                }]
        }] });
