// @rx-angular/cdk/esm2015/template/lib/template-manager.js
import { EMPTY, merge, of } from 'rxjs';
import { catchError, ignoreElements, switchMap, withLatestFrom, } from 'rxjs/operators';
import { onStrategy, strategyHandling } from '@rx-angular/cdk/render-strategies';
import { createErrorHandler } from './render-error';
import { getTNode, notifyAllParentsIfNeeded, notifyInjectingParentIfNeeded, templateHandling, } from './utils';
/**
 * @internal
 *
 * A factory function that returns a map of projections to turn a notification of a Observable (next, error, complete)
 *
 * @param customNextContext - projection function to provide custom properties as well as override existing
 */
export function notificationKindToViewContext(customNextContext) {
    // @TODO rethink overrides
    return {
        suspense: (notification) => {
            const $implicit = notification.value;
            return Object.assign({ $implicit, $suspense: true, $error: false, $complete: false }, customNextContext($implicit));
        },
        next: (notification) => {
            const $implicit = notification.value;
            return Object.assign({ $implicit, $suspense: false, $error: false, $complete: false }, customNextContext($implicit));
        },
        error: (notification) => {
            const $implicit = notification.value;
            return Object.assign({ $implicit, $complete: false, $error: notification.error, $suspense: false }, customNextContext($implicit));
        },
        complete: (notification) => {
            const $implicit = notification.value;
            return Object.assign({ $implicit, $error: false, $complete: true, $suspense: false }, customNextContext($implicit));
        },
    };
}
export function createTemplateManager(config) {
    const { renderSettings, notificationToTemplateName, templateSettings } = config;
    const { defaultStrategyName, strategies, cdRef: injectingViewCdRef, patchZone, parent, eRef, } = renderSettings;
    const errorHandler = createErrorHandler(renderSettings.errorHandler);
    const tNode = parent
        ? getTNode(injectingViewCdRef, eRef.nativeElement)
        : false;
    let activeTemplate;
    const strategyHandling$ = strategyHandling(defaultStrategyName, strategies);
    const templates = templateHandling(templateSettings.viewContainerRef);
    const viewContainerRef = templateSettings.viewContainerRef;
    const triggerHandling = config.templateTrigger$ || EMPTY;
    const getContext = notificationKindToViewContext(templateSettings.customContext || ((v) => { }));
    const workFactory = patchZone
        ? (work) => patchZone.run(work)
        : (work) => work();
    return {
        addTemplateRef: templates.add,
        // addTrigger: triggerHandling.next,
        nextStrategy: strategyHandling$.next,
        render(values$) {
            return values$.pipe(
            /* tslint:disable */
            // mergeWith(triggerHandling.trigger$ || EMPTY),
            /* tslint:enable */
            withLatestFrom(strategyHandling$.strategy$), 
            // Cancel old renders
            switchMap(([notification, strategy]) => {
                const kind = notification.kind;
                const value = notification.value;
                const templateName = notificationToTemplateName[kind](value, templates);
                const template = templates.get(templateName);
                const isNewTemplate = activeTemplate !== templateName;
                const notifyParent = isNewTemplate && parent;
                return merge(onStrategy(value, strategy, (v, work, options) => {
                    const context = getContext[kind](notification);
                    if (isNewTemplate) {
                        // template has changed (undefined => next; suspense => next; ...)
                        // handle remove & insert
                        // remove current view if there is any
                        if (viewContainerRef.length > 0) {
                            // patch removal if needed
                            workFactory(() => viewContainerRef.clear());
                        }
                        // create new view if any
                        if (template) {
                            // createEmbeddedView is already patched, no need for workFactory
                            workFactory(() => templates.createEmbeddedView(templateName, context));
                        }
                    }
                    else if (template) {
                        // template didn't change, update it
                        // handle update
                        const view = viewContainerRef.get(0);
                        Object.keys(context).forEach((k) => {
                            view.context[k] = context[k];
                        });
                        // update view context, patch if needed
                        workFactory(() => work(view, options.scope, notification));
                    }
                    activeTemplate = templateName;
                }
                // we don't need to specify any scope here. The template manager is the only one
                // who will call `viewRef#detectChanges` on any of the templates it manages.
                // whenever a new value comes in, any pre-scheduled work of this taskManager will
                // be nooped before a new work will be scheduled. This happens because of the implementation
                // of `StrategyCredential#behavior`
                ).pipe(notifyAllParentsIfNeeded(tNode, injectingViewCdRef, strategy, () => notifyParent)), notifyInjectingParentIfNeeded(injectingViewCdRef, strategy, isNewTemplate).pipe(ignoreElements())).pipe(catchError((e) => {
                    errorHandler.handleError(e);
                    return of(e);
                }));
            }));
        },
    };
}
