// @angular-three/core/esm2020/geometries/lib/instanced-buffer-geometry/instanced-buffer-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInstancedBufferGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.InstancedBufferGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtInstancedBufferGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInstancedBufferGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInstancedBufferGeometry, selector: "ngt-instanced-buffer-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtInstancedBufferGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtInstancedBufferGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-instanced-buffer-geometry',
                    exportAs: 'ngtInstancedBufferGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtInstancedBufferGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInstancedBufferGeometryModule {
}
NgtInstancedBufferGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInstancedBufferGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometryModule, declarations: [NgtInstancedBufferGeometry], exports: [NgtInstancedBufferGeometry] });
NgtInstancedBufferGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInstancedBufferGeometry],
                    exports: [NgtInstancedBufferGeometry],
                }]
        }] });
