// @angular-three/core/esm2020/curves/lib/cubic-bezier-curve3/cubic-bezier-curve3.directive.mjs
// GENERATED
import { NgtCurve } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtCubicBezierCurve3 extends NgtCurve {
    constructor() {
        super(...arguments);
        this.curveType = THREE.CubicBezierCurve3;
    }
    set args(v) {
        this.curveArgs = v;
    }
}
NgtCubicBezierCurve3.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtCubicBezierCurve3.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtCubicBezierCurve3, selector: "ngt-cubic-bezier-curve3", inputs: { args: "args" }, providers: [
        {
            provide: NgtCurve,
            useExisting: NgtCubicBezierCurve3,
        },
    ], exportAs: ["ngtCubicBezierCurve3"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-cubic-bezier-curve3',
                    exportAs: 'ngtCubicBezierCurve3',
                    providers: [
                        {
                            provide: NgtCurve,
                            useExisting: NgtCubicBezierCurve3,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtCubicBezierCurve3Module {
}
NgtCubicBezierCurve3Module.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtCubicBezierCurve3Module.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, declarations: [NgtCubicBezierCurve3], exports: [NgtCubicBezierCurve3] });
NgtCubicBezierCurve3Module.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtCubicBezierCurve3Module, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtCubicBezierCurve3],
                    exports: [NgtCubicBezierCurve3],
                }]
        }] });
