// @angular-three/core/esm2020/lines/lib/line-segments/line-segments.directive.mjs
// GENERATED
import { NgtCommonLine, NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER, NGT_OBJECT_TYPE, NgtMaterialGeometryControllerModule, } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLineSegments extends NgtCommonLine {
}
NgtLineSegments.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegments, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineSegments.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineSegments, selector: "ngt-line-segments", providers: [
        {
            provide: NgtCommonLine,
            useExisting: NgtLineSegments,
        },
        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
        {
            provide: NGT_OBJECT_TYPE,
            useValue: THREE.LineSegments,
        },
    ], exportAs: ["ngtLineSegments"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegments, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-segments',
                    exportAs: 'ngtLineSegments',
                    providers: [
                        {
                            provide: NgtCommonLine,
                            useExisting: NgtLineSegments,
                        },
                        NGT_MATERIAL_GEOMETRY_CONTROLLER_PROVIDER,
                        {
                            provide: NGT_OBJECT_TYPE,
                            useValue: THREE.LineSegments,
                        },
                    ],
                }]
        }] });
export class NgtLineSegmentsModule {
}
NgtLineSegmentsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineSegmentsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, declarations: [NgtLineSegments], exports: [NgtLineSegments, NgtMaterialGeometryControllerModule] });
NgtLineSegmentsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, imports: [NgtMaterialGeometryControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineSegmentsModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineSegments],
                    exports: [NgtLineSegments, NgtMaterialGeometryControllerModule],
                }]
        }] });
