// @angular-three/core/esm2020/geometries/lib/dodecahedron-geometry/dodecahedron-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDodecahedronGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.DodecahedronGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtDodecahedronGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDodecahedronGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDodecahedronGeometry, selector: "ngt-dodecahedron-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtDodecahedronGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtDodecahedronGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-dodecahedron-geometry',
                    exportAs: 'ngtDodecahedronGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtDodecahedronGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtDodecahedronGeometryModule {
}
NgtDodecahedronGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDodecahedronGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometryModule, declarations: [NgtDodecahedronGeometry], exports: [NgtDodecahedronGeometry] });
NgtDodecahedronGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDodecahedronGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDodecahedronGeometry],
                    exports: [NgtDodecahedronGeometry],
                }]
        }] });
