// @angular-three/core/esm2020/attributes/lib/instanced-buffer-attribute/instanced-buffer-attribute.directive.mjs
// GENERATED
import { NgtAttribute } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtInstancedBufferAttribute extends NgtAttribute {
    constructor() {
        super(...arguments);
        this.attributeType = THREE.InstancedBufferAttribute;
    }
    set args(v) {
        this.attributeArgs = v;
    }
}
NgtInstancedBufferAttribute.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttribute, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtInstancedBufferAttribute.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtInstancedBufferAttribute, selector: "ngt-instanced-buffer-attribute", inputs: { args: "args" }, providers: [
        {
            provide: NgtAttribute,
            useExisting: NgtInstancedBufferAttribute,
        },
    ], exportAs: ["ngtInstancedBufferAttribute"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttribute, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-instanced-buffer-attribute',
                    exportAs: 'ngtInstancedBufferAttribute',
                    providers: [
                        {
                            provide: NgtAttribute,
                            useExisting: NgtInstancedBufferAttribute,
                        },
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtInstancedBufferAttributeModule {
}
NgtInstancedBufferAttributeModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtInstancedBufferAttributeModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, declarations: [NgtInstancedBufferAttribute], exports: [NgtInstancedBufferAttribute] });
NgtInstancedBufferAttributeModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtInstancedBufferAttributeModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtInstancedBufferAttribute],
                    exports: [NgtInstancedBufferAttribute],
                }]
        }] });
