// @angular-three/core/esm2020/geometries/lib/shape-geometry/shape-geometry.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtGeometry } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtShapeGeometry extends NgtGeometry {
    constructor() {
        super(...arguments);
        this.geometryType = THREE.ShapeGeometry;
    }
    set args(v) {
        this.geometryArgs = v;
    }
}
NgtShapeGeometry.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometry, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtShapeGeometry.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtShapeGeometry, selector: "ngt-shape-geometry", inputs: { args: "args" }, providers: [
        {
            provide: NgtGeometry,
            useExisting: NgtShapeGeometry,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtShapeGeometry"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometry, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-shape-geometry',
                    exportAs: 'ngtShapeGeometry',
                    providers: [
                        {
                            provide: NgtGeometry,
                            useExisting: NgtShapeGeometry,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtShapeGeometryModule {
}
NgtShapeGeometryModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometryModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtShapeGeometryModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometryModule, declarations: [NgtShapeGeometry], exports: [NgtShapeGeometry] });
NgtShapeGeometryModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometryModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtShapeGeometryModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtShapeGeometry],
                    exports: [NgtShapeGeometry],
                }]
        }] });
