// @angular-three/core/esm2020/helpers/lib/directional-light-helper/directional-light-helper.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtObjectHelper, } from '@angular-three/core';
import { Directive, Input, NgModule } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtDirectionalLightHelper extends NgtObjectHelper {
    constructor() {
        super(...arguments);
        this.objectHelperType = THREE.DirectionalLightHelper;
    }
    set ngtDirectionalLightHelper(v) {
        if (v) {
            this.objectHelperArgs = v;
        }
    }
}
NgtDirectionalLightHelper.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelper, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtDirectionalLightHelper.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtDirectionalLightHelper, selector: "[ngtDirectionalLightHelper]", inputs: { ngtDirectionalLightHelper: "ngtDirectionalLightHelper" }, providers: [
        {
            provide: NgtObjectHelper,
            useExisting: NgtDirectionalLightHelper,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtDirectionalLightHelper"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelper, decorators: [{
            type: Directive,
            args: [{
                    selector: '[ngtDirectionalLightHelper]',
                    exportAs: 'ngtDirectionalLightHelper',
                    providers: [
                        {
                            provide: NgtObjectHelper,
                            useExisting: NgtDirectionalLightHelper,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }], propDecorators: { ngtDirectionalLightHelper: [{
                type: Input
            }] } });
export class NgtDirectionalLightHelperModule {
}
NgtDirectionalLightHelperModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtDirectionalLightHelperModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, declarations: [NgtDirectionalLightHelper], exports: [NgtDirectionalLightHelper] });
NgtDirectionalLightHelperModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtDirectionalLightHelperModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtDirectionalLightHelper],
                    exports: [NgtDirectionalLightHelper],
                }]
        }] });
