// @angular-three/core/esm2020/lights/lib/hemisphere-light/hemisphere-light.directive.mjs
// GENERATED
import { NgtLight, NGT_OBJECT_CONTROLLER_PROVIDER, NgtObject3dControllerModule, } from '@angular-three/core';
import { NgModule, Directive, Input } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtHemisphereLight extends NgtLight {
    constructor() {
        super(...arguments);
        this.lightType = THREE.HemisphereLight;
    }
    set args(v) {
        this.lightArgs = v;
    }
}
NgtHemisphereLight.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLight, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtHemisphereLight.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtHemisphereLight, selector: "ngt-hemisphere-light", inputs: { args: "args" }, providers: [
        {
            provide: NgtLight,
            useExisting: NgtHemisphereLight,
        },
        NGT_OBJECT_CONTROLLER_PROVIDER,
    ], exportAs: ["ngtHemisphereLight"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLight, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-hemisphere-light',
                    exportAs: 'ngtHemisphereLight',
                    providers: [
                        {
                            provide: NgtLight,
                            useExisting: NgtHemisphereLight,
                        },
                        NGT_OBJECT_CONTROLLER_PROVIDER,
                    ],
                }]
        }], propDecorators: { args: [{
                type: Input
            }] } });
export class NgtHemisphereLightModule {
}
NgtHemisphereLightModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtHemisphereLightModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, declarations: [NgtHemisphereLight], exports: [NgtHemisphereLight, NgtObject3dControllerModule] });
NgtHemisphereLightModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, imports: [NgtObject3dControllerModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtHemisphereLightModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtHemisphereLight],
                    exports: [NgtHemisphereLight, NgtObject3dControllerModule],
                }]
        }] });
