// @angular-three/core/esm2020/materials/lib/line-basic-material/line-basic-material.directive.mjs
// GENERATED
import { NGT_OBJECT_3D_PROVIDER, NgtMaterial } from '@angular-three/core';
import { NgModule, Directive } from '@angular/core';
import * as THREE from 'three';
import * as i0 from "@angular/core";
export class NgtLineBasicMaterial extends NgtMaterial {
    constructor() {
        super(...arguments);
        this.materialType = THREE.LineBasicMaterial;
    }
}
NgtLineBasicMaterial.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterial, deps: null, target: i0.ɵɵFactoryTarget.Directive });
NgtLineBasicMaterial.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "12.0.0", version: "13.1.1", type: NgtLineBasicMaterial, selector: "ngt-line-basic-material", providers: [
        {
            provide: NgtMaterial,
            useExisting: NgtLineBasicMaterial,
        },
        NGT_OBJECT_3D_PROVIDER,
    ], exportAs: ["ngtLineBasicMaterial"], usesInheritance: true, ngImport: i0 });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterial, decorators: [{
            type: Directive,
            args: [{
                    selector: 'ngt-line-basic-material',
                    exportAs: 'ngtLineBasicMaterial',
                    providers: [
                        {
                            provide: NgtMaterial,
                            useExisting: NgtLineBasicMaterial,
                        },
                        NGT_OBJECT_3D_PROVIDER,
                    ],
                }]
        }] });
export class NgtLineBasicMaterialModule {
}
NgtLineBasicMaterialModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgtLineBasicMaterialModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, declarations: [NgtLineBasicMaterial], exports: [NgtLineBasicMaterial] });
NgtLineBasicMaterialModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.1.1", ngImport: i0, type: NgtLineBasicMaterialModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgtLineBasicMaterial],
                    exports: [NgtLineBasicMaterial],
                }]
        }] });
